{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "name": "2021-04-11 RA NLP drug label processing Program.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.7.4"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "-yGLO7WOl-aP"
      },
      "source": [
        "# First of all: on Colabl, switch runtime type to \"GPU\"\n",
        "# -- under Runtime tag, click 'change runtime type', select 'GPU' for Hardware accelerator. "
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wN23caxVlwS2",
        "outputId": "626e38b2-e6a1-4ac1-fd5a-7e20f5094ba9"
      },
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Mounted at /content/drive\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Z5H_LXWKP0v4"
      },
      "source": [
        "!pip install scispacy\n",
        "#!pip uninstall spacy\n",
        "!pip install -U spacy==2.3.1"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uIxUVkrIlr07"
      },
      "source": [
        "# Installations needed\n",
        "!pip install transformers\n",
        "!pip install knockknock==0.1.7\n",
        "!pip install sentencepiece"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-zLldfwUP0v5"
      },
      "source": [
        "# install the pre-tained model\n",
        "# online\n",
        "!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_core_sci_sm-0.2.4.tar.gz\n",
        "!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.3.0/en_ner_craft_md-0.3.0.tar.gz\n",
        "!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.3.0/en_ner_jnlpba_md-0.3.0.tar.gz\n",
        "!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.3.0/en_ner_bc5cdr_md-0.3.0.tar.gz\n",
        "!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.3.0/en_ner_bionlp13cg_md-0.3.0.tar.gz"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1gjgZdWoP0vz",
        "outputId": "1aff8400-dae0-4e97-ae1b-825d4b6db7eb"
      },
      "source": [
        "# import necessary packages\n",
        "from lxml import etree\n",
        "import requests\n",
        "import lxml.html\n",
        "from urllib.parse import urlparse\n",
        "import nltk\n",
        "from nltk import tokenize\n",
        "import re\n",
        "nltk.download('punkt')\n",
        "\n",
        "import scispacy\n",
        "import spacy\n",
        "from spacy import displacy\n",
        "import en_core_sci_sm\n",
        "import en_ner_craft_md\n",
        "import en_ner_jnlpba_md\n",
        "import en_ner_bc5cdr_md\n",
        "import en_ner_bionlp13cg_md"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[nltk_data] Downloading package punkt to /root/nltk_data...\n",
            "[nltk_data]   Unzipping tokenizers/punkt.zip.\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/spacy/util.py:275: UserWarning: [W031] Model 'en_core_sci_sm' (0.2.4) requires spaCy v2.2 and is incompatible with the current spaCy version (2.3.1). This may lead to unexpected results or runtime errors. To resolve this, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate\n",
            "  warnings.warn(warn_msg)\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jV0pICPGP0v6",
        "outputId": "cdf5cf95-9b06-4e2f-ae62-b2dcd4aa5922"
      },
      "source": [
        "# load the pre-tained NER models\n",
        "nlp = en_core_sci_sm.load()\n",
        "nlp_craft = en_ner_craft_md.load()\n",
        "nlp_jnlpha = en_ner_jnlpba_md.load()\n",
        "nlp_bc5cdr = en_ner_bc5cdr_md.load()\n",
        "nlp_bionlp13cg = en_ner_bionlp13cg_md.load()"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/spacy/util.py:275: UserWarning: [W031] Model 'en_core_sci_sm' (0.2.4) requires spaCy v2.2 and is incompatible with the current spaCy version (2.3.1). This may lead to unexpected results or runtime errors. To resolve this, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate\n",
            "  warnings.warn(warn_msg)\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GBZEr125lvZF"
      },
      "source": [
        "# Imports needed\n",
        "import os, re, torch, html, tempfile, copy, json, math, shutil, tarfile, tempfile, sys, random, pickle\n",
        "from torch import nn\n",
        "from torch.nn import functional as F\n",
        "from torch.nn import CrossEntropyLoss, ReLU\n",
        "from torch.optim import Adam\n",
        "from torch.utils.data import TensorDataset, DataLoader, RandomSampler, SequentialSampler\n",
        "from keras.preprocessing.sequence import pad_sequences\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.metrics import f1_score, classification_report\n",
        "from transformers import RobertaTokenizer, BertForTokenClassification, BertTokenizer, BertConfig, BertModel, WordpieceTokenizer, XLNetTokenizer\n",
        "from transformers.file_utils import cached_path\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from tqdm import tqdm\n",
        "import matplotlib.pyplot as plt\n",
        "%matplotlib inline\n",
        "from scipy import stats\n",
        "from knockknock import email_sender, telegram_sender"
      ],
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "qU3hQ3dOl4QG"
      },
      "source": [
        "# load Negbert pre-train models\n",
        "modelCue = torch.load('/content/drive/MyDrive/RA_NLP_drug_label/Negation_extraction/modelCue') # path to the model file on Google drive\n",
        "modelScope = torch.load('/content/drive/MyDrive/RA_NLP_drug_label/Negation_extraction/modelScope') # path to the model file on Google drive"
      ],
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pOG0lpwavMlq"
      },
      "source": [
        "Aditya and Suraj's NegBERT programs."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "N20pvbh7lyao"
      },
      "source": [
        "MAX_LEN = 128\n",
        "bs = 8\n",
        "EPOCHS = 60\n",
        "PATIENCE = 6\n",
        "INITIAL_LEARNING_RATE = 3e-5\n",
        "NUM_RUNS = 1 #Number of times to run the training and evaluation code\n",
        "\n",
        "CUE_MODEL = 'bert-base-uncased'\n",
        "SCOPE_MODEL = 'xlnet-base-cased'\n",
        "SCOPE_METHOD = 'augment' # Options: augment, replace\n",
        "F1_METHOD = 'average' # Options: average, first_token\n",
        "TASK = 'negation' # Options: negation, speculation\n",
        "SUBTASK = 'scope_resolution' # Options: cue_detection, scope_resolution\n",
        "TRAIN_DATASETS = ['bioscope_abstracts','bioscope_full_papers']\n",
        "TEST_DATASETS = ['bioscope_full_papers','bioscope_abstracts']\n",
        "\n",
        "\n",
        "BERT_PRETRAINED_CONFIG_ARCHIVE_MAP = {\n",
        "    'bert-base-uncased': \"https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-config.json\"\n",
        "}\n",
        "\n",
        "BERT_PRETRAINED_MODEL_ARCHIVE_MAP = {\n",
        "    'bert-base-uncased': \"https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-pytorch_model.bin\"\n",
        "}\n",
        "\n",
        "ROBERTA_PRETRAINED_MODEL_ARCHIVE_MAP = {\n",
        "    'roberta-base': \"https://s3.amazonaws.com/models.huggingface.co/bert/roberta-base-pytorch_model.bin\"\n",
        "}\n",
        "\n",
        "ROBERTA_PRETRAINED_CONFIG_ARCHIVE_MAP = {\n",
        "    'roberta-base': \"https://s3.amazonaws.com/models.huggingface.co/bert/roberta-base-config.json\"\n",
        "}\n",
        "\n",
        "XLNET_PRETRAINED_CONFIG_ARCHIVE_MAP = {\n",
        "    'xlnet-base-cased': \"https://s3.amazonaws.com/models.huggingface.co/bert/xlnet-base-cased-config.json\"\n",
        "}\n",
        "\n",
        "XLNET_PRETRAINED_MODEL_ARCHIVE_MAP = {\n",
        "    'xlnet-base-cased': \"https://s3.amazonaws.com/models.huggingface.co/bert/xlnet-base-cased-pytorch_model.bin\"\n",
        "}\n",
        "\n",
        "TF_WEIGHTS_NAME = 'model.ckpt'\n",
        "CONFIG_NAME = \"config.json\"\n",
        "WEIGHTS_NAME = \"pytorch_model.bin\"\n",
        "\n",
        "device = torch.device(\"cuda\")\n",
        "n_gpu = torch.cuda.device_count()\n",
        "\n",
        "class Data:\n",
        "    def __init__(self, file, dataset_name = 'sfu', frac_no_cue_sents = 1.0):\n",
        "        '''\n",
        "        file: The path of the data file.\n",
        "        dataset_name: The name of the dataset to be preprocessed. Values supported: sfu, bioscope, starsem.\n",
        "        frac_no_cue_sents: The fraction of sentences to be included in the data object which have no negation/speculation cues.\n",
        "        '''\n",
        "        def starsem(f_path, cue_sents_only=False, frac_no_cue_sents = 1.0):\n",
        "            raw_data = open(f_path)\n",
        "            sentence = []\n",
        "            labels = []\n",
        "            label = []\n",
        "            scope_sents = []\n",
        "            data_scope = []\n",
        "            scope = []\n",
        "            scope_cues = []\n",
        "            data = []\n",
        "            cue_only_data = []\n",
        "            \n",
        "            for line in raw_data:\n",
        "                label = []\n",
        "                sentence = []\n",
        "                tokens = line.strip().split()\n",
        "                if len(tokens)==8: #This line has no cues\n",
        "                        sentence.append(tokens[3])\n",
        "                        label.append(3) #Not a cue\n",
        "                        for line in raw_data:\n",
        "                            tokens = line.strip().split()\n",
        "                            if len(tokens)==0:\n",
        "                                break\n",
        "                            else:\n",
        "                                sentence.append(tokens[3])\n",
        "                                label.append(3)\n",
        "                        cue_only_data.append([sentence, label])\n",
        "                        \n",
        "                    \n",
        "                else: #The line has 1 or more cues\n",
        "                    num_cues = (len(tokens)-7)//3\n",
        "                    #cue_count+=num_cues\n",
        "                    scope = [[] for i in range(num_cues)]\n",
        "                    label = [[],[]] #First list is the real labels, second list is to modify if it is a multi-word cue.\n",
        "                    label[0].append(3) #Generally not a cue, if it is will be set ahead.\n",
        "                    label[1].append(-1) #Since not a cue, for now.\n",
        "                    for i in range(num_cues):\n",
        "                        if tokens[7+3*i] != '_': #Cue field is active\n",
        "                            if tokens[8+3*i] != '_': #Check for affix\n",
        "                                label[0][-1] = 0 #Affix\n",
        "                                affix_list.append(tokens[7+3*i])\n",
        "                                label[1][-1] = i #Cue number\n",
        "                                #sentence.append(tokens[7+3*i])\n",
        "                                #new_word = '##'+tokens[8+3*i]\n",
        "                            else:\n",
        "                                label[0][-1] = 1 #Maybe a normal or multiword cue. The next few words will determine which.\n",
        "                                label[1][-1] = i #Which cue field, for multiword cue altering.\n",
        "                                \n",
        "                        if tokens[8+3*i] != '_':\n",
        "                            scope[i].append(1)\n",
        "                        else:\n",
        "                            scope[i].append(0)\n",
        "                    sentence.append(tokens[3])\n",
        "                    for line in raw_data:\n",
        "                        tokens = line.strip().split()\n",
        "                        if len(tokens)==0:\n",
        "                            break\n",
        "                        else:\n",
        "                            sentence.append(tokens[3])\n",
        "                            label[0].append(3) #Generally not a cue, if it is will be set ahead.\n",
        "                            label[1].append(-1) #Since not a cue, for now.   \n",
        "                            for i in range(num_cues):\n",
        "                                if tokens[7+3*i] != '_': #Cue field is active\n",
        "                                    if tokens[8+3*i] != '_': #Check for affix\n",
        "                                        label[0][-1] = 0 #Affix\n",
        "                                        label[1][-1] = i #Cue number\n",
        "                                    else:\n",
        "                                        label[0][-1] = 1 #Maybe a normal or multiword cue. The next few words will determine which.\n",
        "                                        label[1][-1] = i #Which cue field, for multiword cue altering.\n",
        "                                if tokens[8+3*i] != '_':\n",
        "                                    scope[i].append(1)\n",
        "                                else:\n",
        "                                    scope[i].append(0)\n",
        "                    for i in range(num_cues):\n",
        "                        indices = [index for index,j in enumerate(label[1]) if i==j]\n",
        "                        count = len(indices)\n",
        "                        if count>1:\n",
        "                            for j in indices:\n",
        "                                label[0][j] = 2\n",
        "                    for i in range(num_cues):\n",
        "                        sc = []\n",
        "                        for a,b in zip(label[0],label[1]):\n",
        "                            if i==b:\n",
        "                                sc.append(a)\n",
        "                            else:\n",
        "                                sc.append(3)\n",
        "                        scope_cues.append(sc)\n",
        "                        scope_sents.append(sentence)\n",
        "                        data_scope.append(scope[i])\n",
        "                    labels.append(label[0])\n",
        "                    data.append(sentence)\n",
        "            cue_only_samples = random.sample(cue_only_data, k=int(frac_no_cue_sents*len(cue_only_data)))\n",
        "            cue_only_sents = [i[0] for i in cue_only_samples]\n",
        "            cue_only_cues = [i[1] for i in cue_only_samples]\n",
        "            starsem_cues = (data+cue_only_sents,labels+cue_only_cues)\n",
        "            starsem_scopes = (scope_sents, scope_cues, data_scope)\n",
        "            return [starsem_cues, starsem_scopes]\n",
        "            \n",
        "        def bioscope(f_path, cue_sents_only=False, frac_no_cue_sents = 1.0):\n",
        "            file = open(f_path, encoding = 'utf-8')\n",
        "            sentences = []\n",
        "            for s in file:\n",
        "                sentences+=re.split(\"(<.*?>)\", html.unescape(s))\n",
        "            cue_sentence = []\n",
        "            cue_cues = []\n",
        "            cue_only_data = []\n",
        "            scope_cues = []\n",
        "            scope_scopes = []\n",
        "            scope_sentence = []\n",
        "            sentence = []\n",
        "            cue = {}\n",
        "            scope = {}\n",
        "            in_scope = []\n",
        "            in_cue = []\n",
        "            word_num = 0\n",
        "            c_idx = []\n",
        "            s_idx = []\n",
        "            in_sentence = 0\n",
        "            for token in sentences:\n",
        "                if token == '':\n",
        "                    continue\n",
        "                elif '<sentence' in token:\n",
        "                    in_sentence = 1\n",
        "                elif '<cue' in token:\n",
        "                    if TASK in token:\n",
        "                        in_cue.append(str(re.split('(ref=\".*?\")',token)[1][4:]))\n",
        "                        c_idx.append(str(re.split('(ref=\".*?\")',token)[1][4:]))\n",
        "                        if c_idx[-1] not in cue.keys():\n",
        "                            cue[c_idx[-1]] = []\n",
        "                elif '</cue' in token:\n",
        "                    in_cue = in_cue[:-1]\n",
        "                elif '<xcope' in token:\n",
        "                    #print(re.split('(id=\".*?\")',token)[1][3:])\n",
        "                    in_scope.append(str(re.split('(id=\".*?\")',token)[1][3:]))\n",
        "                    s_idx.append(str(re.split('(id=\".*?\")',token)[1][3:]))\n",
        "                    scope[s_idx[-1]] = []\n",
        "                elif '</xcope' in token:\n",
        "                    in_scope = in_scope[:-1]\n",
        "                elif '</sentence' in token:\n",
        "                    #print(cue, scope)\n",
        "                    if len(cue.keys())==0:\n",
        "                        cue_only_data.append([sentence, [3]*len(sentence)])\n",
        "                    else:\n",
        "                        cue_sentence.append(sentence)\n",
        "                        cue_cues.append([3]*len(sentence))\n",
        "                        for i in cue.keys():\n",
        "                            scope_sentence.append(sentence)\n",
        "                            scope_cues.append([3]*len(sentence))\n",
        "                            if len(cue[i])==1:\n",
        "                                cue_cues[-1][cue[i][0]] = 1\n",
        "                                scope_cues[-1][cue[i][0]] = 1\n",
        "                            else:\n",
        "                                for c in cue[i]:\n",
        "                                    cue_cues[-1][c] = 2\n",
        "                                    scope_cues[-1][c] = 2\n",
        "                            scope_scopes.append([0]*len(sentence))\n",
        "\n",
        "                            if i in scope.keys():\n",
        "                                for s in scope[i]:\n",
        "                                    scope_scopes[-1][s] = 1\n",
        "\n",
        "                    sentence = []\n",
        "                    cue = {}\n",
        "                    scope = {}\n",
        "                    in_scope = []\n",
        "                    in_cue = []\n",
        "                    word_num = 0\n",
        "                    in_sentence = 0\n",
        "                    c_idx = []\n",
        "                    s_idx = []\n",
        "                elif '<' not in token:\n",
        "                    if in_sentence==1:\n",
        "                        words = token.split()\n",
        "                        sentence+=words\n",
        "                        if len(in_cue)!=0:\n",
        "                            for i in in_cue:\n",
        "                                cue[i]+=[word_num+i for i in range(len(words))]\n",
        "                        elif len(in_scope)!=0:\n",
        "                            for i in in_scope:\n",
        "                                scope[i]+=[word_num+i for i in range(len(words))]\n",
        "                        word_num+=len(words)\n",
        "            cue_only_samples = random.sample(cue_only_data, k=int(frac_no_cue_sents*len(cue_only_data)))\n",
        "            cue_only_sents = [i[0] for i in cue_only_samples]\n",
        "            cue_only_cues = [i[1] for i in cue_only_samples]\n",
        "            return [(cue_sentence+cue_only_sents, cue_cues+cue_only_cues),(scope_sentence, scope_cues, scope_scopes)]\n",
        "        \n",
        "        def sfu_review(f_path, cue_sents_only=False, frac_no_cue_sents = 1.0):\n",
        "            file = open(f_path, encoding = 'utf-8')\n",
        "            sentences = []\n",
        "            for s in file:\n",
        "                sentences+=re.split(\"(<.*?>)\", html.unescape(s))\n",
        "            cue_sentence = []\n",
        "            cue_cues = []\n",
        "            scope_cues = []\n",
        "            scope_scopes = []\n",
        "            scope_sentence = []\n",
        "            sentence = []\n",
        "            cue = {}\n",
        "            scope = {}\n",
        "            in_scope = []\n",
        "            in_cue = []\n",
        "            word_num = 0\n",
        "            c_idx = []\n",
        "            cue_only_data = []\n",
        "            s_idx = []\n",
        "            in_word = 0\n",
        "            for token in sentences:\n",
        "                if token == '':\n",
        "                    continue\n",
        "                elif token == '<W>':\n",
        "                    in_word = 1\n",
        "                elif token == '</W>':\n",
        "                    in_word = 0\n",
        "                    word_num += 1\n",
        "                elif '<cue' in token:\n",
        "                    if TASK in token:\n",
        "                        in_cue.append(int(re.split('(ID=\".*?\")',token)[1][4:-1]))\n",
        "                        c_idx.append(int(re.split('(ID=\".*?\")',token)[1][4:-1]))\n",
        "                        if c_idx[-1] not in cue.keys():\n",
        "                            cue[c_idx[-1]] = []\n",
        "                elif '</cue' in token:\n",
        "                    in_cue = in_cue[:-1]\n",
        "                elif '<xcope' in token:\n",
        "                    continue\n",
        "                elif '</xcope' in token:\n",
        "                    in_scope = in_scope[:-1]\n",
        "                elif '<ref' in token:\n",
        "                    in_scope.append([int(i) for i in re.split('(SRC=\".*?\")',token)[1][5:-1].split(' ')])\n",
        "                    s_idx.append([int(i) for i in re.split('(SRC=\".*?\")',token)[1][5:-1].split(' ')])\n",
        "                    for i in s_idx[-1]:\n",
        "                        scope[i] = []\n",
        "                elif '</SENTENCE' in token:\n",
        "                    if len(cue.keys())==0:\n",
        "                        cue_only_data.append([sentence, [3]*len(sentence)])\n",
        "                    else:\n",
        "                        cue_sentence.append(sentence)\n",
        "                        cue_cues.append([3]*len(sentence))\n",
        "                        for i in cue.keys():\n",
        "                            scope_sentence.append(sentence)\n",
        "                            scope_cues.append([3]*len(sentence))\n",
        "                            if len(cue[i])==1:\n",
        "                                cue_cues[-1][cue[i][0]] = 1\n",
        "                                scope_cues[-1][cue[i][0]] = 1\n",
        "                            else:\n",
        "                                for c in cue[i]:\n",
        "                                    cue_cues[-1][c] = 2\n",
        "                                    scope_cues[-1][c] = 2\n",
        "                            scope_scopes.append([0]*len(sentence))\n",
        "                            if i in scope.keys():\n",
        "                                for s in scope[i]:\n",
        "                                    scope_scopes[-1][s] = 1\n",
        "                    sentence = []\n",
        "                    cue = {}\n",
        "                    scope = {}\n",
        "                    in_scope = []\n",
        "                    in_cue = []\n",
        "                    word_num = 0\n",
        "                    in_word = 0\n",
        "                    c_idx = []\n",
        "                    s_idx = []\n",
        "                elif '<' not in token:\n",
        "                    if in_word == 1:\n",
        "                        if len(in_cue)!=0:\n",
        "                            for i in in_cue:\n",
        "                                cue[i].append(word_num)\n",
        "                        if len(in_scope)!=0:\n",
        "                            for i in in_scope:\n",
        "                                for j in i:\n",
        "                                    scope[j].append(word_num)\n",
        "                        sentence.append(token)\n",
        "            cue_only_samples = random.sample(cue_only_data, k=int(frac_no_cue_sents*len(cue_only_data)))\n",
        "            cue_only_sents = [i[0] for i in cue_only_samples]\n",
        "            cue_only_cues = [i[1] for i in cue_only_samples]\n",
        "            return [(cue_sentence+cue_only_sents, cue_cues+cue_only_cues),(scope_sentence, scope_cues, scope_scopes)]\n",
        "        \n",
        "        \n",
        "        if dataset_name == 'bioscope':\n",
        "            ret_val = bioscope(file, frac_no_cue_sents=frac_no_cue_sents)\n",
        "            self.cue_data = Cues(ret_val[0])\n",
        "            self.scope_data = Scopes(ret_val[1])\n",
        "        elif dataset_name == 'sfu':\n",
        "            sfu_cues = [[], []]\n",
        "            sfu_scopes = [[], [], []]\n",
        "            for dir_name in os.listdir(file):\n",
        "                if '.' not in dir_name:\n",
        "                    for f_name in os.listdir(file+\"//\"+dir_name):\n",
        "                        r_val = sfu_review(file+\"//\"+dir_name+'//'+f_name, frac_no_cue_sents=frac_no_cue_sents)\n",
        "                        sfu_cues = [a+b for a,b in zip(sfu_cues, r_val[0])]\n",
        "                        sfu_scopes = [a+b for a,b in zip(sfu_scopes, r_val[1])]\n",
        "            self.cue_data = Cues(sfu_cues)\n",
        "            self.scope_data = Scopes(sfu_scopes)\n",
        "        elif dataset_name == 'starsem':\n",
        "            if TASK == 'negation':\n",
        "                ret_val = starsem(file, frac_no_cue_sents=frac_no_cue_sents)\n",
        "                self.cue_data = Cues(ret_val[0])\n",
        "                self.scope_data = Scopes(ret_val[1])\n",
        "            else:\n",
        "                raise ValueError(\"Starsem 2012 dataset only supports negation annotations\")\n",
        "        else:\n",
        "            raise ValueError(\"Supported Dataset types are:\\n\\tbioscope\\n\\tsfu\\n\\tconll_cue\")\n",
        "    \n",
        "    def get_cue_dataloader(self, val_size = 0.15, test_size = 0.15, other_datasets = []):\n",
        "        '''\n",
        "        This function returns the dataloader for the cue detection.\n",
        "        val_size: The size of the validation dataset (Fraction between 0 to 1)\n",
        "        test_size: The size of the test dataset (Fraction between 0 to 1)\n",
        "        other_datasets: Other datasets to use to get one combined train dataloader\n",
        "        Returns: train_dataloader, list of validation dataloaders, list of test dataloaders\n",
        "        '''\n",
        "        do_lower_case = True\n",
        "        if 'uncased' not in CUE_MODEL:\n",
        "            do_lower_case = False\n",
        "        if 'xlnet' in CUE_MODEL:\n",
        "            tokenizer = XLNetTokenizer.from_pretrained(CUE_MODEL, do_lower_case=do_lower_case, cache_dir='xlnet_tokenizer')\n",
        "        elif 'roberta' in CUE_MODEL:\n",
        "            tokenizer = RobertaTokenizer.from_pretrained(CUE_MODEL, do_lower_case=do_lower_case, cache_dir='roberta_tokenizer')\n",
        "        elif 'bert' in CUE_MODEL:\n",
        "            tokenizer = BertTokenizer.from_pretrained(CUE_MODEL, do_lower_case=do_lower_case, cache_dir='bert_tokenizer')\n",
        "        def preprocess_data(obj, tokenizer):\n",
        "            dl_sents = obj.cue_data.sentences\n",
        "            dl_cues = obj.cue_data.cues\n",
        "                \n",
        "            sentences = [\" \".join(sent) for sent in dl_sents]\n",
        "\n",
        "            mytexts = []\n",
        "            mylabels = []\n",
        "            mymasks = []\n",
        "            if do_lower_case == True:\n",
        "                sentences_clean = [sent.lower() for sent in sentences]\n",
        "            else:\n",
        "                sentences_clean = sentences\n",
        "            for sent, tags in zip(sentences_clean,dl_cues):\n",
        "                new_tags = []\n",
        "                new_text = []\n",
        "                new_masks = []\n",
        "                for word, tag in zip(sent.split(),tags):\n",
        "                    sub_words = tokenizer._tokenize(word)\n",
        "                    for count, sub_word in enumerate(sub_words):\n",
        "                        if type(tag)!=int:\n",
        "                            raise ValueError(tag)\n",
        "                        mask = 1\n",
        "                        if count > 0:\n",
        "                            mask = 0\n",
        "                        new_masks.append(mask)\n",
        "                        new_tags.append(tag)\n",
        "                        new_text.append(sub_word)\n",
        "                mymasks.append(new_masks)\n",
        "                mytexts.append(new_text)\n",
        "                mylabels.append(new_tags)\n",
        "\n",
        "            \n",
        "            input_ids = pad_sequences([[tokenizer._convert_token_to_id(word) for word in txt] for txt in mytexts],\n",
        "                                      maxlen=MAX_LEN, dtype=\"long\", truncating=\"post\", padding=\"post\").tolist()\n",
        "\n",
        "            tags = pad_sequences(mylabels,\n",
        "                                maxlen=MAX_LEN, value=4, padding=\"post\",\n",
        "                                dtype=\"long\", truncating=\"post\").tolist()\n",
        "            \n",
        "            mymasks = pad_sequences(mymasks, maxlen=MAX_LEN, value=0, padding='post', dtype='long', truncating='post').tolist()\n",
        "            \n",
        "            attention_masks = [[float(i>0) for i in ii] for ii in input_ids]\n",
        "            \n",
        "            random_state = np.random.randint(1,2019)\n",
        "\n",
        "            tra_inputs, test_inputs, tra_tags, test_tags = train_test_split(input_ids, tags, test_size=test_size, random_state = random_state)\n",
        "            tra_masks, test_masks, _, _ = train_test_split(attention_masks, input_ids, test_size=test_size, random_state = random_state)\n",
        "            tra_mymasks, test_mymasks, _, _ = train_test_split(mymasks, input_ids, test_size=test_size, random_state = random_state)\n",
        "            \n",
        "            random_state_2 = np.random.randint(1,2019)\n",
        "\n",
        "            tr_inputs, val_inputs, tr_tags, val_tags = train_test_split(tra_inputs, tra_tags, test_size=(val_size/(1-test_size)), random_state = random_state_2)\n",
        "            tr_masks, val_masks, _, _ = train_test_split(tra_masks, tra_inputs, test_size=(val_size/(1-test_size)), random_state = random_state_2)\n",
        "            tr_mymasks, val_mymasks, _, _ = train_test_split(tra_mymasks, tra_inputs, test_size=(val_size/(1-test_size)), random_state = random_state_2)\n",
        "            return [tr_inputs, tr_tags, tr_masks, tr_mymasks], [val_inputs, val_tags, val_masks, val_mymasks], [test_inputs, test_tags, test_masks, test_mymasks]\n",
        "\n",
        "        tr_inputs = []\n",
        "        tr_tags = []\n",
        "        tr_masks = []\n",
        "        tr_mymasks = []\n",
        "        val_inputs = [[] for i in range(len(other_datasets)+1)]\n",
        "        test_inputs = [[] for i in range(len(other_datasets)+1)]\n",
        "\n",
        "        train_ret_val, val_ret_val, test_ret_val = preprocess_data(self, tokenizer)\n",
        "        tr_inputs+=train_ret_val[0]\n",
        "        tr_tags+=train_ret_val[1]\n",
        "        tr_masks+=train_ret_val[2]\n",
        "        tr_mymasks+=train_ret_val[3]\n",
        "        val_inputs[0].append(val_ret_val[0])\n",
        "        val_inputs[0].append(val_ret_val[1])\n",
        "        val_inputs[0].append(val_ret_val[2])\n",
        "        val_inputs[0].append(val_ret_val[3])\n",
        "        test_inputs[0].append(test_ret_val[0])\n",
        "        test_inputs[0].append(test_ret_val[1])\n",
        "        test_inputs[0].append(test_ret_val[2])\n",
        "        test_inputs[0].append(test_ret_val[3])\n",
        "        \n",
        "        for idx, arg in enumerate(other_datasets, 1):\n",
        "            train_ret_val, val_ret_val, test_ret_val = preprocess_data(arg, tokenizer)\n",
        "            tr_inputs+=train_ret_val[0]\n",
        "            tr_tags+=train_ret_val[1]\n",
        "            tr_masks+=train_ret_val[2]\n",
        "            tr_mymasks+=train_ret_val[3]\n",
        "            val_inputs[idx].append(val_ret_val[0])\n",
        "            val_inputs[idx].append(val_ret_val[1])\n",
        "            val_inputs[idx].append(val_ret_val[2])\n",
        "            val_inputs[idx].append(val_ret_val[3])\n",
        "            test_inputs[idx].append(test_ret_val[0])\n",
        "            test_inputs[idx].append(test_ret_val[1])\n",
        "            test_inputs[idx].append(test_ret_val[2])\n",
        "            test_inputs[idx].append(test_ret_val[3])\n",
        "        \n",
        "        tr_inputs = torch.LongTensor(tr_inputs)\n",
        "        tr_tags = torch.LongTensor(tr_tags)\n",
        "        tr_masks = torch.LongTensor(tr_masks)\n",
        "        tr_mymasks = torch.LongTensor(tr_mymasks)\n",
        "        val_inputs = [[torch.LongTensor(i) for i in j] for j in val_inputs]\n",
        "        test_inputs = [[torch.LongTensor(i) for i in j] for j in test_inputs]\n",
        "\n",
        "        train_data = TensorDataset(tr_inputs, tr_masks, tr_tags, tr_mymasks)\n",
        "        train_sampler = RandomSampler(train_data)\n",
        "        train_dataloader = DataLoader(train_data, sampler=train_sampler, batch_size=bs)\n",
        "\n",
        "        val_dataloaders = []\n",
        "        for i,j,k,l in val_inputs:\n",
        "            val_data = TensorDataset(i, k, j, l)\n",
        "            val_sampler = RandomSampler(val_data)\n",
        "            val_dataloaders.append(DataLoader(val_data, sampler=val_sampler, batch_size=bs))\n",
        "\n",
        "        test_dataloaders = []\n",
        "        for i,j,k,l in test_inputs:\n",
        "            test_data = TensorDataset(i, k, j, l)\n",
        "            test_sampler = RandomSampler(test_data)\n",
        "            test_dataloaders.append(DataLoader(test_data, sampler=test_sampler, batch_size=bs))\n",
        "\n",
        "        return train_dataloader, val_dataloaders, test_dataloaders\n",
        "\n",
        "    def get_scope_dataloader(self, val_size = 0.15, test_size=0.15, other_datasets = []):\n",
        "        '''\n",
        "        This function returns the dataloader for the cue detection.\n",
        "        val_size: The size of the validation dataset (Fraction between 0 to 1)\n",
        "        test_size: The size of the test dataset (Fraction between 0 to 1)\n",
        "        other_datasets: Other datasets to use to get one combined train dataloader\n",
        "        Returns: train_dataloader, list of validation dataloaders, list of test dataloaders\n",
        "        '''\n",
        "        method = SCOPE_METHOD\n",
        "        do_lower_case = True\n",
        "        if 'uncased' not in SCOPE_MODEL:\n",
        "            do_lower_case = False\n",
        "        if 'xlnet' in SCOPE_MODEL:\n",
        "            tokenizer = XLNetTokenizer.from_pretrained(SCOPE_MODEL, do_lower_case=do_lower_case, cache_dir='xlnet_tokenizer')\n",
        "        elif 'roberta' in SCOPE_MODEL:\n",
        "            tokenizer = RobertaTokenizer.from_pretrained(SCOPE_MODEL, do_lower_case=do_lower_case, cache_dir='roberta_tokenizer')\n",
        "        elif 'bert' in SCOPE_MODEL:\n",
        "            tokenizer = BertTokenizer.from_pretrained(SCOPE_MODEL, do_lower_case=do_lower_case, cache_dir='bert_tokenizer')\n",
        "        def preprocess_data(obj, tokenizer_obj):\n",
        "            dl_sents = obj.scope_data.sentences\n",
        "            dl_cues = obj.scope_data.cues\n",
        "            dl_scopes = obj.scope_data.scopes\n",
        "            \n",
        "            sentences = [\" \".join([s for s in sent]) for sent in dl_sents]\n",
        "            mytexts = []\n",
        "            mylabels = []\n",
        "            mycues = []\n",
        "            mymasks = []\n",
        "            if do_lower_case == True:\n",
        "                sentences_clean = [sent.lower() for sent in sentences]\n",
        "            else:\n",
        "                sentences_clean = sentences\n",
        "            \n",
        "            for sent, tags, cues in zip(sentences_clean,dl_scopes, dl_cues):\n",
        "                new_tags = []\n",
        "                new_text = []\n",
        "                new_cues = []\n",
        "                new_masks = []\n",
        "                for word, tag, cue in zip(sent.split(),tags,cues):\n",
        "                    sub_words = tokenizer_obj._tokenize(word)\n",
        "                    for count, sub_word in enumerate(sub_words):\n",
        "                        mask = 1\n",
        "                        if count > 0:\n",
        "                            mask = 0\n",
        "                        new_masks.append(mask)\n",
        "                        new_tags.append(tag)\n",
        "                        new_cues.append(cue)\n",
        "                        new_text.append(sub_word)\n",
        "                mymasks.append(new_masks)\n",
        "                mytexts.append(new_text)\n",
        "                mylabels.append(new_tags)\n",
        "                mycues.append(new_cues)\n",
        "            final_sentences = []\n",
        "            final_labels = []\n",
        "            final_masks = []\n",
        "            if method == 'replace':\n",
        "                for sent,cues in zip(mytexts, mycues):\n",
        "                    temp_sent = []\n",
        "                    for token,cue in zip(sent,cues):\n",
        "                        if cue==3:\n",
        "                            temp_sent.append(token)\n",
        "                        else:\n",
        "                            temp_sent.append(f'[unused{cue+1}]')\n",
        "                    final_sentences.append(temp_sent)\n",
        "                final_labels = mylabels\n",
        "                final_masks = mymasks\n",
        "            elif method == 'augment':\n",
        "                for sent,cues,labels,masks in zip(mytexts, mycues, mylabels, mymasks):\n",
        "                    temp_sent = []\n",
        "                    temp_label = []\n",
        "                    temp_masks = []\n",
        "                    first_part = 0\n",
        "                    for token,cue,label,mask in zip(sent,cues,labels,masks):\n",
        "                        if cue!=3:\n",
        "                            if first_part == 0:\n",
        "                                first_part = 1\n",
        "                                temp_sent.append(f'[unused{cue+1}]')\n",
        "                                temp_masks.append(1)\n",
        "                                temp_label.append(label)\n",
        "                                temp_sent.append(token)\n",
        "                                temp_masks.append(0)\n",
        "                                temp_label.append(label)\n",
        "                                continue\n",
        "                            temp_sent.append(f'[unused{cue+1}]')\n",
        "                            temp_masks.append(0)\n",
        "                            temp_label.append(label)\n",
        "                        else:\n",
        "                            first_part = 0\n",
        "                        temp_masks.append(mask)\n",
        "                        temp_sent.append(token)\n",
        "                        temp_label.append(label)\n",
        "                    final_sentences.append(temp_sent)\n",
        "                    final_labels.append(temp_label)\n",
        "                    final_masks.append(temp_masks)\n",
        "            else:\n",
        "                raise ValueError(\"Supported methods for scope detection are:\\nreplace\\naugment\")\n",
        "            input_ids = pad_sequences([[tokenizer_obj._convert_token_to_id(word) for word in txt] for txt in final_sentences],\n",
        "                                      maxlen=MAX_LEN, dtype=\"long\", truncating=\"post\", padding=\"post\").tolist()\n",
        "\n",
        "            tags = pad_sequences(final_labels,\n",
        "                                maxlen=MAX_LEN, value=0, padding=\"post\",\n",
        "                                dtype=\"long\", truncating=\"post\").tolist()\n",
        "            \n",
        "            final_masks = pad_sequences(final_masks,\n",
        "                                maxlen=MAX_LEN, value=0, padding=\"post\",\n",
        "                                dtype=\"long\", truncating=\"post\").tolist()\n",
        "\n",
        "            attention_masks = [[float(i>0) for i in ii] for ii in input_ids]\n",
        "            \n",
        "            random_state = np.random.randint(1,2019)\n",
        "\n",
        "            tra_inputs, test_inputs, tra_tags, test_tags = train_test_split(input_ids, tags, test_size=test_size, random_state = random_state)\n",
        "            tra_masks, test_masks, _, _ = train_test_split(attention_masks, input_ids, test_size=test_size, random_state = random_state)\n",
        "            tra_mymasks, test_mymasks, _, _ = train_test_split(final_masks, input_ids, test_size=test_size, random_state = random_state)\n",
        "            \n",
        "            random_state_2 = np.random.randint(1,2019)\n",
        "\n",
        "            tr_inputs, val_inputs, tr_tags, val_tags = train_test_split(tra_inputs, tra_tags, test_size=(val_size/(1-test_size)), random_state = random_state_2)\n",
        "            tr_masks, val_masks, _, _ = train_test_split(tra_masks, tra_inputs, test_size=(val_size/(1-test_size)), random_state = random_state_2)\n",
        "            tr_mymasks, val_mymasks, _, _ = train_test_split(tra_mymasks, tra_inputs, test_size=(val_size/(1-test_size)), random_state = random_state_2)\n",
        "\n",
        "            return [tr_inputs, tr_tags, tr_masks, tr_mymasks], [val_inputs, val_tags, val_masks, val_mymasks], [test_inputs, test_tags, test_masks, test_mymasks]\n",
        "\n",
        "        tr_inputs = []\n",
        "        tr_tags = []\n",
        "        tr_masks = []\n",
        "        tr_mymasks = []\n",
        "        val_inputs = [[] for i in range(len(other_datasets)+1)]\n",
        "        test_inputs = [[] for i in range(len(other_datasets)+1)]\n",
        "\n",
        "        train_ret_val, val_ret_val, test_ret_val = preprocess_data(self, tokenizer)\n",
        "        tr_inputs+=train_ret_val[0]\n",
        "        tr_tags+=train_ret_val[1]\n",
        "        tr_masks+=train_ret_val[2]\n",
        "        tr_mymasks+=train_ret_val[3]\n",
        "        val_inputs[0].append(val_ret_val[0])\n",
        "        val_inputs[0].append(val_ret_val[1])\n",
        "        val_inputs[0].append(val_ret_val[2])\n",
        "        val_inputs[0].append(val_ret_val[3])\n",
        "        test_inputs[0].append(test_ret_val[0])\n",
        "        test_inputs[0].append(test_ret_val[1])\n",
        "        test_inputs[0].append(test_ret_val[2])\n",
        "        test_inputs[0].append(test_ret_val[3])\n",
        "        \n",
        "        for idx, arg in enumerate(other_datasets, 1):\n",
        "            train_ret_val, val_ret_val, test_ret_val = preprocess_data(arg, tokenizer)\n",
        "            tr_inputs+=train_ret_val[0]\n",
        "            tr_tags+=train_ret_val[1]\n",
        "            tr_masks+=train_ret_val[2]\n",
        "            tr_mymasks+=train_ret_val[3]\n",
        "            val_inputs[idx].append(val_ret_val[0])\n",
        "            val_inputs[idx].append(val_ret_val[1])\n",
        "            val_inputs[idx].append(val_ret_val[2])\n",
        "            val_inputs[idx].append(val_ret_val[3])\n",
        "            test_inputs[idx].append(test_ret_val[0])\n",
        "            test_inputs[idx].append(test_ret_val[1])\n",
        "            test_inputs[idx].append(test_ret_val[2])\n",
        "            test_inputs[idx].append(test_ret_val[3])\n",
        "\n",
        "        tr_inputs = torch.LongTensor(tr_inputs)\n",
        "        tr_tags = torch.LongTensor(tr_tags)\n",
        "        tr_masks = torch.LongTensor(tr_masks)\n",
        "        tr_mymasks = torch.LongTensor(tr_mymasks)\n",
        "        val_inputs = [[torch.LongTensor(i) for i in j] for j in val_inputs]\n",
        "        test_inputs = [[torch.LongTensor(i) for i in j] for j in test_inputs]\n",
        "\n",
        "        train_data = TensorDataset(tr_inputs, tr_masks, tr_tags, tr_mymasks)\n",
        "        train_sampler = RandomSampler(train_data)\n",
        "        train_dataloader = DataLoader(train_data, sampler=train_sampler, batch_size=bs)\n",
        "\n",
        "        val_dataloaders = []\n",
        "        for i,j,k,l in val_inputs:\n",
        "            val_data = TensorDataset(i, k, j, l)\n",
        "            val_sampler = RandomSampler(val_data)\n",
        "            val_dataloaders.append(DataLoader(val_data, sampler=val_sampler, batch_size=bs))\n",
        "\n",
        "        test_dataloaders = []\n",
        "        for i,j,k,l in test_inputs:\n",
        "            test_data = TensorDataset(i, k, j, l)\n",
        "            test_sampler = RandomSampler(test_data)\n",
        "            test_dataloaders.append(DataLoader(test_data, sampler=test_sampler, batch_size=bs))\n",
        "\n",
        "        return train_dataloader, val_dataloaders, test_dataloaders\n",
        "\n",
        "class CustomData:\n",
        "    def __init__(self, sentences, cues = None):\n",
        "        self.sentences = sentences\n",
        "        self.cues = cues\n",
        "    def get_cue_dataloader(self):\n",
        "        do_lower_case = True\n",
        "        if 'uncased' not in CUE_MODEL:\n",
        "            do_lower_case = False\n",
        "        if 'xlnet' in CUE_MODEL:\n",
        "            tokenizer = XLNetTokenizer.from_pretrained(CUE_MODEL, do_lower_case=do_lower_case, cache_dir='xlnet_tokenizer')\n",
        "        elif 'roberta' in CUE_MODEL:\n",
        "            tokenizer = RobertaTokenizer.from_pretrained(CUE_MODEL, do_lower_case=do_lower_case, cache_dir='roberta_tokenizer')\n",
        "        elif 'bert' in CUE_MODEL:\n",
        "            tokenizer = BertTokenizer.from_pretrained(CUE_MODEL, do_lower_case=do_lower_case, cache_dir='bert_tokenizer')\n",
        "        \n",
        "        dl_sents = self.sentences    \n",
        "        sentences = dl_sents # sentences = [\" \".join(sent) for sent in dl_sents]\n",
        "\n",
        "        mytexts = []\n",
        "        mylabels = []\n",
        "        mymasks = []\n",
        "        if do_lower_case == True:\n",
        "            sentences_clean = [sent.lower() for sent in sentences]\n",
        "        else:\n",
        "            sentences_clean = sentences\n",
        "        for sent in sentences_clean:\n",
        "            new_text = []\n",
        "            new_masks = []\n",
        "            for word in sent.split():\n",
        "                sub_words = tokenizer._tokenize(word)\n",
        "                for count, sub_word in enumerate(sub_words):\n",
        "                    mask = 1\n",
        "                    if count > 0:\n",
        "                        mask = 0\n",
        "                    new_masks.append(mask)\n",
        "                    new_text.append(sub_word)\n",
        "            mymasks.append(new_masks)\n",
        "            mytexts.append(new_text)\n",
        "            \n",
        "        input_ids = pad_sequences([tokenizer.convert_tokens_to_ids(txt) for txt in mytexts],\n",
        "                                  maxlen=MAX_LEN, dtype=\"long\", truncating=\"post\", padding=\"post\")\n",
        "\n",
        "        mymasks = pad_sequences(mymasks, maxlen=MAX_LEN, value=0, padding='post', dtype='long', truncating='post').tolist()\n",
        "\n",
        "        attention_masks = [[float(i>0) for i in ii] for ii in input_ids]\n",
        "\n",
        "        inputs = torch.LongTensor(input_ids)\n",
        "        masks = torch.LongTensor(attention_masks)\n",
        "        mymasks = torch.LongTensor(mymasks)\n",
        "\n",
        "        data = TensorDataset(inputs, masks, mymasks)\n",
        "        dataloader = DataLoader(data, batch_size=bs)\n",
        "\n",
        "        return dataloader\n",
        "    \n",
        "    def get_scope_dataloader(self, cues = None):\n",
        "        if cues != None:\n",
        "            self.cues = cues\n",
        "        if self.cues == None:\n",
        "            raise ValueError(\"Need Cues Data to Generate the Scope Dataloader\")\n",
        "        method = SCOPE_METHOD\n",
        "        do_lower_case = True\n",
        "        if 'uncased' not in SCOPE_MODEL:\n",
        "            do_lower_case = False\n",
        "        if 'xlnet' in SCOPE_MODEL:\n",
        "            tokenizer = XLNetTokenizer.from_pretrained(SCOPE_MODEL, do_lower_case=do_lower_case, cache_dir='xlnet_tokenizer')\n",
        "        elif 'roberta' in SCOPE_MODEL:\n",
        "            tokenizer = RobertaTokenizer.from_pretrained(SCOPE_MODEL, do_lower_case=do_lower_case, cache_dir='roberta_tokenizer')\n",
        "        elif 'bert' in SCOPE_MODEL:\n",
        "            tokenizer = BertTokenizer.from_pretrained(SCOPE_MODEL, do_lower_case=do_lower_case, cache_dir='bert_tokenizer')\n",
        "        dl_sents = self.sentences\n",
        "        dl_cues = self.cues\n",
        "        \n",
        "        sentences = dl_sents\n",
        "        mytexts = []\n",
        "        mycues = []\n",
        "        mymasks = []\n",
        "        if do_lower_case == True:\n",
        "            sentences_clean = [sent.lower() for sent in sentences]\n",
        "        else:\n",
        "            sentences_clean = sentences\n",
        "        \n",
        "        for sent, cues in zip(sentences_clean, dl_cues):\n",
        "            new_text = []\n",
        "            new_cues = []\n",
        "            new_masks = []\n",
        "            for word, cue in zip(sent.split(), cues):\n",
        "                sub_words = tokenizer._tokenize(word)\n",
        "                for count, sub_word in enumerate(sub_words):\n",
        "                    mask = 1\n",
        "                    if count > 0:\n",
        "                        mask = 0\n",
        "                    new_masks.append(mask)\n",
        "                    new_cues.append(cue)\n",
        "                    new_text.append(sub_word)\n",
        "            mymasks.append(new_masks)\n",
        "            mytexts.append(new_text)\n",
        "            mycues.append(new_cues)\n",
        "        final_sentences = []\n",
        "        final_masks = []\n",
        "        if method == 'replace':\n",
        "            for sent,cues in zip(mytexts, mycues):\n",
        "                temp_sent = []\n",
        "                for token,cue in zip(sent,cues):\n",
        "                    if cue==3:\n",
        "                        temp_sent.append(token)\n",
        "                    else:\n",
        "                        temp_sent.append(f'[unused{cue+1}]')\n",
        "                final_sentences.append(temp_sent)\n",
        "            final_masks = mymasks\n",
        "        elif method == 'augment':\n",
        "            for sent,cues,masks in zip(mytexts, mycues, mymasks):\n",
        "                temp_sent = []\n",
        "                temp_masks = []\n",
        "                first_part = 0\n",
        "                for token,cue,mask in zip(sent,cues,masks):\n",
        "                    if cue!=3:\n",
        "                        if first_part == 0:\n",
        "                            first_part = 1\n",
        "                            temp_sent.append(f'[unused{cue+1}]')\n",
        "                            temp_masks.append(1)\n",
        "                            #temp_label.append(label)\n",
        "                            temp_sent.append(token)\n",
        "                            temp_masks.append(0)\n",
        "                            #temp_label.append(label)\n",
        "                            continue\n",
        "                        temp_sent.append(f'[unused{cue+1}]')\n",
        "                        temp_masks.append(0)\n",
        "                        #temp_label.append(label)\n",
        "                    else:\n",
        "                        first_part = 0\n",
        "                    temp_masks.append(mask)\n",
        "                    temp_sent.append(token)\n",
        "                final_sentences.append(temp_sent)\n",
        "                final_masks.append(temp_masks)\n",
        "        else:\n",
        "            raise ValueError(\"Supported methods for scope detection are:\\nreplace\\naugment\")\n",
        "    \n",
        "        input_ids = pad_sequences([tokenizer.convert_tokens_to_ids(txt) for txt in final_sentences],\n",
        "                                  maxlen=MAX_LEN, dtype=\"long\", truncating=\"post\", padding=\"post\")\n",
        "        \n",
        "        final_masks = pad_sequences(final_masks,\n",
        "                                maxlen=MAX_LEN, value=0, padding=\"post\",\n",
        "                                dtype=\"long\", truncating=\"post\").tolist()\n",
        "\n",
        "        attention_masks = [[float(i>0) for i in ii] for ii in input_ids]\n",
        "\n",
        "        inputs = torch.LongTensor(input_ids)\n",
        "        masks = torch.LongTensor(attention_masks)\n",
        "        final_masks = torch.LongTensor(final_masks)\n",
        "\n",
        "        data = TensorDataset(inputs, masks, final_masks)\n",
        "        dataloader = DataLoader(data, batch_size=bs)\n",
        "        #print(final_sentences, mycues)\n",
        "\n",
        "        return dataloader\n",
        "\n",
        "def load_tf_weights_in_bert(model, config, tf_checkpoint_path):\n",
        "    \"\"\" Load tf checkpoints in a pytorch model.\n",
        "    \"\"\"\n",
        "    import re\n",
        "    import tensorflow as tf\n",
        "    tf_path = os.path.abspath(tf_checkpoint_path)\n",
        "    # Load weights from TF model\n",
        "    init_vars = tf.train.list_variables(tf_path)\n",
        "    names = []\n",
        "    arrays = []\n",
        "    for name, shape in init_vars:\n",
        "        array = tf.train.load_variable(tf_path, name)\n",
        "        names.append(name)\n",
        "        arrays.append(array)\n",
        "\n",
        "    for name, array in zip(names, arrays):\n",
        "        name = name.split('/')\n",
        "        # adam_v and adam_m are variables used in AdamWeightDecayOptimizer to calculated m and v\n",
        "        # which are not required for using pretrained model\n",
        "        if any(n in [\"adam_v\", \"adam_m\", \"global_step\"] for n in name):\n",
        "            continue\n",
        "        pointer = model\n",
        "        for m_name in name:\n",
        "            if re.fullmatch(r'[A-Za-z]+_\\d+', m_name):\n",
        "                l = re.split(r'_(\\d+)', m_name)\n",
        "            else:\n",
        "                l = [m_name]\n",
        "            if l[0] == 'kernel' or l[0] == 'gamma':\n",
        "                pointer = getattr(pointer, 'weight')\n",
        "            elif l[0] == 'output_bias' or l[0] == 'beta':\n",
        "                pointer = getattr(pointer, 'bias')\n",
        "            elif l[0] == 'output_weights':\n",
        "                pointer = getattr(pointer, 'weight')\n",
        "            elif l[0] == 'squad':\n",
        "                pointer = getattr(pointer, 'classifier')\n",
        "            else:\n",
        "                try:\n",
        "                    pointer = getattr(pointer, l[0])\n",
        "                except AttributeError:\n",
        "                    continue\n",
        "            if len(l) >= 2:\n",
        "                num = int(l[1])\n",
        "                pointer = pointer[num]\n",
        "        if m_name[-11:] == '_embeddings':\n",
        "            pointer = getattr(pointer, 'weight')\n",
        "        elif m_name == 'kernel':\n",
        "            array = np.transpose(array)\n",
        "        try:\n",
        "            assert pointer.shape == array.shape\n",
        "        except AssertionError as e:\n",
        "            e.args += (pointer.shape, array.shape)\n",
        "            raise\n",
        "        pointer.data = torch.from_numpy(array)\n",
        "    return model\n",
        "\n",
        "def load_tf_weights_in_xlnet(model, config, tf_path):\n",
        "    \"\"\" Load tf checkpoints in a pytorch model\n",
        "    \"\"\"\n",
        "    import numpy as np\n",
        "    import tensorflow as tf\n",
        "    # Load weights from TF model\n",
        "    init_vars = tf.train.list_variables(tf_path)\n",
        "    tf_weights = {}\n",
        "    for name, shape in init_vars:\n",
        "        array = tf.train.load_variable(tf_path, name)\n",
        "        tf_weights[name] = array\n",
        "\n",
        "    # Build TF to PyTorch weights loading map\n",
        "    tf_to_pt_map = build_tf_xlnet_to_pytorch_map(model, config, tf_weights)\n",
        "\n",
        "    for name, pointer in tf_to_pt_map.items():\n",
        "        if name not in tf_weights:\n",
        "            continue\n",
        "        array = tf_weights[name]\n",
        "        # adam_v and adam_m are variables used in AdamWeightDecayOptimizer to calculated m and v\n",
        "        # which are not required for using pretrained model\n",
        "        if 'kernel' in name and ('ff' in name or 'summary' in name or 'logit' in name):\n",
        "            array = np.transpose(array)\n",
        "        if isinstance(pointer, list):\n",
        "            # Here we will split the TF weigths\n",
        "            assert len(pointer) == array.shape[0]\n",
        "            for i, p_i in enumerate(pointer):\n",
        "                arr_i = array[i, ...]\n",
        "                try:\n",
        "                    assert p_i.shape == arr_i.shape\n",
        "                except AssertionError as e:\n",
        "                    e.args += (p_i.shape, arr_i.shape)\n",
        "                    raise\n",
        "                p_i.data = torch.from_numpy(arr_i)\n",
        "        else:\n",
        "            try:\n",
        "                assert pointer.shape == array.shape\n",
        "            except AssertionError as e:\n",
        "                e.args += (pointer.shape, array.shape)\n",
        "                raise\n",
        "            pointer.data = torch.from_numpy(array)\n",
        "        tf_weights.pop(name, None)\n",
        "        tf_weights.pop(name + '/Adam', None)\n",
        "        tf_weights.pop(name + '/Adam_1', None)\n",
        "\n",
        "    return model\n",
        "\n",
        "def gelu(x):\n",
        "    \"\"\" Original Implementation of the gelu activation function in Google Bert repo when initially created.\n",
        "        For information: OpenAI GPT's gelu is slightly different (and gives slightly different results):\n",
        "        0.5 * x * (1 + torch.tanh(math.sqrt(2 / math.pi) * (x + 0.044715 * torch.pow(x, 3))))\n",
        "        Also see https://arxiv.org/abs/1606.08415\n",
        "    \"\"\"\n",
        "    return x * 0.5 * (1.0 + torch.erf(x / math.sqrt(2.0)))\n",
        "\n",
        "def gelu_new(x):\n",
        "    \"\"\" Implementation of the gelu activation function currently in Google Bert repo (identical to OpenAI GPT).\n",
        "        Also see https://arxiv.org/abs/1606.08415\n",
        "    \"\"\"\n",
        "    return 0.5 * x * (1 + torch.tanh(math.sqrt(2 / math.pi) * (x + 0.044715 * torch.pow(x, 3))))\n",
        "\n",
        "def swish(x):\n",
        "    return x * torch.sigmoid(x)\n",
        "\n",
        "ACT2FN = {\"gelu\": gelu, \"relu\": torch.nn.functional.relu, \"swish\": swish, \"gelu_new\": gelu_new}\n",
        "\n",
        "class PretrainedConfig(object):\n",
        "    r\"\"\" Base class for all configuration classes.\n",
        "        Handles a few parameters common to all models' configurations as well as methods for loading/downloading/saving configurations.\n",
        "        Note:\n",
        "            A configuration file can be loaded and saved to disk. Loading the configuration file and using this file to initialize a model does **not** load the model weights.\n",
        "            It only affects the model's configuration.\n",
        "        Class attributes (overridden by derived classes):\n",
        "            - ``pretrained_config_archive_map``: a python ``dict`` of with `short-cut-names` (string) as keys and `url` (string) of associated pretrained model configurations as values.\n",
        "        Parameters:\n",
        "            ``finetuning_task``: string, default `None`. Name of the task used to fine-tune the model. This can be used when converting from an original (TensorFlow or PyTorch) checkpoint.\n",
        "            ``num_labels``: integer, default `2`. Number of classes to use when the model is a classification model (sequences/tokens)\n",
        "            ``output_attentions``: boolean, default `False`. Should the model returns attentions weights.\n",
        "            ``output_hidden_states``: string, default `False`. Should the model returns all hidden-states.\n",
        "            ``torchscript``: string, default `False`. Is the model used with Torchscript.\n",
        "    \"\"\"\n",
        "    pretrained_config_archive_map = {}\n",
        "\n",
        "    def __init__(self, **kwargs):\n",
        "        self.finetuning_task = kwargs.pop('finetuning_task', None)\n",
        "        self.num_labels = kwargs.pop('num_labels', 2)\n",
        "        self.output_attentions = kwargs.pop('output_attentions', False)\n",
        "        self.output_hidden_states = kwargs.pop('output_hidden_states', False)\n",
        "        self.output_past = kwargs.pop('output_past', True)  # Not used by all models\n",
        "        self.torchscript = kwargs.pop('torchscript', False)  # Only used by PyTorch models\n",
        "        self.use_bfloat16 = kwargs.pop('use_bfloat16', False)\n",
        "        self.pruned_heads = kwargs.pop('pruned_heads', {})\n",
        "\n",
        "    def save_pretrained(self, save_directory):\n",
        "        \"\"\" Save a configuration object to the directory `save_directory`, so that it\n",
        "            can be re-loaded using the :func:`~transformers.PretrainedConfig.from_pretrained` class method.\n",
        "        \"\"\"\n",
        "        assert os.path.isdir(save_directory), \"Saving path should be a directory where the model and configuration can be saved\"\n",
        "\n",
        "        # If we save using the predefined names, we can load using `from_pretrained`\n",
        "        output_config_file = os.path.join(save_directory, CONFIG_NAME)\n",
        "\n",
        "        self.to_json_file(output_config_file)\n",
        "        \n",
        "    @classmethod\n",
        "    def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):\n",
        "        r\"\"\" Instantiate a :class:`~transformers.PretrainedConfig` (or a derived class) from a pre-trained model configuration.\n",
        "        Parameters:\n",
        "            pretrained_model_name_or_path: either:\n",
        "                - a string with the `shortcut name` of a pre-trained model configuration to load from cache or download, e.g.: ``bert-base-uncased``.\n",
        "                - a path to a `directory` containing a configuration file saved using the :func:`~transformers.PretrainedConfig.save_pretrained` method, e.g.: ``./my_model_directory/``.\n",
        "                - a path or url to a saved configuration JSON `file`, e.g.: ``./my_model_directory/configuration.json``.\n",
        "            cache_dir: (`optional`) string:\n",
        "                Path to a directory in which a downloaded pre-trained model\n",
        "                configuration should be cached if the standard cache should not be used.\n",
        "            kwargs: (`optional`) dict: key/value pairs with which to update the configuration object after loading.\n",
        "                - The values in kwargs of any keys which are configuration attributes will be used to override the loaded values.\n",
        "                - Behavior concerning key/value pairs whose keys are *not* configuration attributes is controlled by the `return_unused_kwargs` keyword parameter.\n",
        "            force_download: (`optional`) boolean, default False:\n",
        "                Force to (re-)download the model weights and configuration files and override the cached versions if they exists.\n",
        "            proxies: (`optional`) dict, default None:\n",
        "                A dictionary of proxy servers to use by protocol or endpoint, e.g.: {'http': 'foo.bar:3128', 'http://hostname': 'foo.bar:4012'}.\n",
        "                The proxies are used on each request.\n",
        "            return_unused_kwargs: (`optional`) bool:\n",
        "                - If False, then this function returns just the final configuration object.\n",
        "                - If True, then this functions returns a tuple `(config, unused_kwargs)` where `unused_kwargs` is a dictionary consisting of the key/value pairs whose keys are not configuration attributes: ie the part of kwargs which has not been used to update `config` and is otherwise ignored.\n",
        "        Examples::\n",
        "            # We can't instantiate directly the base class `PretrainedConfig` so let's show the examples on a\n",
        "            # derived class: BertConfig\n",
        "            config = BertConfig.from_pretrained('bert-base-uncased')    # Download configuration from S3 and cache.\n",
        "            config = BertConfig.from_pretrained('./test/saved_model/')  # E.g. config (or model) was saved using `save_pretrained('./test/saved_model/')`\n",
        "            config = BertConfig.from_pretrained('./test/saved_model/my_configuration.json')\n",
        "            config = BertConfig.from_pretrained('bert-base-uncased', output_attention=True, foo=False)\n",
        "            assert config.output_attention == True\n",
        "            config, unused_kwargs = BertConfig.from_pretrained('bert-base-uncased', output_attention=True,\n",
        "                                                               foo=False, return_unused_kwargs=True)\n",
        "            assert config.output_attention == True\n",
        "            assert unused_kwargs == {'foo': False}\n",
        "        \"\"\"\n",
        "        cache_dir = kwargs.pop('cache_dir', None)\n",
        "        force_download = kwargs.pop('force_download', False)\n",
        "        proxies = kwargs.pop('proxies', None)\n",
        "        return_unused_kwargs = kwargs.pop('return_unused_kwargs', False)\n",
        "\n",
        "        if pretrained_model_name_or_path in cls.pretrained_config_archive_map:\n",
        "            config_file = cls.pretrained_config_archive_map[pretrained_model_name_or_path]\n",
        "        elif os.path.isdir(pretrained_model_name_or_path):\n",
        "            config_file = os.path.join(pretrained_model_name_or_path, CONFIG_NAME)\n",
        "        else:\n",
        "            config_file = pretrained_model_name_or_path\n",
        "        # redirect to the cache, if necessary\n",
        "        try:\n",
        "            resolved_config_file = cached_path(config_file, cache_dir=cache_dir, force_download=force_download, proxies=proxies)\n",
        "        except EnvironmentError:\n",
        "            if pretrained_model_name_or_path in cls.pretrained_config_archive_map:\n",
        "                msg = \"Couldn't reach server at '{}' to download pretrained model configuration file.\".format(\n",
        "                        config_file)\n",
        "            else:\n",
        "                msg = \"Model name '{}' was not found in model name list ({}). \" \\\n",
        "                      \"We assumed '{}' was a path or url to a configuration file named {} or \" \\\n",
        "                      \"a directory containing such a file but couldn't find any such file at this path or url.\".format(\n",
        "                        pretrained_model_name_or_path,\n",
        "                        ', '.join(cls.pretrained_config_archive_map.keys()),\n",
        "                        config_file, CONFIG_NAME)\n",
        "            raise EnvironmentError(msg)\n",
        "\n",
        "        \n",
        "        # Load config\n",
        "        config = cls.from_json_file(resolved_config_file)\n",
        "\n",
        "        if hasattr(config, 'pruned_heads'):\n",
        "            config.pruned_heads = dict((int(key), value) for key, value in config.pruned_heads.items())\n",
        "\n",
        "        # Update config with kwargs if needed\n",
        "        to_remove = []\n",
        "        for key, value in kwargs.items():\n",
        "            if hasattr(config, key):\n",
        "                setattr(config, key, value)\n",
        "                to_remove.append(key)\n",
        "        for key in to_remove:\n",
        "            kwargs.pop(key, None)\n",
        "\n",
        "        if return_unused_kwargs:\n",
        "            return config, kwargs\n",
        "        else:\n",
        "            return config\n",
        "\n",
        "    @classmethod\n",
        "    def from_dict(cls, json_object):\n",
        "        \"\"\"Constructs a `Config` from a Python dictionary of parameters.\"\"\"\n",
        "        config = cls(vocab_size_or_config_json_file=-1)\n",
        "        for key, value in json_object.items():\n",
        "            setattr(config, key, value)\n",
        "        return config\n",
        "\n",
        "    @classmethod\n",
        "    def from_json_file(cls, json_file):\n",
        "        \"\"\"Constructs a `BertConfig` from a json file of parameters.\"\"\"\n",
        "        with open(json_file, \"r\", encoding='utf-8') as reader:\n",
        "            text = reader.read()\n",
        "        return cls.from_dict(json.loads(text))\n",
        "\n",
        "    def __eq__(self, other):\n",
        "        return self.__dict__ == other.__dict__\n",
        "\n",
        "    def __repr__(self):\n",
        "        return str(self.to_json_string())\n",
        "\n",
        "    def to_dict(self):\n",
        "        \"\"\"Serializes this instance to a Python dictionary.\"\"\"\n",
        "        output = copy.deepcopy(self.__dict__)\n",
        "        return output\n",
        "\n",
        "    def to_json_string(self):\n",
        "        \"\"\"Serializes this instance to a JSON string.\"\"\"\n",
        "        return json.dumps(self.to_dict(), indent=2, sort_keys=True) + \"\\n\"\n",
        "\n",
        "    def to_json_file(self, json_file_path):\n",
        "        \"\"\" Save this instance to a json file.\"\"\"\n",
        "        with open(json_file_path, \"w\", encoding='utf-8') as writer:\n",
        "            writer.write(self.to_json_string())\n",
        "\n",
        "class BertConfig(PretrainedConfig):\n",
        "    r\"\"\"\n",
        "        :class:`~transformers.BertConfig` is the configuration class to store the configuration of a\n",
        "        `BertModel`.\n",
        "        Arguments:\n",
        "            vocab_size_or_config_json_file: Vocabulary size of `inputs_ids` in `BertModel`.\n",
        "            hidden_size: Size of the encoder layers and the pooler layer.\n",
        "            num_hidden_layers: Number of hidden layers in the Transformer encoder.\n",
        "            num_attention_heads: Number of attention heads for each attention layer in\n",
        "                the Transformer encoder.\n",
        "            intermediate_size: The size of the \"intermediate\" (i.e., feed-forward)\n",
        "                layer in the Transformer encoder.\n",
        "            hidden_act: The non-linear activation function (function or string) in the\n",
        "                encoder and pooler. If string, \"gelu\", \"relu\", \"swish\" and \"gelu_new\" are supported.\n",
        "            hidden_dropout_prob: The dropout probabilitiy for all fully connected\n",
        "                layers in the embeddings, encoder, and pooler.\n",
        "            attention_probs_dropout_prob: The dropout ratio for the attention\n",
        "                probabilities.\n",
        "            max_position_embeddings: The maximum sequence length that this model might\n",
        "                ever be used with. Typically set this to something large just in case\n",
        "                (e.g., 512 or 1024 or 2048).\n",
        "            type_vocab_size: The vocabulary size of the `token_type_ids` passed into\n",
        "                `BertModel`.\n",
        "            initializer_range: The sttdev of the truncated_normal_initializer for\n",
        "                initializing all weight matrices.\n",
        "            layer_norm_eps: The epsilon used by LayerNorm.\n",
        "    \"\"\"\n",
        "    pretrained_config_archive_map = BERT_PRETRAINED_CONFIG_ARCHIVE_MAP\n",
        "\n",
        "    def __init__(self,\n",
        "                 vocab_size_or_config_json_file=30522,\n",
        "                 hidden_size=768,\n",
        "                 num_hidden_layers=12,\n",
        "                 num_attention_heads=12,\n",
        "                 intermediate_size=3072,\n",
        "                 hidden_act=\"gelu\",\n",
        "                 hidden_dropout_prob=0.1,\n",
        "                 attention_probs_dropout_prob=0.1,\n",
        "                 max_position_embeddings=512,\n",
        "                 type_vocab_size=2,\n",
        "                 initializer_range=0.02,\n",
        "                 layer_norm_eps=1e-12,\n",
        "                 **kwargs):\n",
        "        super(BertConfig, self).__init__(**kwargs)\n",
        "        if isinstance(vocab_size_or_config_json_file, str) or (sys.version_info[0] == 2\n",
        "                        and isinstance(vocab_size_or_config_json_file, unicode)):\n",
        "            with open(vocab_size_or_config_json_file, \"r\", encoding='utf-8') as reader:\n",
        "                json_config = json.loads(reader.read())\n",
        "            for key, value in json_config.items():\n",
        "                self.__dict__[key] = value\n",
        "        elif isinstance(vocab_size_or_config_json_file, int):\n",
        "            self.vocab_size = vocab_size_or_config_json_file\n",
        "            self.hidden_size = hidden_size\n",
        "            self.num_hidden_layers = num_hidden_layers\n",
        "            self.num_attention_heads = num_attention_heads\n",
        "            self.hidden_act = hidden_act\n",
        "            self.intermediate_size = intermediate_size\n",
        "            self.hidden_dropout_prob = hidden_dropout_prob\n",
        "            self.attention_probs_dropout_prob = attention_probs_dropout_prob\n",
        "            self.max_position_embeddings = max_position_embeddings\n",
        "            self.type_vocab_size = type_vocab_size\n",
        "            self.initializer_range = initializer_range\n",
        "            self.layer_norm_eps = layer_norm_eps\n",
        "        else:\n",
        "            raise ValueError(\"First argument must be either a vocabulary size (int)\"\n",
        "                             \" or the path to a pretrained model config file (str)\")\n",
        "\n",
        "class RobertaConfig(BertConfig):\n",
        "    pretrained_config_archive_map = ROBERTA_PRETRAINED_CONFIG_ARCHIVE_MAP\n",
        "\n",
        "class XLNetConfig(PretrainedConfig):\n",
        "    \"\"\"Configuration class to store the configuration of a ``XLNetModel``.\n",
        "    Args:\n",
        "        vocab_size_or_config_json_file: Vocabulary size of ``inputs_ids`` in ``XLNetModel``.\n",
        "        d_model: Size of the encoder layers and the pooler layer.\n",
        "        n_layer: Number of hidden layers in the Transformer encoder.\n",
        "        n_head: Number of attention heads for each attention layer in\n",
        "            the Transformer encoder.\n",
        "        d_inner: The size of the \"intermediate\" (i.e., feed-forward)\n",
        "            layer in the Transformer encoder.\n",
        "        ff_activation: The non-linear activation function (function or string) in the\n",
        "            encoder and pooler. If string, \"gelu\", \"relu\" and \"swish\" are supported.\n",
        "        untie_r: untie relative position biases\n",
        "        attn_type: 'bi' for XLNet, 'uni' for Transformer-XL\n",
        "        dropout: The dropout probabilitiy for all fully connected\n",
        "            layers in the embeddings, encoder, and pooler.\n",
        "        initializer_range: The sttdev of the truncated_normal_initializer for\n",
        "            initializing all weight matrices.\n",
        "        layer_norm_eps: The epsilon used by LayerNorm.\n",
        "        dropout: float, dropout rate.\n",
        "        init: str, the initialization scheme, either \"normal\" or \"uniform\".\n",
        "        init_range: float, initialize the parameters with a uniform distribution\n",
        "            in [-init_range, init_range]. Only effective when init=\"uniform\".\n",
        "        init_std: float, initialize the parameters with a normal distribution\n",
        "            with mean 0 and stddev init_std. Only effective when init=\"normal\".\n",
        "        mem_len: int, the number of tokens to cache.\n",
        "        reuse_len: int, the number of tokens in the currect batch to be cached\n",
        "            and reused in the future.\n",
        "        bi_data: bool, whether to use bidirectional input pipeline.\n",
        "            Usually set to True during pretraining and False during finetuning.\n",
        "        clamp_len: int, clamp all relative distances larger than clamp_len.\n",
        "            -1 means no clamping.\n",
        "        same_length: bool, whether to use the same attention length for each token.\n",
        "        finetuning_task: name of the glue task on which the model was fine-tuned if any\n",
        "    \"\"\"\n",
        "    pretrained_config_archive_map = XLNET_PRETRAINED_CONFIG_ARCHIVE_MAP\n",
        "\n",
        "    def __init__(self,\n",
        "                 vocab_size_or_config_json_file=32000,\n",
        "                 d_model=1024,\n",
        "                 n_layer=24,\n",
        "                 n_head=16,\n",
        "                 d_inner=4096,\n",
        "                 max_position_embeddings=512,\n",
        "                 ff_activation=\"gelu\",\n",
        "                 untie_r=True,\n",
        "                 attn_type=\"bi\",\n",
        "\n",
        "                 initializer_range=0.02,\n",
        "                 layer_norm_eps=1e-12,\n",
        "\n",
        "                 dropout=0.1,\n",
        "                 mem_len=None,\n",
        "                 reuse_len=None,\n",
        "                 bi_data=False,\n",
        "                 clamp_len=-1,\n",
        "                 same_length=False,\n",
        "\n",
        "                 finetuning_task=None,\n",
        "                 num_labels=2,\n",
        "                 summary_type='last',\n",
        "                 summary_use_proj=True,\n",
        "                 summary_activation='tanh',\n",
        "                 summary_last_dropout=0.1,\n",
        "                 start_n_top=5,\n",
        "                 end_n_top=5,\n",
        "                 **kwargs):\n",
        "        \"\"\"Constructs XLNetConfig.\n",
        "        \"\"\"\n",
        "        super(XLNetConfig, self).__init__(**kwargs)\n",
        "\n",
        "        if isinstance(vocab_size_or_config_json_file, str) or (sys.version_info[0] == 2\n",
        "                        and isinstance(vocab_size_or_config_json_file, unicode)):\n",
        "            with open(vocab_size_or_config_json_file, \"r\", encoding='utf-8') as reader:\n",
        "                json_config = json.loads(reader.read())\n",
        "            for key, value in json_config.items():\n",
        "                setattr(config, key, value)\n",
        "        elif isinstance(vocab_size_or_config_json_file, int):\n",
        "            self.n_token = vocab_size_or_config_json_file\n",
        "            self.d_model = d_model\n",
        "            self.n_layer = n_layer\n",
        "            self.n_head = n_head\n",
        "            assert d_model % n_head == 0\n",
        "            self.d_head = d_model // n_head\n",
        "            self.ff_activation = ff_activation\n",
        "            self.d_inner = d_inner\n",
        "            self.untie_r = untie_r\n",
        "            self.attn_type = attn_type\n",
        "\n",
        "            self.initializer_range = initializer_range\n",
        "            self.layer_norm_eps = layer_norm_eps\n",
        "\n",
        "            self.dropout = dropout\n",
        "            self.mem_len = mem_len\n",
        "            self.reuse_len = reuse_len\n",
        "            self.bi_data = bi_data\n",
        "            self.clamp_len = clamp_len\n",
        "            self.same_length = same_length\n",
        "\n",
        "            self.finetuning_task = finetuning_task\n",
        "            self.num_labels = num_labels\n",
        "            self.summary_type = summary_type\n",
        "            self.summary_use_proj = summary_use_proj\n",
        "            self.summary_activation = summary_activation\n",
        "            self.summary_last_dropout = summary_last_dropout\n",
        "            self.start_n_top = start_n_top\n",
        "            self.end_n_top = end_n_top\n",
        "        else:\n",
        "            raise ValueError(\"First argument must be either a vocabulary size (int)\"\n",
        "                             \" or the path to a pretrained model config file (str)\")\n",
        "\n",
        "    @property\n",
        "    def max_position_embeddings(self):\n",
        "        return -1\n",
        "\n",
        "    @property\n",
        "    def vocab_size(self):\n",
        "        return self.n_token\n",
        "\n",
        "    @vocab_size.setter\n",
        "    def vocab_size(self, value):\n",
        "        self.n_token = value\n",
        "\n",
        "    @property\n",
        "    def hidden_size(self):\n",
        "        return self.d_model\n",
        "\n",
        "    @property\n",
        "    def num_attention_heads(self):\n",
        "        return self.n_head\n",
        "\n",
        "    @property\n",
        "    def num_hidden_layers(self):\n",
        "        return self.n_layer\n",
        "\n",
        "class PreTrainedModel(nn.Module):\n",
        "    r\"\"\" Base class for all models.\n",
        "        :class:`~transformers.PreTrainedModel` takes care of storing the configuration of the models and handles methods for loading/downloading/saving models\n",
        "        as well as a few methods commons to all models to (i) resize the input embeddings and (ii) prune heads in the self-attention heads.\n",
        "        Class attributes (overridden by derived classes):\n",
        "            - ``config_class``: a class derived from :class:`~transformers.PretrainedConfig` to use as configuration class for this model architecture.\n",
        "            - ``pretrained_model_archive_map``: a python ``dict`` of with `short-cut-names` (string) as keys and `url` (string) of associated pretrained weights as values.\n",
        "            - ``load_tf_weights``: a python ``method`` for loading a TensorFlow checkpoint in a PyTorch model, taking as arguments:\n",
        "                - ``model``: an instance of the relevant subclass of :class:`~transformers.PreTrainedModel`,\n",
        "                - ``config``: an instance of the relevant subclass of :class:`~transformers.PretrainedConfig`,\n",
        "                - ``path``: a path (string) to the TensorFlow checkpoint.\n",
        "            - ``base_model_prefix``: a string indicating the attribute associated to the base model in derived classes of the same architecture adding modules on top of the base model.\n",
        "    \"\"\"\n",
        "    config_class = None\n",
        "    pretrained_model_archive_map = {}\n",
        "    load_tf_weights = lambda model, config, path: None\n",
        "    base_model_prefix = \"\"\n",
        "\n",
        "    def __init__(self, config, *inputs, **kwargs):\n",
        "        super(PreTrainedModel, self).__init__()\n",
        "        if not isinstance(config, PretrainedConfig):\n",
        "            raise ValueError(\n",
        "                \"Parameter config in `{}(config)` should be an instance of class `PretrainedConfig`. \"\n",
        "                \"To create a model from a pretrained model use \"\n",
        "                \"`model = {}.from_pretrained(PRETRAINED_MODEL_NAME)`\".format(\n",
        "                    self.__class__.__name__, self.__class__.__name__\n",
        "                ))\n",
        "        # Save config in model\n",
        "        self.config = config\n",
        "\n",
        "    def _get_resized_embeddings(self, old_embeddings, new_num_tokens=None):\n",
        "        \"\"\" Build a resized Embedding Module from a provided token Embedding Module.\n",
        "            Increasing the size will add newly initialized vectors at the end\n",
        "            Reducing the size will remove vectors from the end\n",
        "        Args:\n",
        "            new_num_tokens: (`optional`) int\n",
        "                New number of tokens in the embedding matrix.\n",
        "                Increasing the size will add newly initialized vectors at the end\n",
        "                Reducing the size will remove vectors from the end\n",
        "                If not provided or None: return the provided token Embedding Module.\n",
        "        Return: ``torch.nn.Embeddings``\n",
        "            Pointer to the resized Embedding Module or the old Embedding Module if new_num_tokens is None\n",
        "        \"\"\"\n",
        "        if new_num_tokens is None:\n",
        "            return old_embeddings\n",
        "\n",
        "        old_num_tokens, old_embedding_dim = old_embeddings.weight.size()\n",
        "        if old_num_tokens == new_num_tokens:\n",
        "            return old_embeddings\n",
        "\n",
        "        # Build new embeddings\n",
        "        new_embeddings = nn.Embedding(new_num_tokens, old_embedding_dim)\n",
        "        new_embeddings.to(old_embeddings.weight.device)\n",
        "\n",
        "        # initialize all new embeddings (in particular added tokens)\n",
        "        self._init_weights(new_embeddings)\n",
        "\n",
        "        # Copy word embeddings from the previous weights\n",
        "        num_tokens_to_copy = min(old_num_tokens, new_num_tokens)\n",
        "        new_embeddings.weight.data[:num_tokens_to_copy, :] = old_embeddings.weight.data[:num_tokens_to_copy, :]\n",
        "\n",
        "        return new_embeddings\n",
        "\n",
        "    def _tie_or_clone_weights(self, first_module, second_module):\n",
        "        \"\"\" Tie or clone module weights depending of weither we are using TorchScript or not\n",
        "        \"\"\"\n",
        "        if self.config.torchscript:\n",
        "            first_module.weight = nn.Parameter(second_module.weight.clone())\n",
        "        else:\n",
        "            first_module.weight = second_module.weight\n",
        "\n",
        "        if hasattr(first_module, 'bias') and first_module.bias is not None:\n",
        "            first_module.bias.data = torch.nn.functional.pad(\n",
        "                first_module.bias.data,\n",
        "                (0, first_module.weight.shape[0] - first_module.bias.shape[0]),\n",
        "                'constant',\n",
        "                0\n",
        "            )\n",
        "\n",
        "    def resize_token_embeddings(self, new_num_tokens=None):\n",
        "        \"\"\" Resize input token embeddings matrix of the model if new_num_tokens != config.vocab_size.\n",
        "        Take care of tying weights embeddings afterwards if the model class has a `tie_weights()` method.\n",
        "        Arguments:\n",
        "            new_num_tokens: (`optional`) int:\n",
        "                New number of tokens in the embedding matrix. Increasing the size will add newly initialized vectors at the end. Reducing the size will remove vectors from the end.\n",
        "                If not provided or None: does nothing and just returns a pointer to the input tokens ``torch.nn.Embeddings`` Module of the model.\n",
        "        Return: ``torch.nn.Embeddings``\n",
        "            Pointer to the input tokens Embeddings Module of the model\n",
        "        \"\"\"\n",
        "        base_model = getattr(self, self.base_model_prefix, self)  # get the base model if needed\n",
        "        model_embeds = base_model._resize_token_embeddings(new_num_tokens)\n",
        "        if new_num_tokens is None:\n",
        "            return model_embeds\n",
        "\n",
        "        # Update base model and current model config\n",
        "        self.config.vocab_size = new_num_tokens\n",
        "        base_model.vocab_size = new_num_tokens\n",
        "\n",
        "        # Tie weights again if needed\n",
        "        if hasattr(self, 'tie_weights'):\n",
        "            self.tie_weights()\n",
        "\n",
        "        return model_embeds\n",
        "\n",
        "    def init_weights(self):\n",
        "        \"\"\" Initialize and prunes weights if needed. \"\"\"\n",
        "        # Initialize weights\n",
        "        self.apply(self._init_weights)\n",
        "\n",
        "        # Prune heads if needed\n",
        "        if self.config.pruned_heads:\n",
        "            self.prune_heads(self.config.pruned_heads)\n",
        "\n",
        "    def prune_heads(self, heads_to_prune):\n",
        "        \"\"\" Prunes heads of the base model.\n",
        "            Arguments:\n",
        "                heads_to_prune: dict with keys being selected layer indices (`int`) and associated values being the list of heads to prune in said layer (list of `int`).\n",
        "                E.g. {1: [0, 2], 2: [2, 3]} will prune heads 0 and 2 on layer 1 and heads 2 and 3 on layer 2.\n",
        "        \"\"\"\n",
        "        base_model = getattr(self, self.base_model_prefix, self)  # get the base model if needed\n",
        "\n",
        "        # save new sets of pruned heads as union of previously stored pruned heads and newly pruned heads\n",
        "        for layer, heads in heads_to_prune.items():\n",
        "            union_heads = set(self.config.pruned_heads.get(layer, [])) | set(heads)\n",
        "            self.config.pruned_heads[layer] = list(union_heads)  # Unfortunately we have to store it as list for JSON\n",
        "\n",
        "        base_model._prune_heads(heads_to_prune)\n",
        "\n",
        "    def save_pretrained(self, save_directory):\n",
        "        \"\"\" Save a model and its configuration file to a directory, so that it\n",
        "            can be re-loaded using the `:func:`~transformers.PreTrainedModel.from_pretrained`` class method.\n",
        "        \"\"\"\n",
        "        assert os.path.isdir(save_directory), \"Saving path should be a directory where the model and configuration can be saved\"\n",
        "\n",
        "        # Only save the model it-self if we are using distributed training\n",
        "        model_to_save = self.module if hasattr(self, 'module') else self\n",
        "\n",
        "        # Save configuration file\n",
        "        model_to_save.config.save_pretrained(save_directory)\n",
        "\n",
        "        # If we save using the predefined names, we can load using `from_pretrained`\n",
        "        output_model_file = os.path.join(save_directory, WEIGHTS_NAME)\n",
        "        torch.save(model_to_save.state_dict(), output_model_file)\n",
        "        \n",
        "    @classmethod\n",
        "    def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):\n",
        "        r\"\"\"Instantiate a pretrained pytorch model from a pre-trained model configuration.\n",
        "        The model is set in evaluation mode by default using ``model.eval()`` (Dropout modules are deactivated)\n",
        "        To train the model, you should first set it back in training mode with ``model.train()``\n",
        "        The warning ``Weights from XXX not initialized from pretrained model`` means that the weights of XXX do not come pre-trained with the rest of the model.\n",
        "        It is up to you to train those weights with a downstream fine-tuning task.\n",
        "        The warning ``Weights from XXX not used in YYY`` means that the layer XXX is not used by YYY, therefore those weights are discarded.\n",
        "        Parameters:\n",
        "            pretrained_model_name_or_path: either:\n",
        "                - a string with the `shortcut name` of a pre-trained model to load from cache or download, e.g.: ``bert-base-uncased``.\n",
        "                - a path to a `directory` containing model weights saved using :func:`~transformers.PreTrainedModel.save_pretrained`, e.g.: ``./my_model_directory/``.\n",
        "                - a path or url to a `tensorflow index checkpoint file` (e.g. `./tf_model/model.ckpt.index`). In this case, ``from_tf`` should be set to True and a configuration object should be provided as ``config`` argument. This loading path is slower than converting the TensorFlow checkpoint in a PyTorch model using the provided conversion scripts and loading the PyTorch model afterwards.\n",
        "                - None if you are both providing the configuration and state dictionary (resp. with keyword arguments ``config`` and ``state_dict``)\n",
        "            model_args: (`optional`) Sequence of positional arguments:\n",
        "                All remaning positional arguments will be passed to the underlying model's ``__init__`` method\n",
        "            config: (`optional`) instance of a class derived from :class:`~transformers.PretrainedConfig`:\n",
        "                Configuration for the model to use instead of an automatically loaded configuation. Configuration can be automatically loaded when:\n",
        "                - the model is a model provided by the library (loaded with the ``shortcut-name`` string of a pretrained model), or\n",
        "                - the model was saved using :func:`~transformers.PreTrainedModel.save_pretrained` and is reloaded by suppling the save directory.\n",
        "                - the model is loaded by suppling a local directory as ``pretrained_model_name_or_path`` and a configuration JSON file named `config.json` is found in the directory.\n",
        "            state_dict: (`optional`) dict:\n",
        "                an optional state dictionnary for the model to use instead of a state dictionary loaded from saved weights file.\n",
        "                This option can be used if you want to create a model from a pretrained configuration but load your own weights.\n",
        "                In this case though, you should check if using :func:`~transformers.PreTrainedModel.save_pretrained` and :func:`~transformers.PreTrainedModel.from_pretrained` is not a simpler option.\n",
        "            cache_dir: (`optional`) string:\n",
        "                Path to a directory in which a downloaded pre-trained model\n",
        "                configuration should be cached if the standard cache should not be used.\n",
        "            force_download: (`optional`) boolean, default False:\n",
        "                Force to (re-)download the model weights and configuration files and override the cached versions if they exists.\n",
        "            proxies: (`optional`) dict, default None:\n",
        "                A dictionary of proxy servers to use by protocol or endpoint, e.g.: {'http': 'foo.bar:3128', 'http://hostname': 'foo.bar:4012'}.\n",
        "                The proxies are used on each request.\n",
        "            output_loading_info: (`optional`) boolean:\n",
        "                Set to ``True`` to also return a dictionnary containing missing keys, unexpected keys and error messages.\n",
        "            kwargs: (`optional`) Remaining dictionary of keyword arguments:\n",
        "                Can be used to update the configuration object (after it being loaded) and initiate the model. (e.g. ``output_attention=True``). Behave differently depending on whether a `config` is provided or automatically loaded:\n",
        "                - If a configuration is provided with ``config``, ``**kwargs`` will be directly passed to the underlying model's ``__init__`` method (we assume all relevant updates to the configuration have already been done)\n",
        "                - If a configuration is not provided, ``kwargs`` will be first passed to the configuration class initialization function (:func:`~transformers.PretrainedConfig.from_pretrained`). Each key of ``kwargs`` that corresponds to a configuration attribute will be used to override said attribute with the supplied ``kwargs`` value. Remaining keys that do not correspond to any configuration attribute will be passed to the underlying model's ``__init__`` function.\n",
        "        Examples::\n",
        "            model = BertModel.from_pretrained('bert-base-uncased')    # Download model and configuration from S3 and cache.\n",
        "            model = BertModel.from_pretrained('./test/saved_model/')  # E.g. model was saved using `save_pretrained('./test/saved_model/')`\n",
        "            model = BertModel.from_pretrained('bert-base-uncased', output_attention=True)  # Update configuration during loading\n",
        "            assert model.config.output_attention == True\n",
        "            # Loading from a TF checkpoint file instead of a PyTorch model (slower)\n",
        "            config = BertConfig.from_json_file('./tf_model/my_tf_model_config.json')\n",
        "            model = BertModel.from_pretrained('./tf_model/my_tf_checkpoint.ckpt.index', from_tf=True, config=config)\n",
        "        \"\"\"\n",
        "        config = kwargs.pop('config', None)\n",
        "        state_dict = kwargs.pop('state_dict', None)\n",
        "        cache_dir = kwargs.pop('cache_dir', None)\n",
        "        from_tf = kwargs.pop('from_tf', False)\n",
        "        force_download = kwargs.pop('force_download', False)\n",
        "        proxies = kwargs.pop('proxies', None)\n",
        "        output_loading_info = kwargs.pop('output_loading_info', False)\n",
        "\n",
        "        # Load config\n",
        "        if config is None:\n",
        "            config, model_kwargs = cls.config_class.from_pretrained(\n",
        "                pretrained_model_name_or_path, *model_args,\n",
        "                cache_dir=cache_dir, return_unused_kwargs=True,\n",
        "                force_download=force_download,\n",
        "                **kwargs\n",
        "            )\n",
        "        else:\n",
        "            model_kwargs = kwargs\n",
        "\n",
        "        # Load model\n",
        "        if pretrained_model_name_or_path is not None:\n",
        "            if pretrained_model_name_or_path in cls.pretrained_model_archive_map:\n",
        "                archive_file = cls.pretrained_model_archive_map[pretrained_model_name_or_path]\n",
        "            elif os.path.isdir(pretrained_model_name_or_path):\n",
        "                if from_tf and os.path.isfile(os.path.join(pretrained_model_name_or_path, TF_WEIGHTS_NAME + \".index\")):\n",
        "                    # Load from a TF 1.0 checkpoint\n",
        "                    archive_file = os.path.join(pretrained_model_name_or_path, TF_WEIGHTS_NAME + \".index\")\n",
        "                elif from_tf and os.path.isfile(os.path.join(pretrained_model_name_or_path, TF2_WEIGHTS_NAME)):\n",
        "                    # Load from a TF 2.0 checkpoint\n",
        "                    archive_file = os.path.join(pretrained_model_name_or_path, TF2_WEIGHTS_NAME)\n",
        "                elif os.path.isfile(os.path.join(pretrained_model_name_or_path, WEIGHTS_NAME)):\n",
        "                    # Load from a PyTorch checkpoint\n",
        "                    archive_file = os.path.join(pretrained_model_name_or_path, WEIGHTS_NAME)\n",
        "                else:\n",
        "                    raise EnvironmentError(\"Error no file named {} found in directory {} or `from_tf` set to False\".format(\n",
        "                        [WEIGHTS_NAME, TF2_WEIGHTS_NAME, TF_WEIGHTS_NAME + \".index\"],\n",
        "                        pretrained_model_name_or_path))\n",
        "            elif os.path.isfile(pretrained_model_name_or_path):\n",
        "                archive_file = pretrained_model_name_or_path\n",
        "            else:\n",
        "                assert from_tf, \"Error finding file {}, no file or TF 1.X checkpoint found\".format(pretrained_model_name_or_path)\n",
        "                archive_file = pretrained_model_name_or_path + \".index\"\n",
        "\n",
        "            # redirect to the cache, if necessary\n",
        "            try:\n",
        "                resolved_archive_file = cached_path(archive_file, cache_dir=cache_dir, force_download=force_download, proxies=proxies)\n",
        "            except EnvironmentError:\n",
        "                if pretrained_model_name_or_path in cls.pretrained_model_archive_map:\n",
        "                    msg = \"Couldn't reach server at '{}' to download pretrained weights.\".format(\n",
        "                            archive_file)\n",
        "                else:\n",
        "                    msg = \"Model name '{}' was not found in model name list ({}). \" \\\n",
        "                        \"We assumed '{}' was a path or url to model weight files named one of {} but \" \\\n",
        "                        \"couldn't find any such file at this path or url.\".format(\n",
        "                            pretrained_model_name_or_path,\n",
        "                            ', '.join(cls.pretrained_model_archive_map.keys()),\n",
        "                            archive_file,\n",
        "                            [WEIGHTS_NAME, TF2_WEIGHTS_NAME, TF_WEIGHTS_NAME])\n",
        "                raise EnvironmentError(msg)\n",
        "\n",
        "            \n",
        "        else:\n",
        "            resolved_archive_file = None\n",
        "\n",
        "        # Instantiate model.\n",
        "        model = cls(config, *model_args, **model_kwargs)\n",
        "\n",
        "        if state_dict is None and not from_tf:\n",
        "            state_dict = torch.load(resolved_archive_file, map_location='cpu')\n",
        "\n",
        "        missing_keys = []\n",
        "        unexpected_keys = []\n",
        "        error_msgs = []\n",
        "\n",
        "        if from_tf:\n",
        "            if resolved_archive_file.endswith('.index'):\n",
        "                # Load from a TensorFlow 1.X checkpoint - provided by original authors\n",
        "                model = cls.load_tf_weights(model, config, resolved_archive_file[:-6])  # Remove the '.index'\n",
        "            else:\n",
        "                # Load from our TensorFlow 2.0 checkpoints\n",
        "                try:\n",
        "                    from transformers import load_tf2_checkpoint_in_pytorch_model\n",
        "                    model = load_tf2_checkpoint_in_pytorch_model(model, resolved_archive_file, allow_missing_keys=True)\n",
        "                except ImportError as e:\n",
        "                    raise e\n",
        "        else:\n",
        "            # Convert old format to new format if needed from a PyTorch state_dict\n",
        "            old_keys = []\n",
        "            new_keys = []\n",
        "            for key in state_dict.keys():\n",
        "                new_key = None\n",
        "                if 'gamma' in key:\n",
        "                    new_key = key.replace('gamma', 'weight')\n",
        "                if 'beta' in key:\n",
        "                    new_key = key.replace('beta', 'bias')\n",
        "                if new_key:\n",
        "                    old_keys.append(key)\n",
        "                    new_keys.append(new_key)\n",
        "            for old_key, new_key in zip(old_keys, new_keys):\n",
        "                state_dict[new_key] = state_dict.pop(old_key)\n",
        "\n",
        "            # copy state_dict so _load_from_state_dict can modify it\n",
        "            metadata = getattr(state_dict, '_metadata', None)\n",
        "            state_dict = state_dict.copy()\n",
        "            if metadata is not None:\n",
        "                state_dict._metadata = metadata\n",
        "\n",
        "            def load(module, prefix=''):\n",
        "                local_metadata = {} if metadata is None else metadata.get(prefix[:-1], {})\n",
        "                module._load_from_state_dict(\n",
        "                    state_dict, prefix, local_metadata, True, missing_keys, unexpected_keys, error_msgs)\n",
        "                for name, child in module._modules.items():\n",
        "                    if child is not None:\n",
        "                        load(child, prefix + name + '.')\n",
        "\n",
        "            # Make sure we are able to load base models as well as derived models (with heads)\n",
        "            start_prefix = ''\n",
        "            model_to_load = model\n",
        "            if not hasattr(model, cls.base_model_prefix) and any(s.startswith(cls.base_model_prefix) for s in state_dict.keys()):\n",
        "                start_prefix = cls.base_model_prefix + '.'\n",
        "            if hasattr(model, cls.base_model_prefix) and not any(s.startswith(cls.base_model_prefix) for s in state_dict.keys()):\n",
        "                model_to_load = getattr(model, cls.base_model_prefix)\n",
        "\n",
        "            load(model_to_load, prefix=start_prefix)\n",
        "            if len(error_msgs) > 0:\n",
        "                raise RuntimeError('Error(s) in loading state_dict for {}:\\n\\t{}'.format(\n",
        "                                model.__class__.__name__, \"\\n\\t\".join(error_msgs)))\n",
        "\n",
        "        if hasattr(model, 'tie_weights'):\n",
        "            model.tie_weights()  # make sure word embedding weights are still tied\n",
        "\n",
        "        # Set model in evaluation mode to desactivate DropOut modules by default\n",
        "        model.eval()\n",
        "\n",
        "        if output_loading_info:\n",
        "            loading_info = {\"missing_keys\": missing_keys, \"unexpected_keys\": unexpected_keys, \"error_msgs\": error_msgs}\n",
        "            return model, loading_info\n",
        "\n",
        "        return model\n",
        "\n",
        "XLNetLayerNorm = nn.LayerNorm\n",
        "class XLNetRelativeAttention(nn.Module):\n",
        "    def __init__(self, config):\n",
        "        super(XLNetRelativeAttention, self).__init__()\n",
        "        self.output_attentions = config.output_attentions\n",
        "\n",
        "        if config.d_model % config.n_head != 0:\n",
        "            raise ValueError(\n",
        "                \"The hidden size (%d) is not a multiple of the number of attention \"\n",
        "                \"heads (%d)\" % (config.d_model, config.n_head))\n",
        "\n",
        "        self.n_head = config.n_head\n",
        "        self.d_head = config.d_head\n",
        "        self.d_model = config.d_model\n",
        "        self.scale = 1 / (config.d_head ** 0.5)\n",
        "\n",
        "        self.q = nn.Parameter(torch.FloatTensor(config.d_model, self.n_head, self.d_head))\n",
        "        self.k = nn.Parameter(torch.FloatTensor(config.d_model, self.n_head, self.d_head))\n",
        "        self.v = nn.Parameter(torch.FloatTensor(config.d_model, self.n_head, self.d_head))\n",
        "        self.o = nn.Parameter(torch.FloatTensor(config.d_model, self.n_head, self.d_head))\n",
        "        self.r = nn.Parameter(torch.FloatTensor(config.d_model, self.n_head, self.d_head))\n",
        "\n",
        "        self.r_r_bias = nn.Parameter(torch.FloatTensor(self.n_head, self.d_head))\n",
        "        self.r_s_bias = nn.Parameter(torch.FloatTensor(self.n_head, self.d_head))\n",
        "        self.r_w_bias = nn.Parameter(torch.FloatTensor(self.n_head, self.d_head))\n",
        "        self.seg_embed = nn.Parameter(torch.FloatTensor(2, self.n_head, self.d_head))\n",
        "\n",
        "        self.layer_norm = XLNetLayerNorm(config.d_model, eps=config.layer_norm_eps)\n",
        "        self.dropout = nn.Dropout(config.dropout)\n",
        "\n",
        "    def prune_heads(self, heads):\n",
        "        raise NotImplementedError\n",
        "\n",
        "    @staticmethod\n",
        "    def rel_shift(x, klen=-1):\n",
        "        \"\"\"perform relative shift to form the relative attention score.\"\"\"\n",
        "        x_size = x.shape\n",
        "\n",
        "        x = x.reshape(x_size[1], x_size[0], x_size[2], x_size[3])\n",
        "        x = x[1:, ...]\n",
        "        x = x.reshape(x_size[0], x_size[1] - 1, x_size[2], x_size[3])\n",
        "        # x = x[:, 0:klen, :, :]\n",
        "        x = torch.index_select(x, 1, torch.arange(klen, device=x.device, dtype=torch.long))\n",
        "\n",
        "        return x\n",
        "\n",
        "    @staticmethod\n",
        "    def rel_shift_bnij(x, klen=-1):\n",
        "        x_size = x.shape\n",
        "\n",
        "        x = x.reshape(x_size[0], x_size[1], x_size[3], x_size[2])\n",
        "        x = x[:, :, 1:, :]\n",
        "        x = x.reshape(x_size[0], x_size[1], x_size[2], x_size[3]-1)\n",
        "        # Note: the tensor-slice form was faster in my testing than torch.index_select\n",
        "        #       However, tracing doesn't like the nature of the slice, and if klen changes\n",
        "        #       during the run then it'll fail, whereas index_select will be fine.\n",
        "        x = torch.index_select(x, 3, torch.arange(klen, device=x.device, dtype=torch.long))\n",
        "        # x = x[:, :, :, :klen]\n",
        "\n",
        "        return x\n",
        "\n",
        "    def rel_attn_core(self, q_head, k_head_h, v_head_h, k_head_r, seg_mat=None, attn_mask=None, head_mask=None):\n",
        "        \"\"\"Core relative positional attention operations.\"\"\"\n",
        "\n",
        "        # content based attention score\n",
        "        ac = torch.einsum('ibnd,jbnd->bnij', q_head + self.r_w_bias, k_head_h)\n",
        "\n",
        "        # position based attention score\n",
        "        bd = torch.einsum('ibnd,jbnd->bnij', q_head + self.r_r_bias, k_head_r)\n",
        "        bd = self.rel_shift_bnij(bd, klen=ac.shape[3])\n",
        "\n",
        "        # segment based attention score\n",
        "        if seg_mat is None:\n",
        "            ef = 0\n",
        "        else:\n",
        "            ef = torch.einsum('ibnd,snd->ibns', q_head + self.r_s_bias, self.seg_embed)\n",
        "            ef = torch.einsum('ijbs,ibns->bnij', seg_mat, ef)\n",
        "\n",
        "        # merge attention scores and perform masking\n",
        "        attn_score = (ac + bd + ef) * self.scale\n",
        "        if attn_mask is not None:\n",
        "            # attn_score = attn_score * (1 - attn_mask) - 1e30 * attn_mask\n",
        "            if attn_mask.dtype == torch.float16:\n",
        "                attn_score = attn_score - 65500 * torch.einsum('ijbn->bnij', attn_mask)\n",
        "            else:\n",
        "                attn_score = attn_score - 1e30 * torch.einsum('ijbn->bnij', attn_mask)\n",
        "\n",
        "        # attention probability\n",
        "        attn_prob = F.softmax(attn_score, dim=3)\n",
        "        attn_prob = self.dropout(attn_prob)\n",
        "\n",
        "        # Mask heads if we want to\n",
        "        if head_mask is not None:\n",
        "            attn_prob = attn_prob * torch.einsum('ijbn->bnij', head_mask)\n",
        "\n",
        "        # attention output\n",
        "        attn_vec = torch.einsum('bnij,jbnd->ibnd', attn_prob, v_head_h)\n",
        "\n",
        "        if self.output_attentions:\n",
        "            return attn_vec, torch.einsum('bnij->ijbn', attn_prob)\n",
        "\n",
        "        return attn_vec\n",
        "\n",
        "    def post_attention(self, h, attn_vec, residual=True):\n",
        "        \"\"\"Post-attention processing.\"\"\"\n",
        "        # post-attention projection (back to `d_model`)\n",
        "        attn_out = torch.einsum('ibnd,hnd->ibh', attn_vec, self.o)\n",
        "\n",
        "        attn_out = self.dropout(attn_out)\n",
        "        if residual:\n",
        "            attn_out = attn_out + h\n",
        "        output = self.layer_norm(attn_out)\n",
        "\n",
        "        return output\n",
        "\n",
        "    def forward(self, h, g,\n",
        "                      attn_mask_h, attn_mask_g,\n",
        "                      r, seg_mat,\n",
        "                      mems=None, target_mapping=None, head_mask=None):\n",
        "        if g is not None:\n",
        "            ###### Two-stream attention with relative positional encoding.\n",
        "            # content based attention score\n",
        "            if mems is not None and mems.dim() > 1:\n",
        "                cat = torch.cat([mems, h], dim=0)\n",
        "            else:\n",
        "                cat = h\n",
        "\n",
        "            # content-based key head\n",
        "            k_head_h = torch.einsum('ibh,hnd->ibnd', cat, self.k)\n",
        "\n",
        "            # content-based value head\n",
        "            v_head_h = torch.einsum('ibh,hnd->ibnd', cat, self.v)\n",
        "\n",
        "            # position-based key head\n",
        "            k_head_r = torch.einsum('ibh,hnd->ibnd', r, self.r)\n",
        "\n",
        "            ##### h-stream\n",
        "            # content-stream query head\n",
        "            q_head_h = torch.einsum('ibh,hnd->ibnd', h, self.q)\n",
        "\n",
        "            # core attention ops\n",
        "            attn_vec_h = self.rel_attn_core(\n",
        "                q_head_h, k_head_h, v_head_h, k_head_r, seg_mat=seg_mat, attn_mask=attn_mask_h, head_mask=head_mask)\n",
        "\n",
        "            if self.output_attentions:\n",
        "                attn_vec_h, attn_prob_h = attn_vec_h\n",
        "\n",
        "            # post processing\n",
        "            output_h = self.post_attention(h, attn_vec_h)\n",
        "\n",
        "            ##### g-stream\n",
        "            # query-stream query head\n",
        "            q_head_g = torch.einsum('ibh,hnd->ibnd', g, self.q)\n",
        "\n",
        "            # core attention ops\n",
        "            if target_mapping is not None:\n",
        "                q_head_g = torch.einsum('mbnd,mlb->lbnd', q_head_g, target_mapping)\n",
        "                attn_vec_g = self.rel_attn_core(\n",
        "                    q_head_g, k_head_h, v_head_h, k_head_r, seg_mat=seg_mat, attn_mask=attn_mask_g, head_mask=head_mask)\n",
        "\n",
        "                if self.output_attentions:\n",
        "                    attn_vec_g, attn_prob_g = attn_vec_g\n",
        "\n",
        "                attn_vec_g = torch.einsum('lbnd,mlb->mbnd', attn_vec_g, target_mapping)\n",
        "            else:\n",
        "                attn_vec_g = self.rel_attn_core(\n",
        "                    q_head_g, k_head_h, v_head_h, k_head_r, seg_mat=seg_mat, attn_mask=attn_mask_g, head_mask=head_mask)\n",
        "\n",
        "                if self.output_attentions:\n",
        "                    attn_vec_g, attn_prob_g = attn_vec_g\n",
        "\n",
        "            # post processing\n",
        "            output_g = self.post_attention(g, attn_vec_g)\n",
        "\n",
        "            if self.output_attentions:\n",
        "                attn_prob = attn_prob_h, attn_prob_g\n",
        "\n",
        "        else:\n",
        "            ###### Multi-head attention with relative positional encoding\n",
        "            if mems is not None and mems.dim() > 1:\n",
        "                cat = torch.cat([mems, h], dim=0)\n",
        "            else:\n",
        "                cat = h\n",
        "\n",
        "            # content heads\n",
        "            q_head_h = torch.einsum('ibh,hnd->ibnd', h, self.q)\n",
        "            k_head_h = torch.einsum('ibh,hnd->ibnd', cat, self.k)\n",
        "            v_head_h = torch.einsum('ibh,hnd->ibnd', cat, self.v)\n",
        "\n",
        "            # positional heads\n",
        "            k_head_r = torch.einsum('ibh,hnd->ibnd', r, self.r)\n",
        "\n",
        "            # core attention ops\n",
        "            attn_vec = self.rel_attn_core(\n",
        "                q_head_h, k_head_h, v_head_h, k_head_r, seg_mat=seg_mat, attn_mask=attn_mask_h, head_mask=head_mask)\n",
        "\n",
        "            if self.output_attentions:\n",
        "                attn_vec, attn_prob = attn_vec\n",
        "\n",
        "            # post processing\n",
        "            output_h = self.post_attention(h, attn_vec)\n",
        "            output_g = None\n",
        "\n",
        "        outputs = (output_h, output_g)\n",
        "        if self.output_attentions:\n",
        "            outputs = outputs + (attn_prob,)\n",
        "        return outputs\n",
        "\n",
        "class XLNetFeedForward(nn.Module):\n",
        "    def __init__(self, config):\n",
        "        super(XLNetFeedForward, self).__init__()\n",
        "        self.layer_norm = XLNetLayerNorm(config.d_model, eps=config.layer_norm_eps)\n",
        "        self.layer_1 = nn.Linear(config.d_model, config.d_inner)\n",
        "        self.layer_2 = nn.Linear(config.d_inner, config.d_model)\n",
        "        self.dropout = nn.Dropout(config.dropout)\n",
        "        if isinstance(config.ff_activation, str) or \\\n",
        "                (sys.version_info[0] == 2 and isinstance(config.ff_activation, unicode)):\n",
        "            self.activation_function = ACT2FN[config.ff_activation]\n",
        "        else:\n",
        "            self.activation_function = config.ff_activation\n",
        "\n",
        "    def forward(self, inp):\n",
        "        output = inp\n",
        "        output = self.layer_1(output)\n",
        "        output = self.activation_function(output)\n",
        "        output = self.dropout(output)\n",
        "        output = self.layer_2(output)\n",
        "        output = self.dropout(output)\n",
        "        output = self.layer_norm(output + inp)\n",
        "        return output\n",
        "\n",
        "class XLNetLayer(nn.Module):\n",
        "    def __init__(self, config):\n",
        "        super(XLNetLayer, self).__init__()\n",
        "        self.rel_attn = XLNetRelativeAttention(config)\n",
        "        self.ff = XLNetFeedForward(config)\n",
        "        self.dropout = nn.Dropout(config.dropout)\n",
        "\n",
        "    def forward(self, output_h, output_g,\n",
        "                attn_mask_h, attn_mask_g,\n",
        "                r, seg_mat, mems=None, target_mapping=None, head_mask=None):\n",
        "        outputs = self.rel_attn(output_h, output_g, attn_mask_h, attn_mask_g,\n",
        "                                r, seg_mat, mems=mems, target_mapping=target_mapping,\n",
        "                                head_mask=head_mask)\n",
        "        output_h, output_g = outputs[:2]\n",
        "\n",
        "        if output_g is not None:\n",
        "            output_g = self.ff(output_g)\n",
        "        output_h = self.ff(output_h)\n",
        "\n",
        "        outputs = (output_h, output_g) + outputs[2:]  # Add again attentions if there are there\n",
        "        return outputs\n",
        "\n",
        "\n",
        "class XLNetPreTrainedModel(PreTrainedModel):\n",
        "    \"\"\" An abstract class to handle weights initialization and\n",
        "        a simple interface for dowloading and loading pretrained models.\n",
        "    \"\"\"\n",
        "    config_class = XLNetConfig\n",
        "    pretrained_model_archive_map = XLNET_PRETRAINED_MODEL_ARCHIVE_MAP\n",
        "    load_tf_weights = load_tf_weights_in_xlnet\n",
        "    base_model_prefix = \"transformer\"\n",
        "\n",
        "    def _init_weights(self, module):\n",
        "        \"\"\" Initialize the weights.\n",
        "        \"\"\"\n",
        "        if isinstance(module, (nn.Linear, nn.Embedding)):\n",
        "            # Slightly different from the TF version which uses truncated_normal for initialization\n",
        "            # cf https://github.com/pytorch/pytorch/pull/5617\n",
        "            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)\n",
        "            if isinstance(module, nn.Linear) and module.bias is not None:\n",
        "                module.bias.data.zero_()\n",
        "        elif isinstance(module, XLNetLayerNorm):\n",
        "            module.bias.data.zero_()\n",
        "            module.weight.data.fill_(1.0)\n",
        "        elif isinstance(module, XLNetRelativeAttention):\n",
        "            for param in [module.q, module.k, module.v, module.o, module.r,\n",
        "                          module.r_r_bias, module.r_s_bias, module.r_w_bias,\n",
        "                          module.seg_embed]:\n",
        "                param.data.normal_(mean=0.0, std=self.config.initializer_range)\n",
        "        elif isinstance(module, XLNetModel):\n",
        "                module.mask_emb.data.normal_(mean=0.0, std=self.config.initializer_range)\n",
        "\n",
        "class XLNetModel(XLNetPreTrainedModel):\n",
        "    r\"\"\"\n",
        "    Outputs: `Tuple` comprising various elements depending on the configuration (config) and inputs:\n",
        "        **last_hidden_state**: ``torch.FloatTensor`` of shape ``(batch_size, sequence_length, hidden_size)``\n",
        "            Sequence of hidden-states at the last layer of the model.\n",
        "        **mems**: (`optional`, returned when ``config.mem_len > 0``)\n",
        "            list of ``torch.FloatTensor`` (one for each layer):\n",
        "            that contains pre-computed hidden-states (key and values in the attention blocks) as computed by the model\n",
        "            if config.mem_len > 0 else tuple of None. Can be used to speed up sequential decoding and attend to longer context.\n",
        "            See details in the docstring of the `mems` input above.\n",
        "        **hidden_states**: (`optional`, returned when ``config.output_hidden_states=True``)\n",
        "            list of ``torch.FloatTensor`` (one for the output of each layer + the output of the embeddings)\n",
        "            of shape ``(batch_size, sequence_length, hidden_size)``:\n",
        "            Hidden-states of the model at the output of each layer plus the initial embedding outputs.\n",
        "        **attentions**: (`optional`, returned when ``config.output_attentions=True``)\n",
        "            list of ``torch.FloatTensor`` (one for each layer) of shape ``(batch_size, num_heads, sequence_length, sequence_length)``:\n",
        "            Attentions weights after the attention softmax, used to compute the weighted average in the self-attention heads.\n",
        "    Examples::\n",
        "        tokenizer = XLNetTokenizer.from_pretrained('xlnet-large-cased')\n",
        "        model = XLNetModel.from_pretrained('xlnet-large-cased')\n",
        "        input_ids = torch.tensor(tokenizer.encode(\"Hello, my dog is cute\")).unsqueeze(0)  # Batch size 1\n",
        "        outputs = model(input_ids)\n",
        "        last_hidden_states = outputs[0]  # The last hidden-state is the first element of the output tuple\n",
        "    \"\"\"\n",
        "    def __init__(self, config):\n",
        "        super(XLNetModel, self).__init__(config)\n",
        "        self.output_attentions = config.output_attentions\n",
        "        self.output_hidden_states = config.output_hidden_states\n",
        "        self.output_past = config.output_past\n",
        "\n",
        "        self.mem_len = config.mem_len\n",
        "        self.reuse_len = config.reuse_len\n",
        "        self.d_model = config.d_model\n",
        "        self.same_length = config.same_length\n",
        "        self.attn_type = config.attn_type\n",
        "        self.bi_data = config.bi_data\n",
        "        self.clamp_len = config.clamp_len\n",
        "        self.n_layer = config.n_layer\n",
        "\n",
        "        self.word_embedding = nn.Embedding(config.n_token, config.d_model)\n",
        "        self.mask_emb = nn.Parameter(torch.FloatTensor(1, 1, config.d_model))\n",
        "        self.layer = nn.ModuleList([XLNetLayer(config) for _ in range(config.n_layer)])\n",
        "        self.dropout = nn.Dropout(config.dropout)\n",
        "\n",
        "        self.init_weights()\n",
        "\n",
        "    def _resize_token_embeddings(self, new_num_tokens):\n",
        "        self.word_embedding = self._get_resized_embeddings(self.word_embedding, new_num_tokens)\n",
        "        return self.word_embedding\n",
        "\n",
        "    def _prune_heads(self, heads_to_prune):\n",
        "        raise NotImplementedError\n",
        "\n",
        "    def create_mask(self, qlen, mlen):\n",
        "        \"\"\"\n",
        "        Creates causal attention mask. Float mask where 1.0 indicates masked, 0.0 indicates not-masked.\n",
        "        Args:\n",
        "            qlen: TODO Lysandre didn't fill\n",
        "            mlen: TODO Lysandre didn't fill\n",
        "        ::\n",
        "                  same_length=False:      same_length=True:\n",
        "                  <mlen > <  qlen >       <mlen > <  qlen >\n",
        "               ^ [0 0 0 0 0 1 1 1 1]     [0 0 0 0 0 1 1 1 1]\n",
        "                 [0 0 0 0 0 0 1 1 1]     [1 0 0 0 0 0 1 1 1]\n",
        "            qlen [0 0 0 0 0 0 0 1 1]     [1 1 0 0 0 0 0 1 1]\n",
        "                 [0 0 0 0 0 0 0 0 1]     [1 1 1 0 0 0 0 0 1]\n",
        "               v [0 0 0 0 0 0 0 0 0]     [1 1 1 1 0 0 0 0 0]\n",
        "        \"\"\"\n",
        "        attn_mask = torch.ones([qlen, qlen])\n",
        "        mask_up = torch.triu(attn_mask, diagonal=1)\n",
        "        attn_mask_pad = torch.zeros([qlen, mlen])\n",
        "        ret = torch.cat([attn_mask_pad, mask_up], dim=1)\n",
        "        if self.same_length:\n",
        "            mask_lo = torch.tril(attn_mask, diagonal=-1)\n",
        "            ret = torch.cat([ret[:, :qlen] + mask_lo, ret[:, qlen:]], dim=1)\n",
        "\n",
        "        ret = ret.to(next(self.parameters()))\n",
        "        return ret\n",
        "\n",
        "    def cache_mem(self, curr_out, prev_mem):\n",
        "        \"\"\"cache hidden states into memory.\"\"\"\n",
        "        if self.reuse_len is not None and self.reuse_len > 0:\n",
        "            curr_out = curr_out[:self.reuse_len]\n",
        "\n",
        "        if prev_mem is None:\n",
        "            new_mem = curr_out[-self.mem_len:]\n",
        "        else:\n",
        "            new_mem = torch.cat([prev_mem, curr_out], dim=0)[-self.mem_len:]\n",
        "\n",
        "        return new_mem.detach()\n",
        "\n",
        "    @staticmethod\n",
        "    def positional_embedding(pos_seq, inv_freq, bsz=None):\n",
        "        sinusoid_inp = torch.einsum('i,d->id', pos_seq, inv_freq)\n",
        "        pos_emb = torch.cat([torch.sin(sinusoid_inp), torch.cos(sinusoid_inp)], dim=-1)\n",
        "        pos_emb = pos_emb[:, None, :]\n",
        "\n",
        "        if bsz is not None:\n",
        "            pos_emb = pos_emb.expand(-1, bsz, -1)\n",
        "\n",
        "        return pos_emb\n",
        "\n",
        "    def relative_positional_encoding(self, qlen, klen, bsz=None):\n",
        "        \"\"\"create relative positional encoding.\"\"\"\n",
        "        freq_seq = torch.arange(0, self.d_model, 2.0, dtype=torch.float)\n",
        "        inv_freq = 1 / torch.pow(10000, (freq_seq / self.d_model))\n",
        "\n",
        "        if self.attn_type == 'bi':\n",
        "            # beg, end = klen - 1, -qlen\n",
        "            beg, end = klen, -qlen\n",
        "        elif self.attn_type == 'uni':\n",
        "            # beg, end = klen - 1, -1\n",
        "            beg, end = klen, -1\n",
        "        else:\n",
        "            raise ValueError('Unknown `attn_type` {}.'.format(self.attn_type))\n",
        "\n",
        "        if self.bi_data:\n",
        "            fwd_pos_seq = torch.arange(beg, end, -1.0, dtype=torch.float)\n",
        "            bwd_pos_seq = torch.arange(-beg, -end, 1.0, dtype=torch.float)\n",
        "\n",
        "            if self.clamp_len > 0:\n",
        "                fwd_pos_seq = fwd_pos_seq.clamp(-self.clamp_len, self.clamp_len)\n",
        "                bwd_pos_seq = bwd_pos_seq.clamp(-self.clamp_len, self.clamp_len)\n",
        "\n",
        "            if bsz is not None:\n",
        "                fwd_pos_emb = self.positional_embedding(fwd_pos_seq, inv_freq, bsz//2)\n",
        "                bwd_pos_emb = self.positional_embedding(bwd_pos_seq, inv_freq, bsz//2)\n",
        "            else:\n",
        "                fwd_pos_emb = self.positional_embedding(fwd_pos_seq, inv_freq)\n",
        "                bwd_pos_emb = self.positional_embedding(bwd_pos_seq, inv_freq)\n",
        "\n",
        "            pos_emb = torch.cat([fwd_pos_emb, bwd_pos_emb], dim=1)\n",
        "        else:\n",
        "            fwd_pos_seq = torch.arange(beg, end, -1.0)\n",
        "            if self.clamp_len > 0:\n",
        "                fwd_pos_seq = fwd_pos_seq.clamp(-self.clamp_len, self.clamp_len)\n",
        "            pos_emb = self.positional_embedding(fwd_pos_seq, inv_freq, bsz)\n",
        "\n",
        "        pos_emb = pos_emb.to(next(self.parameters()))\n",
        "        return pos_emb\n",
        "\n",
        "    def forward(self, input_ids, attention_mask=None, mems=None, perm_mask=None, target_mapping=None,\n",
        "                token_type_ids=None, input_mask=None, head_mask=None):\n",
        "        # the original code for XLNet uses shapes [len, bsz] with the batch dimension at the end\n",
        "        # but we want a unified interface in the library with the batch size on the first dimension\n",
        "        # so we move here the first dimension (batch) to the end\n",
        "        input_ids = input_ids.transpose(0, 1).contiguous()\n",
        "        token_type_ids = token_type_ids.transpose(0, 1).contiguous() if token_type_ids is not None else None\n",
        "        input_mask = input_mask.transpose(0, 1).contiguous() if input_mask is not None else None\n",
        "        attention_mask = attention_mask.transpose(0, 1).contiguous() if attention_mask is not None else None\n",
        "        perm_mask = perm_mask.permute(1, 2, 0).contiguous() if perm_mask is not None else None\n",
        "        target_mapping = target_mapping.permute(1, 2, 0).contiguous() if target_mapping is not None else None\n",
        "\n",
        "        qlen, bsz = input_ids.shape[0], input_ids.shape[1]\n",
        "        mlen = mems[0].shape[0] if mems is not None and mems[0] is not None else 0\n",
        "        klen = mlen + qlen\n",
        "\n",
        "        dtype_float = next(self.parameters()).dtype\n",
        "        device = next(self.parameters()).device\n",
        "\n",
        "        ##### Attention mask\n",
        "        # causal attention mask\n",
        "        if self.attn_type == 'uni':\n",
        "            attn_mask = self.create_mask(qlen, mlen)\n",
        "            attn_mask = attn_mask[:, :, None, None]\n",
        "        elif self.attn_type == 'bi':\n",
        "            attn_mask = None\n",
        "        else:\n",
        "            raise ValueError('Unsupported attention type: {}'.format(self.attn_type))\n",
        "\n",
        "        # data mask: input mask & perm mask\n",
        "        assert input_mask is None or attention_mask is None, \"You can only use one of input_mask (uses 1 for padding) \"\n",
        "        \"or attention_mask (uses 0 for padding, added for compatbility with BERT). Please choose one.\"\n",
        "        if input_mask is None and attention_mask is not None:\n",
        "            input_mask = 1.0 - attention_mask\n",
        "        if input_mask is not None and perm_mask is not None:\n",
        "            data_mask = input_mask[None] + perm_mask\n",
        "        elif input_mask is not None and perm_mask is None:\n",
        "            data_mask = input_mask[None]\n",
        "        elif input_mask is None and perm_mask is not None:\n",
        "            data_mask = perm_mask\n",
        "        else:\n",
        "            data_mask = None\n",
        "\n",
        "        if data_mask is not None:\n",
        "            # all mems can be attended to\n",
        "            if mlen > 0:\n",
        "                mems_mask = torch.zeros([data_mask.shape[0], mlen, bsz]).to(data_mask)\n",
        "                data_mask = torch.cat([mems_mask, data_mask], dim=1)\n",
        "            if attn_mask is None:\n",
        "                attn_mask = data_mask[:, :, :, None]\n",
        "            else:\n",
        "                attn_mask += data_mask[:, :, :, None]\n",
        "\n",
        "        if attn_mask is not None:\n",
        "            attn_mask = (attn_mask > 0).to(dtype_float)\n",
        "\n",
        "        if attn_mask is not None:\n",
        "            non_tgt_mask = -torch.eye(qlen).to(attn_mask)\n",
        "            if mlen > 0:\n",
        "                non_tgt_mask = torch.cat([torch.zeros([qlen, mlen]).to(attn_mask), non_tgt_mask], dim=-1)\n",
        "            non_tgt_mask = ((attn_mask + non_tgt_mask[:, :, None, None]) > 0).to(attn_mask)\n",
        "        else:\n",
        "            non_tgt_mask = None\n",
        "\n",
        "        ##### Word embeddings and prepare h & g hidden states\n",
        "        word_emb_k = self.word_embedding(input_ids)\n",
        "        output_h = self.dropout(word_emb_k)\n",
        "        if target_mapping is not None:\n",
        "            word_emb_q = self.mask_emb.expand(target_mapping.shape[0], bsz, -1)\n",
        "        # else:  # We removed the inp_q input which was same as target mapping\n",
        "        #     inp_q_ext = inp_q[:, :, None]\n",
        "        #     word_emb_q = inp_q_ext * self.mask_emb + (1 - inp_q_ext) * word_emb_k\n",
        "            output_g = self.dropout(word_emb_q)\n",
        "        else:\n",
        "            output_g = None\n",
        "\n",
        "        ##### Segment embedding\n",
        "        if token_type_ids is not None:\n",
        "            # Convert `token_type_ids` to one-hot `seg_mat`\n",
        "            if mlen > 0:\n",
        "                mem_pad = torch.zeros([mlen, bsz], dtype=torch.long, device=device)\n",
        "                cat_ids = torch.cat([mem_pad, token_type_ids], dim=0)\n",
        "            else:\n",
        "                cat_ids = token_type_ids\n",
        "\n",
        "            # `1` indicates not in the same segment [qlen x klen x bsz]\n",
        "            seg_mat = (token_type_ids[:, None] != cat_ids[None, :]).long()\n",
        "            seg_mat = F.one_hot(seg_mat, num_classes=2).to(dtype_float)\n",
        "        else:\n",
        "            seg_mat = None\n",
        "\n",
        "        ##### Positional encoding\n",
        "        pos_emb = self.relative_positional_encoding(qlen, klen, bsz=bsz)\n",
        "        pos_emb = self.dropout(pos_emb)\n",
        "\n",
        "        # Prepare head mask if needed\n",
        "        # 1.0 in head_mask indicate we keep the head\n",
        "        # attention_probs has shape bsz x n_heads x N x N\n",
        "        # input head_mask has shape [num_heads] or [num_hidden_layers x num_heads] (a head_mask for each layer)\n",
        "        # and head_mask is converted to shape [num_hidden_layers x qlen x klen x bsz x n_head]\n",
        "        if head_mask is not None:\n",
        "            if head_mask.dim() == 1:\n",
        "                head_mask = head_mask.unsqueeze(0).unsqueeze(0).unsqueeze(0).unsqueeze(0)\n",
        "                head_mask = head_mask.expand(self.n_layer, -1, -1, -1, -1)\n",
        "            elif head_mask.dim() == 2:\n",
        "                head_mask = head_mask.unsqueeze(1).unsqueeze(1).unsqueeze(1)\n",
        "            head_mask = head_mask.to(dtype=next(self.parameters()).dtype) # switch to fload if need + fp16 compatibility\n",
        "        else:\n",
        "            head_mask = [None] * self.n_layer\n",
        "\n",
        "        new_mems = ()\n",
        "        if mems is None:\n",
        "            mems = [None] * len(self.layer)\n",
        "\n",
        "        attentions = []\n",
        "        hidden_states = []\n",
        "        for i, layer_module in enumerate(self.layer):\n",
        "            if self.mem_len is not None and self.mem_len > 0 and self.output_past:\n",
        "                # cache new mems\n",
        "                new_mems = new_mems + (self.cache_mem(output_h, mems[i]),)\n",
        "            if self.output_hidden_states:\n",
        "                hidden_states.append((output_h, output_g) if output_g is not None else output_h)\n",
        "\n",
        "            outputs = layer_module(output_h, output_g, attn_mask_h=non_tgt_mask, attn_mask_g=attn_mask,\n",
        "                                   r=pos_emb, seg_mat=seg_mat, mems=mems[i], target_mapping=target_mapping,\n",
        "                                   head_mask=head_mask[i])\n",
        "            output_h, output_g = outputs[:2]\n",
        "            if self.output_attentions:\n",
        "                attentions.append(outputs[2])\n",
        "\n",
        "        # Add last hidden state\n",
        "        if self.output_hidden_states:\n",
        "            hidden_states.append((output_h, output_g) if output_g is not None else output_h)\n",
        "\n",
        "        output = self.dropout(output_g if output_g is not None else output_h)\n",
        "\n",
        "        # Prepare outputs, we transpose back here to shape [bsz, len, hidden_dim] (cf. beginning of forward() method)\n",
        "        outputs = (output.permute(1, 0, 2).contiguous(),)\n",
        "\n",
        "        if self.mem_len is not None and self.mem_len > 0 and self.output_past:\n",
        "            outputs = outputs + (new_mems,)\n",
        "\n",
        "        if self.output_hidden_states:\n",
        "            if output_g is not None:\n",
        "                hidden_states = tuple(h.permute(1, 0, 2).contiguous() for hs in hidden_states for h in hs)\n",
        "            else:\n",
        "                hidden_states = tuple(hs.permute(1, 0, 2).contiguous() for hs in hidden_states)\n",
        "            outputs = outputs + (hidden_states,)\n",
        "        if self.output_attentions:\n",
        "            attentions = tuple(t.permute(2, 3, 0, 1).contiguous() for t in attentions)\n",
        "            outputs = outputs + (attentions,)\n",
        "\n",
        "        return outputs  # outputs, (new_mems), (hidden_states), (attentions)\n",
        "\n",
        "## Our implementation of XLNetForTokenClassification\n",
        "class XLNetForTokenClassification(XLNetPreTrainedModel):\n",
        "  def __init__(self, config):\n",
        "        super(XLNetForTokenClassification, self).__init__(config)\n",
        "        self.num_labels = config.num_labels\n",
        "        self.transformer = XLNetModel(config)\n",
        "        self.dropout = nn.Dropout(config.dropout)\n",
        "        self.logits_proj = nn.Linear(config.d_model, config.num_labels)\n",
        "        self.init_weights()\n",
        "\n",
        "  def forward(self, input_ids, attention_mask=None, mems=None, perm_mask=None, target_mapping=None,\n",
        "                token_type_ids=None, input_mask=None, head_mask=None, labels=None):\n",
        "        transformer_outputs = self.transformer(input_ids,\n",
        "                                               attention_mask=attention_mask,\n",
        "                                               mems=mems,\n",
        "                                               perm_mask=perm_mask,\n",
        "                                               target_mapping=target_mapping,\n",
        "                                               token_type_ids=token_type_ids,\n",
        "                                               input_mask=input_mask,\n",
        "                                               head_mask=head_mask)\n",
        "        \n",
        "        output = transformer_outputs[0]\n",
        "        output = self.dropout(output)\n",
        "        logits = self.logits_proj(output)\n",
        "\n",
        "        return (logits,)\n",
        "\n",
        "class CueModel:\n",
        "    def __init__(self, full_finetuning = True, train = False, pretrained_model_path = 'Cue_Detection.pickle', device = 'cuda', learning_rate = 3e-5, class_weight = [100, 100, 100, 1, 0], num_labels = 5):\n",
        "        self.model_name = CUE_MODEL\n",
        "        self.task = TASK\n",
        "        if train == True:\n",
        "            if 'xlnet' in CUE_MODEL:\n",
        "                self.model = XLNetForTokenClassification.from_pretrained(CUE_MODEL, num_labels=num_labels, cache_dir = 'xlnet-base-cased-model')\n",
        "            elif 'roberta' in CUE_MODEL:\n",
        "                self.model = RobertaForTokenClassification.from_pretrained(CUE_MODEL, num_labels=num_labels, cache_dir = 'roberta-base-model')\n",
        "            elif 'bert' in CUE_MODEL:\n",
        "                self.model = BertForTokenClassification.from_pretrained(CUE_MODEL, num_labels=num_labels, cache_dir = 'bert_base_uncased_model')\n",
        "            else:\n",
        "                raise ValueError(\"Supported model types are: xlnet, roberta, bert\")\n",
        "        else:\n",
        "            self.model = torch.load(pretrained_model_path)\n",
        "        self.device = torch.device(device)\n",
        "        self.class_weight = class_weight\n",
        "        self.learning_rate = learning_rate\n",
        "        self.num_labels = num_labels\n",
        "        if device == 'cuda':\n",
        "            self.model.cuda()\n",
        "        else:\n",
        "            self.model.cpu()\n",
        "            \n",
        "        if full_finetuning:\n",
        "            param_optimizer = list(self.model.named_parameters())\n",
        "            no_decay = ['bias', 'gamma', 'beta']\n",
        "            optimizer_grouped_parameters = [\n",
        "                {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)],\n",
        "                 'weight_decay_rate': 0.01},\n",
        "                {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)],\n",
        "                 'weight_decay_rate': 0.0}\n",
        "            ]\n",
        "        else:\n",
        "            if intermediate_neurons == None:\n",
        "                param_optimizer = list(self.model.classifier.named_parameters()) \n",
        "            else:\n",
        "                param_optimizer = list(self.model.classifier.named_parameters())+list(self.model.int_layer.named_parameters())\n",
        "            optimizer_grouped_parameters = [{\"params\": [p for n, p in param_optimizer]}]\n",
        "        self.optimizer = Adam(optimizer_grouped_parameters, lr=learning_rate)\n",
        "\n",
        "    #@telegram_sender(token=TELEGRAM_TOKEN, chat_id=TELEGRAM_CHAT_ID)  \n",
        "    def train(self, train_dataloader, valid_dataloaders, train_dl_name, val_dl_name, epochs = 5, max_grad_norm = 1.0, patience = 3):\n",
        "        \n",
        "        self.train_dl_name = train_dl_name\n",
        "        return_dict = {\"Task\": f\"{self.task} Cue Detection\",\n",
        "                       \"Model\": self.model_name,\n",
        "                       \"Train Dataset\": train_dl_name,\n",
        "                       \"Val Dataset\": val_dl_name,\n",
        "                       \"Best Precision\": 0,\n",
        "                       \"Best Recall\": 0,\n",
        "                       \"Best F1\": 0}\n",
        "        train_loss = []\n",
        "        valid_loss = []\n",
        "        early_stopping = EarlyStopping(patience=patience, verbose=True)\n",
        "        loss_fn = CrossEntropyLoss(weight=torch.Tensor(self.class_weight).to(self.device))\n",
        "        for _ in tqdm(range(epochs), desc=\"Epoch\"):\n",
        "            self.model.train()\n",
        "            tr_loss = 0\n",
        "            nb_tr_examples, nb_tr_steps = 0, 0\n",
        "            for step, batch in enumerate(train_dataloader):\n",
        "                batch = tuple(t.to(self.device) for t in batch)\n",
        "                b_input_ids, b_input_mask, b_labels, b_mymasks = batch\n",
        "                logits = self.model(b_input_ids, token_type_ids=None,attention_mask=b_input_mask)[0]\n",
        "                active_loss = b_input_mask.view(-1) == 1\n",
        "                active_logits = logits.view(-1, self.num_labels)[active_loss] #5 is num_labels\n",
        "                active_labels = b_labels.view(-1)[active_loss]\n",
        "                loss = loss_fn(active_logits, active_labels)\n",
        "                loss.backward()\n",
        "                tr_loss += loss.item()\n",
        "                if step % 100 == 0:\n",
        "                    print(f\"Batch {step}, loss {loss.item()}\")\n",
        "                train_loss.append(loss.item())\n",
        "                nb_tr_examples += b_input_ids.size(0)\n",
        "                nb_tr_steps += 1\n",
        "                torch.nn.utils.clip_grad_norm_(parameters=self.model.parameters(), max_norm=max_grad_norm)\n",
        "                self.optimizer.step()\n",
        "                self.model.zero_grad()\n",
        "            print(\"Train loss: {}\".format(tr_loss/nb_tr_steps))\n",
        "            \n",
        "            self.model.eval()\n",
        "            eval_loss, eval_accuracy, eval_scope_accuracy, eval_positive_cue_accuracy = 0, 0, 0, 0\n",
        "            nb_eval_steps, nb_eval_examples, steps_positive_cue_accuracy = 0, 0, 0\n",
        "            predictions , true_labels, ip_mask = [], [], []\n",
        "            for valid_dataloader in valid_dataloaders:\n",
        "                for batch in valid_dataloader:\n",
        "                    batch = tuple(t.to(self.device) for t in batch)\n",
        "                    b_input_ids, b_input_mask, b_labels, b_mymasks = batch\n",
        "\n",
        "                    with torch.no_grad():\n",
        "                        logits = self.model(b_input_ids, token_type_ids=None, attention_mask=b_input_mask)[0]\n",
        "                        active_loss = b_input_mask.view(-1) == 1\n",
        "                        active_logits = logits.view(-1, self.num_labels)[active_loss] #5 is num_labels\n",
        "                        active_labels = b_labels.view(-1)[active_loss]\n",
        "                        tmp_eval_loss = loss_fn(active_logits, active_labels)\n",
        "                        \n",
        "                    logits = logits.detach().cpu().numpy()\n",
        "                    label_ids = b_labels.to('cpu').numpy()\n",
        "                    mymasks = b_mymasks.to('cpu').numpy()\n",
        "                    \n",
        "                    if F1_METHOD == 'first_token':\n",
        "\n",
        "                        logits = [list(p) for p in np.argmax(logits, axis=2)]\n",
        "                        actual_logits = []\n",
        "                        actual_label_ids = []\n",
        "                        for l,lid,m in zip(logits, label_ids, mymasks):\n",
        "                            actual_logits.append([i for i,j in zip(l,m) if j==1])\n",
        "                            actual_label_ids.append([i for i,j in zip(lid, m) if j==1])\n",
        "\n",
        "                        logits = actual_logits\n",
        "                        label_ids = actual_label_ids\n",
        "\n",
        "                        predictions.append(logits)\n",
        "                        true_labels.append(label_ids)\n",
        "                    \n",
        "                    elif F1_METHOD == 'average':\n",
        "\n",
        "                        logits = [list(p) for p in logits]\n",
        "                        \n",
        "                        actual_logits = []\n",
        "                        actual_label_ids = []\n",
        "                        \n",
        "                        for l,lid,m in zip(logits, label_ids, mymasks):\n",
        "                            \n",
        "                            actual_label_ids.append([i for i,j in zip(lid, m) if j==1])\n",
        "                            curr_preds = []\n",
        "                            my_logits = []\n",
        "                            in_split = 0\n",
        "                            for i,j in zip(l,m):\n",
        "                                if j==1:\n",
        "                                    if in_split == 1:\n",
        "                                        if len(my_logits)>0:\n",
        "                                            curr_preds.append(my_logits[-1])\n",
        "                                        mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                                        if len(my_logits)>0:\n",
        "                                            my_logits[-1] = mode_pred\n",
        "                                        else:\n",
        "                                            my_logits.append(mode_pred)\n",
        "                                        curr_preds = []\n",
        "                                        in_split = 0\n",
        "                                    my_logits.append(np.argmax(i))\n",
        "                                if j==0:\n",
        "                                    curr_preds.append(i)\n",
        "                                    in_split = 1\n",
        "                            if in_split == 1:\n",
        "                                if len(my_logits)>0:\n",
        "                                    curr_preds.append(my_logits[-1])\n",
        "                                mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                                if len(my_logits)>0:\n",
        "                                    my_logits[-1] = mode_pred\n",
        "                                else:\n",
        "                                    my_logits.append(mode_pred)\n",
        "                            actual_logits.append(my_logits)\n",
        "                            \n",
        "                        logits = actual_logits\n",
        "                        label_ids = actual_label_ids\n",
        "                        \n",
        "                        predictions.append(logits)\n",
        "                        true_labels.append(label_ids)\n",
        "                    \n",
        "                    tmp_eval_accuracy = flat_accuracy(logits, label_ids)\n",
        "                    tmp_eval_positive_cue_accuracy = flat_accuracy_positive_cues(logits, label_ids)\n",
        "                    eval_loss += tmp_eval_loss.mean().item()\n",
        "                    valid_loss.append(tmp_eval_loss.mean().item())\n",
        "                    eval_accuracy += tmp_eval_accuracy\n",
        "                    if tmp_eval_positive_cue_accuracy!=None:\n",
        "                        eval_positive_cue_accuracy+=tmp_eval_positive_cue_accuracy\n",
        "                        steps_positive_cue_accuracy+=1\n",
        "                    nb_eval_examples += b_input_ids.size(0)\n",
        "                    nb_eval_steps += 1\n",
        "                eval_loss = eval_loss/nb_eval_steps\n",
        "                \n",
        "            print(\"Validation loss: {}\".format(eval_loss))\n",
        "            print(\"Validation Accuracy: {}\".format(eval_accuracy/nb_eval_steps))\n",
        "            print(\"Validation Accuracy for Positive Cues: {}\".format(eval_positive_cue_accuracy/steps_positive_cue_accuracy))\n",
        "            labels_flat = [l_ii for l in true_labels for l_i in l for l_ii in l_i]\n",
        "            pred_flat = [p_ii for p in predictions for p_i in p for p_ii in p_i]\n",
        "            pred_flat = [p for p,l in zip(pred_flat, labels_flat) if l!=4]\n",
        "            labels_flat = [l for l in labels_flat if l!=4]\n",
        "            report_per_class_accuracy(labels_flat, pred_flat)\n",
        "            print(classification_report(labels_flat, pred_flat))\n",
        "            print(\"F1-Score Overall: {}\".format(f1_score(labels_flat,pred_flat, average='weighted')))\n",
        "            p,r,f1 = f1_cues(labels_flat, pred_flat)\n",
        "            if f1>return_dict['Best F1']:\n",
        "                return_dict['Best F1'] = f1\n",
        "                return_dict['Best Precision'] = p\n",
        "                return_dict['Best Recall'] = r\n",
        "            early_stopping(f1, self.model)\n",
        "        \n",
        "            if early_stopping.early_stop:\n",
        "                print(\"Early stopping\")\n",
        "                break\n",
        "\n",
        "            labels_flat = [int(i!=3) for i in labels_flat]\n",
        "            pred_flat = [int(i!=3) for i in pred_flat]\n",
        "            print(\"F1-Score Cue_No Cue: {}\".format(f1_score(labels_flat,pred_flat, average='weighted')))\n",
        "            \n",
        "        self.model.load_state_dict(torch.load('checkpoint.pt'))\n",
        "        plt.xlabel(\"Iteration\")\n",
        "        plt.ylabel(\"Train Loss\")\n",
        "        plt.plot([i for i in range(len(train_loss))], train_loss)\n",
        "        plt.figure()\n",
        "        plt.xlabel(\"Iteration\")\n",
        "        plt.ylabel(\"Validation Loss\")\n",
        "        plt.plot([i for i in range(len(valid_loss))], valid_loss)\n",
        "        return return_dict\n",
        "\n",
        "    #@telegram_sender(token=TELEGRAM_TOKEN, chat_id=TELEGRAM_CHAT_ID)\n",
        "    def evaluate(self, test_dataloader, test_dl_name):\n",
        "        return_dict = {\"Task\": f\"{self.task} Cue Detection\",\n",
        "                       \"Model\": self.model_name,\n",
        "                       \"Train Dataset\": self.train_dl_name,\n",
        "                       \"Test Dataset\": test_dl_name,\n",
        "                       \"Precision\": 0,\n",
        "                       \"Recall\": 0,\n",
        "                       \"F1\": 0}\n",
        "        self.model.eval()\n",
        "        eval_loss, eval_accuracy, eval_scope_accuracy, eval_positive_cue_accuracy = 0, 0, 0, 0\n",
        "        nb_eval_steps, nb_eval_examples, steps_positive_cue_accuracy = 0, 0, 0\n",
        "        predictions , true_labels, ip_mask = [], [], []\n",
        "        loss_fn = CrossEntropyLoss(weight=torch.Tensor(self.class_weight).to(self.device))\n",
        "        for batch in test_dataloader:\n",
        "            batch = tuple(t.to(self.device) for t in batch)\n",
        "            b_input_ids, b_input_mask, b_labels, b_mymasks = batch\n",
        "            \n",
        "            with torch.no_grad():\n",
        "                logits = self.model(b_input_ids, token_type_ids=None,attention_mask=b_input_mask)[0]\n",
        "                active_loss = b_input_mask.view(-1) == 1\n",
        "                active_logits = logits.view(-1, self.num_labels)[active_loss] #5 is num_labels\n",
        "                active_labels = b_labels.view(-1)[active_loss]\n",
        "                tmp_eval_loss = loss_fn(active_logits, active_labels)\n",
        "                logits = logits.detach().cpu().numpy()\n",
        "            label_ids = b_labels.to('cpu').numpy()\n",
        "            mymasks = b_mymasks.to('cpu').numpy()\n",
        "\n",
        "            if F1_METHOD == 'first_token':\n",
        "\n",
        "                logits = [list(p) for p in np.argmax(logits, axis=2)]\n",
        "                actual_logits = []\n",
        "                actual_label_ids = []\n",
        "                for l,lid,m in zip(logits, label_ids, mymasks):\n",
        "                    actual_logits.append([i for i,j in zip(l,m) if j==1])\n",
        "                    actual_label_ids.append([i for i,j in zip(lid, m) if j==1])\n",
        "\n",
        "                logits = actual_logits\n",
        "                label_ids = actual_label_ids\n",
        "\n",
        "                predictions.append(logits)\n",
        "                true_labels.append(label_ids)\n",
        "\n",
        "            elif F1_METHOD == 'average':\n",
        "                logits = [list(p) for p in logits]\n",
        "                    \n",
        "                actual_logits = []\n",
        "                actual_label_ids = []\n",
        "                for l,lid,m in zip(logits, label_ids, mymasks):\n",
        "                        \n",
        "                    actual_label_ids.append([i for i,j in zip(lid, m) if j==1])\n",
        "                    my_logits = []\n",
        "                    curr_preds = []\n",
        "                    in_split = 0\n",
        "                    for i,j in zip(l,m):\n",
        "                        if j==1:\n",
        "                            if in_split == 1:\n",
        "                                if len(my_logits)>0:\n",
        "                                    curr_preds.append(my_logits[-1])\n",
        "                                mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                                if len(my_logits)>0:\n",
        "                                    my_logits[-1] = mode_pred\n",
        "                                else:\n",
        "                                    my_logits.append(mode_pred)\n",
        "                                curr_preds = []\n",
        "                                in_split = 0\n",
        "                            my_logits.append(np.argmax(i))\n",
        "                        if j==0:\n",
        "                            curr_preds.append(i)\n",
        "                            in_split = 1\n",
        "                    if in_split == 1:\n",
        "                        if len(my_logits)>0:\n",
        "                            curr_preds.append(my_logits[-1])\n",
        "                        mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                        if len(my_logits)>0:\n",
        "                            my_logits[-1] = mode_pred\n",
        "                        else:\n",
        "                            my_logits.append(mode_pred)\n",
        "                    actual_logits.append(my_logits)\n",
        "\n",
        "                logits = actual_logits\n",
        "                label_ids = actual_label_ids\n",
        "                \n",
        "                predictions.append(logits)\n",
        "                true_labels.append(label_ids)    \n",
        "                \n",
        "            tmp_eval_accuracy = flat_accuracy(logits, label_ids)\n",
        "            tmp_eval_positive_cue_accuracy = flat_accuracy_positive_cues(logits, label_ids)\n",
        "        \n",
        "            eval_loss += tmp_eval_loss.mean().item()\n",
        "            eval_accuracy += tmp_eval_accuracy\n",
        "            if tmp_eval_positive_cue_accuracy != None:\n",
        "                eval_positive_cue_accuracy += tmp_eval_positive_cue_accuracy\n",
        "                steps_positive_cue_accuracy+=1\n",
        "            nb_eval_examples += b_input_ids.size(0)\n",
        "            nb_eval_steps += 1\n",
        "        eval_loss = eval_loss/nb_eval_steps\n",
        "        print(\"Validation loss: {}\".format(eval_loss))\n",
        "        print(\"Validation Accuracy: {}\".format(eval_accuracy/nb_eval_steps))\n",
        "        print(\"Validation Accuracy for Positive Cues: {}\".format(eval_positive_cue_accuracy/steps_positive_cue_accuracy))\n",
        "        labels_flat = [l_ii for l in true_labels for l_i in l for l_ii in l_i]\n",
        "        pred_flat = [p_ii for p in predictions for p_i in p for p_ii in p_i]\n",
        "        pred_flat = [p for p,l in zip(pred_flat, labels_flat) if l!=4]\n",
        "        labels_flat = [l for l in labels_flat if l!=4]\n",
        "        report_per_class_accuracy(labels_flat, pred_flat)\n",
        "        print(classification_report(labels_flat, pred_flat))\n",
        "        print(\"F1-Score: {}\".format(f1_score(labels_flat,pred_flat,average='weighted')))\n",
        "        p,r,f1 = f1_cues(labels_flat, pred_flat)\n",
        "        return_dict['Precision'] = p\n",
        "        return_dict['Recall'] = r\n",
        "        return_dict['F1'] = f1\n",
        "        labels_flat = [int(i!=3) for i in labels_flat]\n",
        "        pred_flat = [int(i!=3) for i in pred_flat]\n",
        "        print(\"F1-Score Cue_No Cue: {}\".format(f1_score(labels_flat,pred_flat,average='weighted')))\n",
        "        return return_dict\n",
        "\n",
        "    def predict(self, dataloader):\n",
        "        self.model.eval()\n",
        "        predictions, ip_mask = [], []\n",
        "        for batch in dataloader:\n",
        "            batch = tuple(t.to(self.device) for t in batch)\n",
        "            b_input_ids, b_input_mask, b_mymasks = batch\n",
        "\n",
        "            with torch.no_grad():\n",
        "                logits = self.model(b_input_ids, token_type_ids=None, attention_mask=b_input_mask)[0]\n",
        "            logits = logits.detach().cpu().numpy()\n",
        "            mymasks = b_mymasks.to('cpu').numpy()\n",
        "            #predictions.extend([list(p) for p in np.argmax(logits, axis=2)])\n",
        "            if F1_METHOD == 'first_token':\n",
        "\n",
        "                logits = [list(p) for p in np.argmax(logits, axis=2)]\n",
        "                actual_logits = []\n",
        "                for l,m in zip(logits, mymasks):\n",
        "                    actual_logits.append([i for i,j in zip(l,m) if j==1])\n",
        "                \n",
        "                predictions.append(actual_logits)\n",
        "\n",
        "            elif F1_METHOD == 'average':\n",
        "                logits = [list(p) for p in logits]\n",
        "                    \n",
        "                actual_logits = []\n",
        "                actual_label_ids = []\n",
        "                for l,m in zip(logits, mymasks):\n",
        "                    my_logits = []\n",
        "                    curr_preds = []\n",
        "                    in_split = 0\n",
        "                    for i,j in zip(l,m):\n",
        "                        if j==1:\n",
        "                            if in_split == 1:\n",
        "                                if len(my_logits)>0:\n",
        "                                    curr_preds.append(my_logits[-1])\n",
        "                                mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                                if len(my_logits)>0:\n",
        "                                    my_logits[-1] = mode_pred\n",
        "                                else:\n",
        "                                    my_logits.append(mode_pred)\n",
        "                                curr_preds = []\n",
        "                                in_split = 0\n",
        "                            my_logits.append(np.argmax(i))\n",
        "                        if j==0:\n",
        "                            curr_preds.append(i)\n",
        "                            in_split = 1\n",
        "                    if in_split == 1:\n",
        "                        if len(my_logits)>0:\n",
        "                            curr_preds.append(my_logits[-1])\n",
        "                        mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                        if len(my_logits)>0:\n",
        "                            my_logits[-1] = mode_pred\n",
        "                        else:\n",
        "                            my_logits.append(mode_pred)\n",
        "                    actual_logits.append(my_logits)\n",
        "                predictions.append(actual_logits)\n",
        "                \n",
        "        return predictions\n",
        "\n",
        "class ScopeModel:\n",
        "    def __init__(self, full_finetuning = True, train = False, pretrained_model_path = 'Scope_Resolution_Augment.pickle', device = 'cuda', learning_rate = 3e-5):\n",
        "        self.model_name = SCOPE_MODEL\n",
        "        self.task = TASK\n",
        "        self.num_labels = 2\n",
        "        if train == True:\n",
        "            if 'xlnet' in SCOPE_MODEL:\n",
        "                self.model = XLNetForTokenClassification.from_pretrained(SCOPE_MODEL, num_labels=self.num_labels, cache_dir = 'xlnet-base-cased-model')\n",
        "            elif 'roberta' in SCOPE_MODEL:\n",
        "                self.model = RobertaForTokenClassification.from_pretrained(SCOPE_MODEL, num_labels=self.num_labels, cache_dir = 'roberta-base-model')\n",
        "            elif 'bert' in SCOPE_MODEL:\n",
        "                self.model = BertForTokenClassification.from_pretrained(SCOPE_MODEL, num_labels=self.num_labels, cache_dir = 'bert_base_uncased_model')\n",
        "            else:\n",
        "                raise ValueError(\"Supported model types are: xlnet, roberta, bert\")\n",
        "        else:\n",
        "            self.model = torch.load(pretrained_model_path)\n",
        "        self.device = torch.device(device)\n",
        "        if device=='cuda':\n",
        "            self.model.cuda()\n",
        "        else:\n",
        "            self.model.cpu()\n",
        "\n",
        "        if full_finetuning:\n",
        "            param_optimizer = list(self.model.named_parameters())\n",
        "            no_decay = ['bias', 'gamma', 'beta']\n",
        "            optimizer_grouped_parameters = [\n",
        "                {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)],\n",
        "                 'weight_decay_rate': 0.01},\n",
        "                {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)],\n",
        "                 'weight_decay_rate': 0.0}\n",
        "            ]\n",
        "        else:\n",
        "            param_optimizer = list(self.model.classifier.named_parameters()) \n",
        "            optimizer_grouped_parameters = [{\"params\": [p for n, p in param_optimizer]}]\n",
        "        self.optimizer = Adam(optimizer_grouped_parameters, lr=learning_rate)\n",
        "\n",
        "    #@telegram_sender(token=TELEGRAM_TOKEN, chat_id=TELEGRAM_CHAT_ID)    \n",
        "    def train(self, train_dataloader, valid_dataloaders, train_dl_name, val_dl_name, epochs = 5, max_grad_norm = 1.0, patience = 3):\n",
        "        self.train_dl_name = train_dl_name\n",
        "        return_dict = {\"Task\": f\"{self.task} Scope Resolution\",\n",
        "                       \"Model\": self.model_name,\n",
        "                       \"Train Dataset\": train_dl_name,\n",
        "                       \"Val Dataset\": val_dl_name,\n",
        "                       \"Best Precision\": 0,\n",
        "                       \"Best Recall\": 0,\n",
        "                       \"Best F1\": 0}\n",
        "        train_loss = []\n",
        "        valid_loss = []\n",
        "        early_stopping = EarlyStopping(patience=patience, verbose=True)\n",
        "        loss_fn = CrossEntropyLoss()\n",
        "        for _ in tqdm(range(epochs), desc=\"Epoch\"):\n",
        "            self.model.train()\n",
        "            tr_loss = 0\n",
        "            nb_tr_examples, nb_tr_steps = 0, 0\n",
        "            for step, batch in enumerate(train_dataloader):\n",
        "                batch = tuple(t.to(self.device) for t in batch)\n",
        "                b_input_ids, b_input_mask, b_labels, b_mymasks = batch\n",
        "                logits = self.model(b_input_ids, token_type_ids=None,\n",
        "                             attention_mask=b_input_mask)[0]\n",
        "                active_loss = b_input_mask.view(-1) == 1\n",
        "                active_logits = logits.view(-1, self.num_labels)[active_loss] #2 is num_labels\n",
        "                active_labels = b_labels.view(-1)[active_loss]\n",
        "                loss = loss_fn(active_logits, active_labels)\n",
        "                loss.backward()\n",
        "                tr_loss += loss.item()\n",
        "                train_loss.append(loss.item())\n",
        "                if step%100 == 0:\n",
        "                    print(f\"Batch {step}, loss {loss.item()}\")\n",
        "                nb_tr_examples += b_input_ids.size(0)\n",
        "                nb_tr_steps += 1\n",
        "                torch.nn.utils.clip_grad_norm_(parameters=self.model.parameters(), max_norm=max_grad_norm)\n",
        "                self.optimizer.step()\n",
        "                self.model.zero_grad()\n",
        "            print(\"Train loss: {}\".format(tr_loss/nb_tr_steps))\n",
        "            \n",
        "            self.model.eval()\n",
        "            \n",
        "            eval_loss, eval_accuracy, eval_scope_accuracy = 0, 0, 0\n",
        "            nb_eval_steps, nb_eval_examples = 0, 0\n",
        "            predictions , true_labels, ip_mask = [], [], []\n",
        "            loss_fn = CrossEntropyLoss()\n",
        "            for valid_dataloader in valid_dataloaders:\n",
        "                for batch in valid_dataloader:\n",
        "                    batch = tuple(t.to(self.device) for t in batch)\n",
        "                    b_input_ids, b_input_mask, b_labels, b_mymasks = batch\n",
        "\n",
        "                    with torch.no_grad():\n",
        "                        logits = self.model(b_input_ids, token_type_ids=None,\n",
        "                                      attention_mask=b_input_mask)[0]\n",
        "                        active_loss = b_input_mask.view(-1) == 1\n",
        "                        active_logits = logits.view(-1, self.num_labels)[active_loss]\n",
        "                        active_labels = b_labels.view(-1)[active_loss]\n",
        "                        tmp_eval_loss = loss_fn(active_logits, active_labels)\n",
        "                        \n",
        "                    logits = logits.detach().cpu().numpy()\n",
        "                    label_ids = b_labels.to('cpu').numpy()\n",
        "                    b_input_ids = b_input_ids.to('cpu').numpy()\n",
        "\n",
        "                    mymasks = b_mymasks.to('cpu').numpy()\n",
        "                        \n",
        "                    if F1_METHOD == 'first_token':\n",
        "\n",
        "                        logits = [list(p) for p in np.argmax(logits, axis=2)]\n",
        "                        actual_logits = []\n",
        "                        actual_label_ids = []\n",
        "                        for l,lid,m in zip(logits, label_ids, mymasks):\n",
        "                            actual_logits.append([i for i,j in zip(l,m) if j==1])\n",
        "                            actual_label_ids.append([i for i,j in zip(lid, m) if j==1])\n",
        "\n",
        "                        logits = actual_logits\n",
        "                        label_ids = actual_label_ids\n",
        "\n",
        "                        predictions.append(logits)\n",
        "                        true_labels.append(label_ids)\n",
        "                    elif F1_METHOD == 'average':\n",
        "                      \n",
        "                        logits = [list(p) for p in logits]\n",
        "                    \n",
        "                        actual_logits = []\n",
        "                        actual_label_ids = []\n",
        "                        \n",
        "                        for l,lid,m,b_ii in zip(logits, label_ids, mymasks, b_input_ids):\n",
        "                                \n",
        "                            actual_label_ids.append([i for i,j in zip(lid, m) if j==1])\n",
        "                            my_logits = []\n",
        "                            curr_preds = []\n",
        "                            in_split = 0\n",
        "                            for i,j,k in zip(l,m, b_ii):\n",
        "                                '''if k == 0:\n",
        "                                    break'''\n",
        "                                if j==1:\n",
        "                                    if in_split == 1:\n",
        "                                        if len(my_logits)>0:\n",
        "                                            curr_preds.append(my_logits[-1])\n",
        "                                        mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                                        if len(my_logits)>0:\n",
        "                                            my_logits[-1] = mode_pred\n",
        "                                        else:\n",
        "                                            my_logits.append(mode_pred)\n",
        "                                        curr_preds = []\n",
        "                                        in_split = 0\n",
        "                                    my_logits.append(np.argmax(i))\n",
        "                                if j==0:\n",
        "                                    curr_preds.append(i)\n",
        "                                    in_split = 1\n",
        "                            if in_split == 1:\n",
        "                                if len(my_logits)>0:\n",
        "                                    curr_preds.append(my_logits[-1])\n",
        "                                mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                                if len(my_logits)>0:\n",
        "                                    my_logits[-1] = mode_pred\n",
        "                                else:\n",
        "                                    my_logits.append(mode_pred)\n",
        "                            actual_logits.append(my_logits)\n",
        "                            \n",
        "                        predictions.append(actual_logits)\n",
        "                        true_labels.append(actual_label_ids)    \n",
        "                        \n",
        "                    tmp_eval_accuracy = flat_accuracy(actual_logits, actual_label_ids)\n",
        "                    tmp_eval_scope_accuracy = scope_accuracy(actual_logits, actual_label_ids)\n",
        "                    eval_scope_accuracy += tmp_eval_scope_accuracy\n",
        "                    valid_loss.append(tmp_eval_loss.mean().item())\n",
        "\n",
        "                    eval_loss += tmp_eval_loss.mean().item()\n",
        "                    eval_accuracy += tmp_eval_accuracy\n",
        "\n",
        "                    nb_eval_examples += len(b_input_ids)\n",
        "                    nb_eval_steps += 1\n",
        "                eval_loss = eval_loss/nb_eval_steps\n",
        "            print(\"Validation loss: {}\".format(eval_loss))\n",
        "            print(\"Validation Accuracy: {}\".format(eval_accuracy/nb_eval_steps))\n",
        "            print(\"Validation Accuracy Scope Level: {}\".format(eval_scope_accuracy/nb_eval_steps))\n",
        "            f1_scope([j for i in true_labels for j in i], [j for i in predictions for j in i], level='scope')\n",
        "            labels_flat = [l_ii for l in true_labels for l_i in l for l_ii in l_i]\n",
        "            pred_flat = [p_ii for p in predictions for p_i in p for p_ii in p_i]\n",
        "            classification_dict = classification_report(labels_flat, pred_flat, output_dict= True)\n",
        "            p = classification_dict[\"1\"][\"precision\"]\n",
        "            r = classification_dict[\"1\"][\"recall\"]\n",
        "            f1 = classification_dict[\"1\"][\"f1-score\"]\n",
        "            if f1>return_dict['Best F1']:\n",
        "                return_dict['Best F1'] = f1\n",
        "                return_dict['Best Precision'] = p\n",
        "                return_dict['Best Recall'] = r\n",
        "            print(\"F1-Score Token: {}\".format(f1))\n",
        "            print(classification_report(labels_flat, pred_flat))\n",
        "            early_stopping(f1, self.model)\n",
        "            if early_stopping.early_stop:\n",
        "                print(\"Early stopping\")\n",
        "                break\n",
        "        \n",
        "        self.model.load_state_dict(torch.load('checkpoint.pt'))\n",
        "        plt.xlabel(\"Iteration\")\n",
        "        plt.ylabel(\"Train Loss\")\n",
        "        plt.plot([i for i in range(len(train_loss))], train_loss)\n",
        "        plt.figure()\n",
        "        plt.xlabel(\"Iteration\")\n",
        "        plt.ylabel(\"Validation Loss\")\n",
        "        plt.plot([i for i in range(len(valid_loss))], valid_loss)\n",
        "        return return_dict\n",
        "\n",
        "    #@telegram_sender(token=TELEGRAM_TOKEN, chat_id=TELEGRAM_CHAT_ID)\n",
        "    def evaluate(self, test_dataloader, test_dl_name = \"SFU\"):\n",
        "        return_dict = {\"Task\": f\"{self.task} Scope Resolution\",\n",
        "                       \"Model\": self.model_name,\n",
        "                       \"Train Dataset\": self.train_dl_name,\n",
        "                       \"Test Dataset\": test_dl_name,\n",
        "                       \"Precision\": 0,\n",
        "                       \"Recall\": 0,\n",
        "                       \"F1\": 0}\n",
        "        self.model.eval()\n",
        "        eval_loss, eval_accuracy, eval_scope_accuracy = 0, 0, 0\n",
        "        nb_eval_steps, nb_eval_examples = 0, 0\n",
        "        predictions , true_labels, ip_mask = [], [], []\n",
        "        loss_fn = CrossEntropyLoss()\n",
        "        for batch in test_dataloader:\n",
        "            batch = tuple(t.to(self.device) for t in batch)\n",
        "            b_input_ids, b_input_mask, b_labels, b_mymasks = batch\n",
        "\n",
        "            with torch.no_grad():\n",
        "                logits = self.model(b_input_ids, token_type_ids=None,\n",
        "                               attention_mask=b_input_mask)[0]\n",
        "                active_loss = b_input_mask.view(-1) == 1\n",
        "                active_logits = logits.view(-1, self.num_labels)[active_loss] #5 is num_labels\n",
        "                active_labels = b_labels.view(-1)[active_loss]\n",
        "                tmp_eval_loss = loss_fn(active_logits, active_labels)\n",
        "                \n",
        "            logits = logits.detach().cpu().numpy()\n",
        "            label_ids = b_labels.to('cpu').numpy()\n",
        "            b_input_ids = b_input_ids.to('cpu').numpy()\n",
        "            \n",
        "            mymasks = b_mymasks.to('cpu').numpy()\n",
        "\n",
        "            if F1_METHOD == 'first_token':\n",
        "\n",
        "                logits = [list(p) for p in np.argmax(logits, axis=2)]\n",
        "                actual_logits = []\n",
        "                actual_label_ids = []\n",
        "                for l,lid,m in zip(logits, label_ids, mymasks):\n",
        "                    actual_logits.append([i for i,j in zip(l,m) if j==1])\n",
        "                    actual_label_ids.append([i for i,j in zip(lid, m) if j==1])\n",
        "\n",
        "                logits = actual_logits\n",
        "                label_ids = actual_label_ids\n",
        "\n",
        "                predictions.append(logits)\n",
        "                true_labels.append(label_ids)\n",
        "\n",
        "            elif F1_METHOD == 'average':\n",
        "                \n",
        "                logits = [list(p) for p in logits]\n",
        "                \n",
        "                actual_logits = []\n",
        "                actual_label_ids = []\n",
        "                \n",
        "                for l,lid,m,b_ii in zip(logits, label_ids, mymasks, b_input_ids):\n",
        "                        \n",
        "                    actual_label_ids.append([i for i,j in zip(lid, m) if j==1])\n",
        "                    my_logits = []\n",
        "                    curr_preds = []\n",
        "                    in_split = 0\n",
        "                    for i,j,k in zip(l,m,b_ii):\n",
        "                        '''if k == 0:\n",
        "                            break'''\n",
        "                        if j==1:\n",
        "                            if in_split == 1:\n",
        "                                if len(my_logits)>0:\n",
        "                                    curr_preds.append(my_logits[-1])\n",
        "                                mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                                if len(my_logits)>0:\n",
        "                                    my_logits[-1] = mode_pred\n",
        "                                else:\n",
        "                                    my_logits.append(mode_pred)\n",
        "                                curr_preds = []\n",
        "                                in_split = 0\n",
        "                            my_logits.append(np.argmax(i))\n",
        "                        if j==0:\n",
        "                            curr_preds.append(i)\n",
        "                            in_split = 1\n",
        "                    if in_split == 1:\n",
        "                        if len(my_logits)>0:\n",
        "                            curr_preds.append(my_logits[-1])\n",
        "                        mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                        if len(my_logits)>0:\n",
        "                            my_logits[-1] = mode_pred\n",
        "                        else:\n",
        "                            my_logits.append(mode_pred)\n",
        "                    actual_logits.append(my_logits)\n",
        "                    \n",
        "                predictions.append(actual_logits)\n",
        "                true_labels.append(actual_label_ids)\n",
        "\n",
        "            tmp_eval_accuracy = flat_accuracy(actual_logits, actual_label_ids)\n",
        "            tmp_eval_scope_accuracy = scope_accuracy(actual_logits, actual_label_ids)\n",
        "            eval_scope_accuracy += tmp_eval_scope_accuracy\n",
        "\n",
        "            eval_loss += tmp_eval_loss.mean().item()\n",
        "            eval_accuracy += tmp_eval_accuracy\n",
        "\n",
        "            nb_eval_examples += len(b_input_ids)\n",
        "            nb_eval_steps += 1\n",
        "        eval_loss = eval_loss/nb_eval_steps\n",
        "        print(\"Validation loss: {}\".format(eval_loss))\n",
        "        print(\"Validation Accuracy: {}\".format(eval_accuracy/nb_eval_steps))\n",
        "        print(\"Validation Accuracy Scope Level: {}\".format(eval_scope_accuracy/nb_eval_steps))\n",
        "        f1_scope([j for i in true_labels for j in i], [j for i in predictions for j in i], level='scope')\n",
        "        labels_flat = [l_ii for l in true_labels for l_i in l for l_ii in l_i]\n",
        "        pred_flat = [p_ii for p in predictions for p_i in p for p_ii in p_i]\n",
        "        classification_dict = classification_report(labels_flat, pred_flat, output_dict= True)\n",
        "        p = classification_dict[\"1\"][\"precision\"]\n",
        "        r = classification_dict[\"1\"][\"recall\"]\n",
        "        f1 = classification_dict[\"1\"][\"f1-score\"]\n",
        "        return_dict['Precision'] = p\n",
        "        return_dict['Recall'] = r\n",
        "        return_dict['F1'] = f1\n",
        "        print(\"Classification Report:\")\n",
        "        print(classification_report(labels_flat, pred_flat))\n",
        "        return return_dict\n",
        "\n",
        "    def predict(self, dataloader):\n",
        "        self.model.eval()\n",
        "        predictions, ip_mask = [], []\n",
        "        for batch in dataloader:\n",
        "            batch = tuple(t.to(self.device) for t in batch)\n",
        "            b_input_ids, b_input_mask, b_mymasks = batch\n",
        "\n",
        "            with torch.no_grad():\n",
        "                logits = self.model(b_input_ids, token_type_ids=None, attention_mask=b_input_mask)[0]\n",
        "            logits = logits.detach().cpu().numpy()\n",
        "            mymasks = b_mymasks.to('cpu').numpy()\n",
        "\n",
        "            if F1_METHOD == 'first_token':\n",
        "\n",
        "                logits = [list(p) for p in np.argmax(logits, axis=2)]\n",
        "                actual_logits = []\n",
        "                for l,lid,m in zip(logits, label_ids, mymasks):\n",
        "                    actual_logits.append([i for i,j in zip(l,m) if j==1])\n",
        "                \n",
        "                logits = actual_logits\n",
        "                label_ids = actual_label_ids\n",
        "\n",
        "                predictions.append(logits)\n",
        "                true_labels.append(label_ids)\n",
        "\n",
        "            elif F1_METHOD == 'average':\n",
        "                \n",
        "                logits = [list(p) for p in logits]\n",
        "                \n",
        "                actual_logits = []\n",
        "                \n",
        "                for l,m in zip(logits, mymasks):\n",
        "                        \n",
        "                    my_logits = []\n",
        "                    curr_preds = []\n",
        "                    in_split = 0\n",
        "                    for i,j in zip(l,m):\n",
        "                        \n",
        "                        if j==1:\n",
        "                            if in_split == 1:\n",
        "                                if len(my_logits)>0:\n",
        "                                    curr_preds.append(my_logits[-1])\n",
        "                                mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                                if len(my_logits)>0:\n",
        "                                    my_logits[-1] = mode_pred\n",
        "                                else:\n",
        "                                    my_logits.append(mode_pred)\n",
        "                                curr_preds = []\n",
        "                                in_split = 0\n",
        "                            my_logits.append(np.argmax(i))\n",
        "                        if j==0:\n",
        "                            curr_preds.append(i)\n",
        "                            in_split = 1\n",
        "                    if in_split == 1:\n",
        "                        if len(my_logits)>0:\n",
        "                            curr_preds.append(my_logits[-1])\n",
        "                        mode_pred = np.argmax(np.average(np.array(curr_preds), axis=0), axis=0)\n",
        "                        if len(my_logits)>0:\n",
        "                            my_logits[-1] = mode_pred\n",
        "                        else:\n",
        "                            my_logits.append(mode_pred)\n",
        "                    actual_logits.append(my_logits)\n",
        "                    \n",
        "                predictions.append(actual_logits)\n",
        "        return predictions\n",
        "\n"
      ],
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "v94vUUL8vqyc"
      },
      "source": [
        "Funtions to process text and extract biomarkers"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "rjctOYdjaACt"
      },
      "source": [
        "# function to extract text from dailyMed in 'INDICATIONS AND USAGE' section with matched disease\n",
        "def diseaseContent(dailyMedURL, disease):\n",
        "  \n",
        "  # parse the drug information page's link\n",
        "  dailyMedRp = requests.get(dailyMedURL)\n",
        "  dailyMedPage = lxml.html.fromstring(dailyMedRp.text)\n",
        "\n",
        "  # locate 'INDICATIONS AND USAGE' section in DailyMed and extract the text content by certain cancer type. \n",
        "  usagePath = \"//li[a[contains(text(),'INDICATIONS AND USAGE')]]/div/div[h2[contains(translate(text(),'ABCDEFGHIJKLMNOPQRSTUVWXYZ','abcdefghijklmnopqrstuvwxyz'),'\" + disease.lower() + \"')]]\"\n",
        "  dailyMedDisContent = dailyMedPage.xpath(usagePath)\n",
        "\n",
        "  dailyMedDisContentStr = \"\"\n",
        "  # extract text content of the matched cancer type\n",
        "  if len(dailyMedDisContent)>0:\n",
        "                                 \n",
        "    for disContect in dailyMedDisContent:\n",
        "      dailyMedDisContentStr = dailyMedDisContentStr + disContect.text_content() + ' '\n",
        "      print(disContect.text_content())\n",
        "                                    \n",
        "    dailyMedDisContentStr = dailyMedDisContentStr.strip()\n",
        "\n",
        "  return dailyMedDisContentStr"
      ],
      "execution_count": 23,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "zceAAtHdIllk"
      },
      "source": [
        "# function to extract brand label and NDC codes from dailyMed drug information page\n",
        "def brandNDCDailyMed(dailyMedURL):\n",
        "  # the input should be an url string\n",
        "\n",
        "  # parse the drug information page's link\n",
        "  dailyMedRp = requests.get(dailyMedURL)\n",
        "  dailyMedPage = lxml.html.fromstring(dailyMedRp.text)\n",
        "\n",
        "  dic = {}\n",
        "\n",
        "  # extract the brand label\n",
        "  therBrName = dailyMedPage.xpath(\"//span[@id='drug-label']/text()\")\n",
        "  if len(therBrName) > 0:\n",
        "    therBrName = therBrName[0]\n",
        "  else: therBrName = \"No brand label found\"\n",
        "  dic['drugLabel'] = therBrName\n",
        "\n",
        "  # extract the NDC codes\n",
        "  NDC = dailyMedPage.xpath(\"//span[@id='item-code-s']/text()\")\n",
        "  NDCCodeList = \" \"\n",
        "  if NDC:\n",
        "    NDCCode = NDC[0]\n",
        "    NDCCodeList = NDCCode.split(',')\n",
        "    for c in range(len(NDCCodeList)):\n",
        "      NDCCodeList[c] = NDCCodeList[c].replace('\\n', '')\n",
        "      NDCCodeList[c] = NDCCodeList[c].replace(' ', '')\n",
        "\n",
        "  # extract the NDC code (view more)\n",
        "  NDCCodeMore = dailyMedPage.xpath(\"//span[@id='item-code-s']//div[@class = 'more-codes']/span/text()\")\n",
        "  if NDCCodeMore:\n",
        "    NDCCodeMore = NDCCodeMore[0]\n",
        "    NDCCodeMoreList = NDCCodeMore.split(',')\n",
        "    for m in range(len(NDCCodeMoreList)):\n",
        "      NDCCodeMoreList[m] = NDCCodeMoreList[m].replace('\\n', '')\n",
        "      NDCCodeMoreList[m] = NDCCodeMoreList[m].replace(' ', '')\n",
        "    \n",
        "    NDCCodeList.extend(NDCCodeMoreList)\n",
        "  dic['NDCCode'] = [x for x in NDCCodeList if x]\n",
        "\n",
        "  return dic\n"
      ],
      "execution_count": 24,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tITbsh1GmNOo"
      },
      "source": [
        "def negScope(text):\n",
        "  # the input text: should be a single sentence\n",
        "\n",
        "  # perform negation cue detection\n",
        "  mydata = CustomData([text])\n",
        "  dl = mydata.get_cue_dataloader()\n",
        "  cueIndex = modelCue.predict(dl)\n",
        "\n",
        "  # string of the scope\n",
        "  scopeStr = []\n",
        "\n",
        "  # check if negation is detected.\n",
        "  if 1 in cueIndex[0][0]:\n",
        "    \n",
        "    # perform negation scope resolution\n",
        "    mydata = CustomData([text], cues = cueIndex[0])\n",
        "    dl = mydata.get_scope_dataloader()\n",
        "    scopeIndex = modelScope.predict(dl)\n",
        "\n",
        "    # extract the negated string of the scope\n",
        "    if len(scopeIndex[0][0]) == len(text.split()):\n",
        "      scopeSubstr = ''\n",
        "      for i in range(len(text.split())):\n",
        "        if scopeIndex[0][0][i] == 1 and i != len(text.split())-1:\n",
        "          scopeSubstr += \" \" + text.split()[i]\n",
        "        elif i == len(text.split())-1 and scopeIndex[0][0][i] == 1:\n",
        "          scopeSubstr += \" \" + text.split()[i]\n",
        "          scopeStr.append(scopeSubstr.lstrip())\n",
        "        else: \n",
        "          if len(scopeSubstr) > 0:\n",
        "            scopeStr.append(scopeSubstr.lstrip())\n",
        "            scopeSubstr = ''\n",
        "  \n",
        "  #return scopeStr\n",
        "  geneProDrDic = {}\n",
        "  if len(scopeStr)>0:\n",
        "    scopeStrConca = ' '.join(st for st in scopeStr)\n",
        "    geneProDrDic = geneProteinDrugExtract(text=scopeStrConca, gene= 'gene', protein = 'protein', drug = 'drug')\n",
        "\n",
        "  return geneProDrDic\n"
      ],
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "qOWV_3UVmQ-Q"
      },
      "source": [
        "# function to extract gene, protein, and drug label from a string\n",
        "def geneProteinDrugExtract(text, gene='1', protein = '1', drug = '1'):\n",
        "  # options: gene= 'gene', protein = 'protein', drug = 'drug'\n",
        "  \n",
        "  dic = {}\n",
        "  geneList = []\n",
        "  proteinList = []\n",
        "  drugList = []\n",
        "\n",
        "  # extract genes\n",
        "  if gene=='gene':\n",
        "    doc = nlp_craft(text)\n",
        "    # filter gene and gene product\n",
        "    for entity in doc.ents:\n",
        "      if entity.label_=='GGP':\n",
        "         #print(entity.text + \" GGP\")\n",
        "         geneList.append(entity.text)\n",
        "    dic['gene'] = geneList\n",
        "\n",
        "  # extract proteins\n",
        "  if protein == 'protein':\n",
        "    doc = nlp_jnlpha(text)\n",
        "    # filter protein\n",
        "    for entity in doc.ents:\n",
        "      if entity.label_=='PROTEIN':\n",
        "        #print(entity.text + \" PROTEIN\")\n",
        "        proteinList.append(entity.text)\n",
        "    dic['protein'] = proteinList\n",
        "\n",
        "  # extract drugs (SIMPLE_CHEMICAL)\n",
        "  if drug == 'drug':\n",
        "    doc = nlp_bionlp13cg(text)\n",
        "    # filter drug (SIMPLE_CHEMICAL)\n",
        "    for entity in doc.ents:\n",
        "      if entity.label_=='SIMPLE_CHEMICAL':\n",
        "        #print(entity.text + \" SIMPLE_CHEMICAL\")\n",
        "        drugList.append(entity.text)\n",
        "    dic['drug'] = drugList\n",
        "\n",
        "  return dic\n"
      ],
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4pdUtNsymmkT"
      },
      "source": [
        "# rule to extract CHEMICAL name in certain patterns ('in conbination with', 'used with', etc)\n",
        "def sent_subtree(text):\n",
        "    \n",
        "    # pattern match\n",
        "    patterns = [r'\\b(?i)'+'in combination with'+r'\\b',\n",
        "           r'\\b(?i)'+'used with'+r'\\b']\n",
        "    \n",
        "    subPatterns = ['combination', 'used']\n",
        "    \n",
        "    schemes = []\n",
        "    doc = nlp_bionlp13cg(text)\n",
        "    flag = 0\n",
        "    # if no pattern present in sentence\n",
        "    for pat in patterns:\n",
        "        \n",
        "        if re.search(pat, text) != None:\n",
        "            flag = 1\n",
        "            break\n",
        "    \n",
        "    if flag == 0:\n",
        "       return schemes\n",
        "\n",
        "    # iterating over sentence tokens\n",
        "    for token in doc:\n",
        "\n",
        "        for pat in subPatterns:\n",
        "                \n",
        "            # if we get a pattern match\n",
        "            if re.search(pat, token.text) != None:\n",
        "\n",
        "                subtree = list(token.subtree)\n",
        "                subtreeSent = ' '.join(word.text for word in subtree)\n",
        "                schemes.append(subtreeSent)\n",
        "    return schemes      \n"
      ],
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Zb9XR6fyP0v6"
      },
      "source": [
        "# get the url and parse the link\n",
        "url = \"https://www.cancer.gov/about-cancer/treatment/types/targeted-therapies/targeted-therapies-fact-sheet\"\n",
        "rp = requests.get(url)\n",
        "page = lxml.html.fromstring(rp.text)"
      ],
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "goDmgBaLP0v6"
      },
      "source": [
        "# get the link of FDA approaved targeted therapies\n",
        "therapies = page.xpath(\"//section[h2[@id = 'what-targeted-therapies-have-been-approved-for-specific-types-of-cancer']]//p//a/@href\")\n",
        "#therapies"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WUaQSJdQv6Q6"
      },
      "source": [
        "Processing program"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dBkY4KhrP0v7",
        "outputId": "7c838b09-0e54-4c50-bc32-7ce15b5bc672"
      },
      "source": [
        "# create empty lists to contain values\n",
        "therapy_list = []\n",
        "web_gov = 'https://www.cancer.gov'\n",
        "geneProtein_label = []\n",
        "drug_label = []\n",
        "NDCCode_label = []\n",
        "disease_label = []\n",
        "dailymed_label = []\n",
        "comDrug_label = []\n",
        "negGeProDr_label = []\n",
        "\n",
        "for i in therapies:\n",
        "    link = \"\"\n",
        "    \n",
        "    # make sure the url is complete\n",
        "    if web_gov in i:\n",
        "        link = i\n",
        "    else:\n",
        "        link = web_gov + i\n",
        "    \n",
        "    # test if the url is valid\n",
        "    # exist\n",
        "    result = urlparse(link)\n",
        "    if all([result.scheme, result.netloc, result.path]):\n",
        "        therName = \"\"\n",
        "        therBrName = \"\"\n",
        "        \n",
        "        # get the targeted therapy url and parse it\n",
        "        #print(link)\n",
        "        therRp = requests.get(link)\n",
        "        therPage = lxml.html.fromstring(therRp.text)\n",
        "        \n",
        "        # extract the therapy's name\n",
        "        therName = therPage.xpath(\"//article/div/h1/text()\")\n",
        "        print(therName)\n",
        "\n",
        "        if len(therName) > 0:\n",
        "            print(therName[0])\n",
        "            # check duplicate therapy\n",
        "            if therName[0] not in therapy_list:\n",
        "                therapy_list.append(therName[0])\n",
        "                \n",
        "                # extract the corresponding diseases (only for bold text, but there are some non-bold disease's name left (could be improved))\n",
        "                diseases = therPage.xpath(\"//article//div[h2[contains(text(),'Use in Cancer')]]/ul/li/strong\")\n",
        "                \n",
        "                punc = '''!()-[]{};:'\"\\, <>./?@#$%^&*_~'''\n",
        "                diseaseList = []\n",
        "                dis = ''\n",
        "                for disease in diseases:\n",
        "                    # get the disease in string\n",
        "                    # remove punctuation at the end of disease\n",
        "                    dis = str(disease.text_content())\n",
        "                    if dis[-1] in punc:  \n",
        "                        dis = dis.replace(dis[-1], \"\")  \n",
        "                        #print(dis)\n",
        "                        # add the diseases from NCI page to the list\n",
        "                        diseaseList.append(dis)\n",
        "                        \n",
        "                    '''\n",
        "                    ## on NCI page\n",
        "                    # obtain the text associated with the disease\n",
        "                    path = \"//article//div[h2[contains(text(),'Use in Cancer')]]/ul/li[strong/*[contains(text(),'\" + dis + \"')]]\"\n",
        "                    diseaseContent = therPage.xpath(path)\n",
        "                    #print(len(diseaseContent))\n",
        "                    \n",
        "\n",
        "                    # do NER to extract GENE and PROTEIN\n",
        "                    geneProtein = ''\n",
        "                    if len(diseaseContent)>0:\n",
        "                        #print(diseaseContent[0].text_content())\n",
        "                            \n",
        "                        text = diseaseContent[0].text_content()\n",
        "                        doc = nlp_craft(str(text))\n",
        "                        # filter GGP (gene and gene product) type\n",
        "                        for entity in doc.ents:\n",
        "                            if entity.label_=='GGP':\n",
        "                                geneProtein = geneProtein + ', ' + entity.text\n",
        "                                print(geneProtein + ' GGP')\n",
        "\n",
        "                        doc = nlp_jnlpha(str(text))\n",
        "                        # filter PROTEIN type\n",
        "                        for entity in doc.ents:\n",
        "                            if entity.label_=='PROTEIN':\n",
        "                                geneProtein = geneProtein + ', ' + entity.text\n",
        "                                print(geneProtein + ' PROTEIN')\n",
        "                        \n",
        "                    # add the gene and proteins for the associated disease\n",
        "                    #geneProtein_label.append(geneProtein[2:])\n",
        "                        \n",
        "                    # add the disease and drug name in NCI to the disease/drug label list\n",
        "                    #disease_label.append(dis)\n",
        "                    #drug_label.append(therName[0]+' (general)')  \n",
        "                    '''\n",
        "                    \n",
        "                    ## on DailyMed pages\n",
        "                    # obtain DailyMed link\n",
        "                    dailyMed = therPage.xpath(\"//article//div/p/a[text() = 'FDA label information for this drug is available at DailyMed.']/@href\")\n",
        "                    \n",
        "                    geneProtein = []\n",
        "                    combDrug = []\n",
        "                    negGeProDrList = []\n",
        "                    if len(dailyMed)>0:\n",
        "                    \n",
        "                        dailyMedURL = dailyMed[0]\n",
        "                \n",
        "                        # make sure the dailymed url is valid\n",
        "                        dailyMedResult = urlparse(dailyMedURL)\n",
        "                        if all([dailyMedResult.scheme, dailyMedResult.netloc, dailyMedResult.path]):\n",
        "                        \n",
        "                            # check if the link is directed to drug info page or multiple drug search page\n",
        "                            dailyMedRp = requests.get(dailyMedURL)\n",
        "                            dailyMedPage = lxml.html.fromstring(dailyMedRp.text)\n",
        "\n",
        "                            # check if the URL link is of the multiple drug selection page\n",
        "                            if 'dailymed/search' in dailyMedURL:\n",
        "                                print('multiple choice')\n",
        "                                print(dailyMedURL)\n",
        "                                \n",
        "                                # multiple drug search page, get each drug link\n",
        "                                searchResults = dailyMedPage.xpath(\"//div[@class = 'results-info']\")\n",
        "                                for s in range(1,len(searchResults)+1):\n",
        "                                  dailyMedDrugURL = dailyMedPage.xpath(\"(//div[@class = 'results-info'])[\" + str(s) + \"]//a[@class='drug-info-link']/@href\")\n",
        "                                  #print(len(dailyMedDrugURL))\n",
        "                                  \n",
        "                                  # extract the url\n",
        "                                  if len(dailyMedDrugURL) > 0:\n",
        "                                      dailyMedDrugURL = dailyMedDrugURL[0]\n",
        "      \n",
        "                                      # make sure it is a link to drug information page\n",
        "                                      if 'dailymed/drugInfo' in dailyMedDrugURL:\n",
        "                                          dailyMedURL = 'https://dailymed.nlm.nih.gov' + dailyMedDrugURL\n",
        "                                          print(dailyMedURL)\n",
        "\n",
        "                                          # extract the brand name\n",
        "                                          therBrName = brandNDCDailyMed(dailyMedURL = dailyMedURL).get('drugLabel')\n",
        "\n",
        "                                          # extract the NDC code\n",
        "                                          NDCCodes = brandNDCDailyMed(dailyMedURL = dailyMedURL).get('NDCCode')\n",
        "\n",
        "                                          # locate 'INDICATIONS AND USAGE' section in DailyMed and extract the text content by certain cancer type. \n",
        "                                          dailyMedDisContentStr = diseaseContext(dailyMedURL = dailyMedURL, disease = dis)\n",
        "\n",
        "                                          if dailyMedDisContentStr:\n",
        "                                              ## do NER\n",
        "                                              geneProteinDic = geneProteinDrugExtract(text = dailyMedDisContentStr, gene='gene', protein = 'protein')\n",
        "                                              \n",
        "                                              # extract gene and protein\n",
        "                                              geneProtein.extend(geneProteinDic.get(\"gene\"))\n",
        "                                              geneProtein.extend(geneProteinDic.get(\"protein\"))\n",
        "\n",
        "                                              # split into sentences\n",
        "                                              dailyMedDisContentStrList = tokenize.sent_tokenize(dailyMedDisContentStr)\n",
        "\n",
        "                                              for st in dailyMedDisContentStrList:\n",
        "\n",
        "                                                ## do 'in conbination with'\n",
        "                                                subTree = sent_subtree(st)\n",
        "                                                if len(subTree) >0:\n",
        "                                                  subTreeConca = ' '.join(tree for tree in subTree)\n",
        "                                                  drug = geneProteinDrugExtract(text = subTreeConca, drug = 'drug')\n",
        "                                                  combDrug.extend(drug.get(\"drug\"))\n",
        "                                                \n",
        "                                                ## do negation\n",
        "                                                negGeProDrDic = negScope(st)\n",
        "                                                if len(negGeProDrDic)>0:\n",
        "                                                  print(negGeProDrDic)\n",
        "                                                  negGeProDrList.extend(negGeProDrDic.get(\"drug\"))\n",
        "                                                  negGeProDrList.extend(negGeProDrDic.get(\"gene\"))\n",
        "                                                  negGeProDrList.extend(negGeProDrDic.get(\"protein\"))\n",
        "\n",
        "                                              # add the gene and proteins for the associated disease\n",
        "                                              geneProteinStr = \"\"\n",
        "                                              if len(geneProtein)>0:\n",
        "                                                for i in geneProtein:\n",
        "                                                  geneProteinStr += \", \" + i\n",
        "                                                geneProteinStr = geneProteinStr[2:]\n",
        "                                              geneProtein_label.append(geneProteinStr)\n",
        "\n",
        "                                              # add the disease and drug name in NCI to the disease/drug label list\n",
        "                                              disease_label.append(dis)\n",
        "                                              drug_label.append(therName[0]+' (' + therBrName + ')')  \n",
        "\n",
        "                                              # add the NDC codes to the list\n",
        "                                              NDCCodesStr = \"\"\n",
        "                                              if len(NDCCodes)>0:\n",
        "                                                for i in NDCCodes:\n",
        "                                                  NDCCodesStr += \", \" + i\n",
        "                                                NDCCodesStr = NDCCodesStr[2:]\n",
        "                                              NDCCode_label.append(NDCCodesStr)\n",
        "                                              \n",
        "                                              # add the combination used drug\n",
        "                                              combDrugStr = \"\"\n",
        "                                              if len(combDrug)>0:\n",
        "                                                for i in combDrug:\n",
        "                                                  combDrugStr += \", \" + i\n",
        "                                                combDrugStr = combDrugStr[2:]\n",
        "                                              comDrug_label.append(combDrugStr)\n",
        "                                              \n",
        "                                              # add negated drug, gene, protein\n",
        "                                              negGeProDrListStr = \"\"\n",
        "                                              if len(negGeProDrList)>0:\n",
        "                                                for i in negGeProDrList:\n",
        "                                                  negGeProDrListStr += \", \" + i\n",
        "                                                negGeProDrListStr = negGeProDrListStr[2:]\n",
        "                                              negGeProDr_label.append(negGeProDrListStr)\n",
        "\n",
        "                            \n",
        "                            # drug information page\n",
        "                            elif 'dailymed/drugInfo' in dailyMedURL:\n",
        "                                print(dailyMedURL)\n",
        "                                \n",
        "                                # extract the brand name\n",
        "                                therBrName = brandNDCDailyMed(dailyMedURL = dailyMedURL).get('drugLabel')\n",
        "\n",
        "                                # extract the NDC code\n",
        "                                NDCCodes = brandNDCDailyMed(dailyMedURL = dailyMedURL).get('NDCCode')\n",
        "\n",
        "                                '''\n",
        "                                therBrName = dailyMedPage.xpath(\"//span[@id='drug-label']/text()\")\n",
        "                                if len(therBrName) > 0:\n",
        "                                    therBrName = therBrName[0]\n",
        "                                else: therBrName = \"No brand found\"\n",
        "                                '''\n",
        "                                \n",
        "                                # locate 'INDICATIONS AND USAGE' section in DailyMed and extract the text content by certain cancer type. \n",
        "                                dailyMedDisContentStr = diseaseContext(dailyMedURL = dailyMedURL, disease = dis)\n",
        "\n",
        "                                '''\n",
        "                                usagePath = \"//li[a[contains(text(),'INDICATIONS AND USAGE')]]/div/div[h2[contains(translate(text(),'ABCDEFGHIJKLMNOPQRSTUVWXYZ','abcdefghijklmnopqrstuvwxyz'),'\" + dis.lower() + \"')]]/p\"\n",
        "                                dailyMedDisContent = dailyMedPage.xpath(usagePath)\n",
        "                                print(len(dailyMedDisContent))\n",
        "                                \n",
        "                                # extract text content of the matched cancer type\n",
        "                                if len(dailyMedDisContent)>0:\n",
        "                                    dailyMedDisContentStr = \"\"\n",
        "                                    \n",
        "                                    for disContect in dailyMedDisContent:\n",
        "                                        dailyMedDisContentStr = dailyMedDisContentStr + disContect.text_content() + ' '\n",
        "                                        print(disContect.text_content())\n",
        "                                    \n",
        "                                    dailyMedDisContentStr = dailyMedDisContentStr.strip()\n",
        "                                    print(dailyMedDisContentStr)\n",
        "                                    '''\n",
        "                                if dailyMedDisContentStr:\n",
        "                                    ## do NER\n",
        "                                    geneProteinDic = geneProteinDrugExtract(text = dailyMedDisContentStr, gene='gene', protein = 'protein')\n",
        "                                    \n",
        "                                    # extract gene and protein\n",
        "                                    geneProtein.extend(geneProteinDic.get(\"gene\"))\n",
        "                                    geneProtein.extend(geneProteinDic.get(\"protein\"))\n",
        "\n",
        "                                    # split into sentences\n",
        "                                    dailyMedDisContentStrList = tokenize.sent_tokenize(dailyMedDisContentStr)\n",
        "\n",
        "                                    for st in dailyMedDisContentStrList:\n",
        "\n",
        "                                      ## do 'in conbination with'\n",
        "                                      subTree = sent_subtree(st)\n",
        "                                      if len(subTree) >0:\n",
        "                                        subTreeConca = ' '.join(tree for tree in subTree)\n",
        "                                        drug = geneProteinDrugExtract(text = subTreeConca, drug = 'drug')\n",
        "                                        combDrug.extend(drug.get(\"drug\"))\n",
        "                                      \n",
        "                                      ## do negation\n",
        "                                      negGeProDrDic = negScope(st)\n",
        "                                      if len(negGeProDrDic)>0:\n",
        "                                        print(negGeProDrDic)\n",
        "                                        negGeProDrList.extend(negGeProDrDic.get(\"drug\"))\n",
        "                                        negGeProDrList.extend(negGeProDrDic.get(\"gene\"))\n",
        "                                        negGeProDrList.extend(negGeProDrDic.get(\"protein\"))\n",
        "\n",
        "                                    # add the gene and proteins for the associated disease\n",
        "                                    geneProteinStr = \"\"\n",
        "                                    if len(geneProtein)>0:\n",
        "                                      for i in geneProtein:\n",
        "                                        geneProteinStr += \", \" + i\n",
        "                                      geneProteinStr = geneProteinStr[2:]\n",
        "                                    geneProtein_label.append(geneProteinStr)\n",
        "\n",
        "                                    # add the disease and drug name in NCI to the disease/drug label list\n",
        "                                    disease_label.append(dis)\n",
        "                                    drug_label.append(therName[0]+' (' + therBrName + ')')  \n",
        "\n",
        "                                    # add the NDC codes to the list\n",
        "                                    NDCCodesStr = \"\"\n",
        "                                    if len(NDCCodes)>0:\n",
        "                                      for i in NDCCodes:\n",
        "                                        NDCCodesStr += \", \" + i\n",
        "                                      NDCCodesStr = NDCCodesStr[2:]\n",
        "                                    NDCCode_label.append(NDCCodesStr)\n",
        "                                    \n",
        "                                    # add the combination used drug\n",
        "                                    combDrugStr = \"\"\n",
        "                                    if len(combDrug)>0:\n",
        "                                      for i in combDrug:\n",
        "                                        combDrugStr += \", \" + i\n",
        "                                      combDrugStr = combDrugStr[2:]\n",
        "                                    comDrug_label.append(combDrugStr)\n",
        "                                    \n",
        "                                    # add negated drug, gene, protein\n",
        "                                    negGeProDrListStr = \"\"\n",
        "                                    if len(negGeProDrList)>0:\n",
        "                                      for i in negGeProDrList:\n",
        "                                        negGeProDrListStr += \", \" + i\n",
        "                                      negGeProDrListStr = negGeProDrListStr[2:]\n",
        "                                    negGeProDr_label.append(negGeProDrListStr)\n",
        "\n",
        "\n",
        "\n",
        "                \n",
        "                \n",
        "    \n",
        "            \n",
        "    "
      ],
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "['Atezolizumab']\n",
            "Atezolizumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6fa682c9-a312-4932-9831-f286908660ee&audience=consumer\n",
            "\n",
            "TECENTRIQ, in combination with paclitaxel protein-bound, is indicated for the treatment of adult patients with unresectable locally advanced or metastatic triple-negative breast cancer (TNBC) whose tumors express PD-L1 (PD-L1 stained tumor-infiltrating immune cells [IC] of any intensity covering  1% of the tumor area), as determined by an FDA-approved test [see Dosage and Administration (2.1)].\n",
            "This indication is approved under accelerated approval based on progression free survival [see Clinical Studies (14.3)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in a confirmatory trial(s).\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2605: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n",
            "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2619: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6fa682c9-a312-4932-9831-f286908660ee&audience=consumer\n",
            "\n",
            "\n",
            "TECENTRIQ, in combination with bevacizumab, is indicated for the treatment of patients with unresectable or metastatic hepatocellular carcinoma (HCC) who have not received prior systemic therapy.\n",
            "\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2989: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n",
            "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3003: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6fa682c9-a312-4932-9831-f286908660ee&audience=consumer\n",
            "\n",
            "\n",
            "TECENTRIQ, in combination with cobimetinib and vemurafenib, is indicated for the treatment of patients with BRAF V600 mutation-positive unresectable or metastatic melanoma [see Dosage and Administration (2.1)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6fa682c9-a312-4932-9831-f286908660ee&audience=consumer\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6fa682c9-a312-4932-9831-f286908660ee&audience=consumer\n",
            "\n",
            "\n",
            "TECENTRIQ, in combination with carboplatin and etoposide, is indicated for the first-line treatment of adult patients with extensive-stage small cell lung cancer (ES-SCLC).\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6fa682c9-a312-4932-9831-f286908660ee&audience=consumer\n",
            "\n",
            "TECENTRIQ is indicated for the treatment of adult patients with locally advanced or metastatic urothelial carcinoma who:\n",
            "This indication is approved under accelerated approval based on tumor response rate and durability of response [see Clinical Studies (14.1)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in a confirmatory trial(s).\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "\n",
            "OPDIVO, as a single agent or in combination with ipilimumab, is indicated for the treatment of adult and pediatric patients 12 years and older with microsatellite instability-high (MSI-H) or mismatch repair deficient (dMMR) metastatic colorectal cancer (CRC) that has progressed following treatment with a fluoropyrimidine, oxaliplatin, and irinotecan.\n",
            "This indication is approved under accelerated approval based on overall response rate and duration of response [see Clinical Studies (14.9)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in confirmatory trials.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "\n",
            "OPDIVO, as a single agent or in combination with ipilimumab, is indicated for the treatment of patients with hepatocellular carcinoma (HCC) who have been previously treated with sorafenib. This indication is approved under accelerated approval based on overall response rate and duration of response [see Clinical Studies (14.10)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in the confirmatory trials.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "\n",
            "\n",
            "OPDIVO, in combination with ipilimumab, is indicated for the first-line treatment of adult patients with unresectable malignant pleural mesothelioma.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "\n",
            "OPDIVO, as a single agent or in combination with ipilimumab, is indicated for the treatment of patients with unresectable or metastatic melanoma.\n",
            "\n",
            "OPDIVO is indicated for the adjuvant treatment of patients with melanoma with involvement of lymph nodes or metastatic disease who have undergone complete resection.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "\n",
            "OPDIVO is indicated for the treatment of patients with recurrent or metastatic squamous cell carcinoma of the head and neck (SCCHN) with disease progression on or after platinum-based therapy.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f570b9c4-6846-4de2-abfa-4d0a4ae4e394&audience=consumer\n",
            "\n",
            "OPDIVO is indicated for the treatment of patients with locally advanced or metastatic urothelial carcinoma who:\n",
            "This indication is approved under accelerated approval based on tumor response rate and duration of response [see Clinical Studies (14.8)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in confirmatory trials. \n",
            "['Avelumab']\n",
            "Avelumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5cd725a1-2fa4-408a-a651-57a7b84b2118&audience=consumer\n",
            "\n",
            "BAVENCIO (avelumab) is indicated for the treatment of adults and pediatric patients 12 years and older with metastatic Merkel cell carcinoma (MCC). This indication is approved under accelerated approval based on tumor response rate and duration of response. Continued approval for this indication may be contingent upon verification and description of clinical benefit in confirmatory trials [see Clinical Studies (14.1)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5cd725a1-2fa4-408a-a651-57a7b84b2118&audience=consumer\n",
            "\n",
            "BAVENCIO in combination with axitinib is indicated for the first-line treatment of patients with advanced renal cell carcinoma (RCC) [see Clinical Studies (14.3)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5cd725a1-2fa4-408a-a651-57a7b84b2118&audience=consumer\n",
            "\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "KEYTRUDA is indicated for the treatment of patients with recurrent or metastatic cervical cancer with disease progression on or after chemotherapy whose tumors express PD-L1 (CPS 1) as determined by an FDA-approved test [see Dosage and Administration (2.1)].\n",
            "This indication is approved under accelerated approval based on tumor response rate and durability of response [see Clinical Studies (14.11)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in the confirmatory trials.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "KEYTRUDA, in combination with lenvatinib, is indicated for the treatment of patients with advanced endometrial carcinoma that is not MSI-H or dMMR, who have disease progression following prior systemic therapy and are not candidates for curative surgery or radiation.\n",
            "This indication is approved under accelerated approval based on tumor response rate and durability of response [see Clinical Studies (14.15)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in the confirmatory trials.\n",
            "{'gene': ['MSI-H'], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "\t\t\t\t\t\t\t\t\tKEYTRUDA is indicated for the treatment of patients with hepatocellular carcinoma (HCC) who have been previously treated with sorafenib.\n",
            "\n",
            "\t\t\t\t\t\t\t\t\tThis indication is approved under accelerated approval based on tumor response rate and durability of response [see Clinical Studies (14.12)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in the confirmatory trials. \t\t\t\t\t\t\t\t\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "KEYTRUDA is indicated for the treatment of patients with unresectable or metastatic melanoma.\n",
            "KEYTRUDA is indicated for the adjuvant treatment of patients with melanoma with involvement of lymph node(s) following complete resection.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "\t\t\t\t\t\t\t\t\tKEYTRUDA is indicated for the treatment of adult and pediatric patients with recurrent locally advanced or metastatic Merkel cell carcinoma (MCC). \t\t\t\t\t\t\t\t\n",
            "\t\t\t\t\t\t\t\t\tThis indication is approved under accelerated approval based on tumor response rate and durability of response [see Clinical Studies (14.13)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in the confirmatory trials. \t\t\t\t\t\t\t\t\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "\t\t\t\t\t\t\t\t\tKEYTRUDA, in combination with pemetrexed and platinum chemotherapy, is indicated for the first-line treatment of patients with metastatic nonsquamous non-small cell lung cancer (NSCLC), with no EGFR or ALK genomic tumor aberrations. \t\t\t\t\t\t\t\t\n",
            "\t\t\t\t\t\t\t\t\tKEYTRUDA, in combination with carboplatin and either paclitaxel or paclitaxel protein-bound, is indicated for the first-line treatment of patients with metastatic squamous NSCLC. \t\t\t\t\t\t\t\t\n",
            "KEYTRUDA, as a single agent, is indicated for the first-line treatment of patients with NSCLC expressing PD-L1 [Tumor Proportion Score (TPS) 1%] as determined by an FDA-approved test [see Dosage and Administration (2.1)], with no EGFR or ALK genomic tumor aberrations, and is:\n",
            "\t\t\t\t\t\t\t\t\n",
            "KEYTRUDA, as a single agent, is indicated for the treatment of patients with metastatic NSCLC whose tumors express PD-L1 (TPS 1%) as determined by an FDA-approved test [see Dosage and Administration (2.1)], with disease progression on or after platinum-containing chemotherapy. Patients with EGFR or ALK genomic tumor aberrations should have disease progression on FDA-approved therapy for these aberrations prior to receiving KEYTRUDA.\n",
            "{'gene': ['EGFR', 'ALK genomic'], 'protein': ['EGFR'], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "\n",
            "KEYTRUDA is indicated for the treatment of adult and pediatric patients with refractory primary mediastinal large B-cell lymphoma (PMBCL), or who have relapsed after 2 or more prior lines of therapy.\n",
            "\n",
            "\n",
            "Limitations of Use: KEYTRUDA is not recommended for treatment of patients with PMBCL who require urgent cytoreductive therapy. \n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "KEYTRUDA, in combination with axitinib, is indicated for the first-line treatment of patients with advanced renal cell carcinoma (RCC).\n",
            "\t\t\t\t\t\t\t\t\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9333c79b-d487-4538-a9f0-71b91a02b287&audience=consumer\n",
            "\n",
            "\t\t\t\t\t\t\t\t\tKEYTRUDA is indicated for the treatment of patients with locally advanced or metastatic urothelial carcinoma who are not eligible for cisplatin-containing chemotherapy and whose tumors express PD-L1 (CPS 10) as determined by an FDA-approved test [see Dosage and Administration (2.1)], or in patients who are not eligible for any platinum-containing chemotherapy regardless of PD-L1 status. \t\t\t\t\t\t\t\t\t \t\t\t\t\t\t\t\t\n",
            "This indication is approved under accelerated approval based on tumor response rate and duration of response [see Clinical Studies (14.6)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in confirmatory trials.\n",
            "KEYTRUDA is indicated for the treatment of patients with locally advanced or metastatic urothelial carcinoma who have disease progression during or following platinum-containing chemotherapy or within 12 months of neoadjuvant or adjuvant treatment with platinum-containing chemotherapy. \t\t\t\t\t\t\t\t\n",
            "KEYTRUDA is indicated for the treatment of patients with Bacillus Calmette-Guerin (BCG)-unresponsive, high-risk, non-muscle invasive bladder cancer (NMIBC) with carcinoma in situ (CIS) with or without papillary tumors who are ineligible for or have elected not to undergo cystectomy. \n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "['Erdafitinib']\n",
            "Erdafitinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2a8aa5c0-6c92-4566-8c45-e8f4d1fc20ee&audience=consumer\n",
            "['Enfortumab Vedotin-ejfv']\n",
            "Enfortumab Vedotin-ejfv\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b5631d3e-4604-4363-8f20-11dfc5a4a8ed&audience=consumer\n",
            "['Bevacizumab']\n",
            "Bevacizumab\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "\n",
            "Avastin, in combination with paclitaxel and cisplatin or paclitaxel and topotecan, is indicated for the treatment of patients with persistent, recurrent, or metastatic cervical cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "\n",
            "MVASI, in combination with paclitaxel and cisplatin or paclitaxel and topotecan, is indicated for the treatment of patients with persistent, recurrent, or metastatic cervical cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "\n",
            "ZIRABEV, in combination with paclitaxel and cisplatin or paclitaxel and topotecan, is indicated for the treatment of patients with persistent, recurrent, or metastatic cervical cancer.\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "\n",
            "Avastin, in combination with intravenous fluorouracil-based chemotherapy, is indicated for the first-or second-line treatment of patients with metastatic colorectal cancer (mCRC).\n",
            "Avastin, in combination with fluoropyrimidine-irinotecan- or fluoropyrimidine-oxaliplatin-based chemotherapy, is indicated for the second-line treatment of patients with mCRC who have progressed on a first-line Avastin-containing regimen.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "\n",
            "MVASI, in combination with intravenous fluorouracil-based chemotherapy, is indicated for the first-or second-line treatment of patients with metastatic colorectal cancer (mCRC).\n",
            "MVASI, in combination with fluoropyrimidine-irinotecan-or fluoropyrimidine-oxaliplatin-based chemotherapy, is indicated for the second-line treatment of patients with mCRC who have progressed on a first-line bevacizumab product-containing regimen.\n",
            "\n",
            "Limitations of Use: MVASI is not indicated for adjuvant treatment of colon cancer[see Clinical Studies (14.2)].\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': ['MVASI']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "\n",
            "ZIRABEV, in combination with intravenous fluorouracil-based chemotherapy, is indicated for the first- or second-line treatment of patients with metastatic colorectal cancer (mCRC).\n",
            "ZIRABEV, in combination with fluoropyrimidine-irinotecan- or fluoropyrimidine-oxaliplatin-based chemotherapy, is indicated for the second-line treatment of patients with mCRC who have progressed on a first-line bevacizumab product-containing regimen.\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "\n",
            "Avastin is indicated for the treatment of recurrent glioblastoma (GBM) in adults.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "\n",
            "\n",
            "MVASI is indicated for the treatment of recurrent glioblastoma (GBM) in adults.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "\n",
            "ZIRABEV is indicated for the treatment of recurrent glioblastoma (GBM) in adults.\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "\n",
            "\n",
            "Avastin, in combination with atezolizumab, is indicated for the treatment of patients with unresectable or metastatic hepatocellular carcinoma (HCC) who have not received prior systemic therapy.\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "\n",
            "Avastin, in combination with carboplatin and paclitaxel, followed by Avastin as a single agent, is indicated for the treatment of patients with stage III or IV epithelial ovarian, fallopian tube, or primary peritoneal cancer following initial surgical resection. \t\t\t\t\t\t\t\t\n",
            "Avastin, in combination with paclitaxel, pegylated liposomal doxorubicin, or topotecan, is indicated for the treatment of patients with platinum-resistant recurrent epithelial ovarian, fallopian tube or primary peritoneal cancer who received no more than 2 prior chemotherapy regimens.\n",
            "Avastin, in combination with carboplatin and paclitaxel, or with carboplatin and gemcitabine, followed by Avastin as a single agent, is indicated for the treatment of patients with platinum-sensitive recurrent epithelial ovarian, fallopian tube, or primary peritoneal cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "\n",
            "\n",
            "ZIRABEV, in combination with carboplatin and paclitaxel, followed by ZIRABEV as a single agent, is indicated for the treatment of patients with stage III or IV epithelial ovarian, fallopian tube, or primary peritoneal cancer following initial surgical resection.\n",
            "\n",
            "\n",
            "ZIRABEV, in combination with paclitaxel, pegylated liposomal doxorubicin, or topotecan, is indicated for the treatment of patients with platinum-resistant recurrent epithelial ovarian, fallopian tube, or primary peritoneal cancer who received no more than 2 prior chemotherapy regimens.\n",
            "\n",
            "\n",
            "ZIRABEV, in combination with carboplatin and paclitaxel, or with carboplatin and gemcitabine, followed by ZIRABEV as a single agent, is indicated for the treatment of patients with platinum-sensitive recurrent epithelial ovarian, fallopian tube, or primary peritoneal cancer.\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "\n",
            "Avastin, in combination with carboplatin and paclitaxel, followed by Avastin as a single agent, is indicated for the treatment of patients with stage III or IV epithelial ovarian, fallopian tube, or primary peritoneal cancer following initial surgical resection. \t\t\t\t\t\t\t\t\n",
            "Avastin, in combination with paclitaxel, pegylated liposomal doxorubicin, or topotecan, is indicated for the treatment of patients with platinum-resistant recurrent epithelial ovarian, fallopian tube or primary peritoneal cancer who received no more than 2 prior chemotherapy regimens.\n",
            "Avastin, in combination with carboplatin and paclitaxel, or with carboplatin and gemcitabine, followed by Avastin as a single agent, is indicated for the treatment of patients with platinum-sensitive recurrent epithelial ovarian, fallopian tube, or primary peritoneal cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "\n",
            "\n",
            "ZIRABEV, in combination with carboplatin and paclitaxel, followed by ZIRABEV as a single agent, is indicated for the treatment of patients with stage III or IV epithelial ovarian, fallopian tube, or primary peritoneal cancer following initial surgical resection.\n",
            "\n",
            "\n",
            "ZIRABEV, in combination with paclitaxel, pegylated liposomal doxorubicin, or topotecan, is indicated for the treatment of patients with platinum-resistant recurrent epithelial ovarian, fallopian tube, or primary peritoneal cancer who received no more than 2 prior chemotherapy regimens.\n",
            "\n",
            "\n",
            "ZIRABEV, in combination with carboplatin and paclitaxel, or with carboplatin and gemcitabine, followed by ZIRABEV as a single agent, is indicated for the treatment of patients with platinum-sensitive recurrent epithelial ovarian, fallopian tube, or primary peritoneal cancer.\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BEVACIZUMAB&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=939b5d1f-9fb2-4499-80ef-0607aa6b114e\n",
            "\n",
            "Avastin, in combination with interferon alfa, is indicated for the treatment of metastatic renal cell carcinoma (mRCC).\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0a562b4e-67ce-4bab-852c-d9ee71e55fb8\n",
            "\n",
            "MVASI, in combination with interferon-alfa, is indicated for the treatment of metastatic renal cell carcinoma (mRCC).\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aa27acbd-d117-4350-aeee-17bc2e2c0ca4\n",
            "\n",
            "ZIRABEV, in combination with interferon alfa, is indicated for the treatment of metastatic renal cell carcinoma (mRCC).\n",
            "['Everolimus']\n",
            "Everolimus\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=EVEROLIMUS&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=01f7b5fb-be0d-4adf-acd2-6899018de966&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5aaf2fb7-93f6-40d6-941a-9583a3794ce7&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cf2dd54b-7a1d-4401-9a01-75e68235e433&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b7c7c9a4-994f-43dc-8cd6-cd6eeed6fa89&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated for the treatment of postmenopausal women with advanced hormone receptor-positive, HER2-negative breast cancer in combination with exemestane, after failure of treatment with letrozole or anastrozole. \n",
            "{'gene': [], 'protein': [], 'drug': ['letrozole']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3dbbc24-7e55-4069-a305-e2cded2db641&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=45d54974-3fc4-4237-a2fa-3158ca1a8105&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3bca36b-29ed-4a2b-a1be-a57b6af805f6&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated for the treatment of postmenopausal women with advanced hormone  receptor-positive, HER2-negative breast cancer in combination with exemestane, after failure of treatment with letrozole or anastrozole.\n",
            "{'gene': [], 'protein': [], 'drug': ['letrozole']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2150f73a-179b-4afc-b8ce-67c85cc72f04&audience=consumer\n",
            "\n",
            "AFINITOR is indicated for the treatment of postmenopausal women with advanced hormone receptor-positive, HER2-negative breast cancer in combination with exemestane, after failure of treatment with letrozole or anastrozole.\n",
            "{'gene': [], 'protein': [], 'drug': ['letrozole']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e082a024-7850-400b-a5c2-2a140612562a&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=EVEROLIMUS&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=01f7b5fb-be0d-4adf-acd2-6899018de966&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5aaf2fb7-93f6-40d6-941a-9583a3794ce7&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cf2dd54b-7a1d-4401-9a01-75e68235e433&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b7c7c9a4-994f-43dc-8cd6-cd6eeed6fa89&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3dbbc24-7e55-4069-a305-e2cded2db641&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=45d54974-3fc4-4237-a2fa-3158ca1a8105&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3bca36b-29ed-4a2b-a1be-a57b6af805f6&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2150f73a-179b-4afc-b8ce-67c85cc72f04&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e082a024-7850-400b-a5c2-2a140612562a&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=EVEROLIMUS&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=01f7b5fb-be0d-4adf-acd2-6899018de966&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5aaf2fb7-93f6-40d6-941a-9583a3794ce7&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cf2dd54b-7a1d-4401-9a01-75e68235e433&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b7c7c9a4-994f-43dc-8cd6-cd6eeed6fa89&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3dbbc24-7e55-4069-a305-e2cded2db641&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=45d54974-3fc4-4237-a2fa-3158ca1a8105&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3bca36b-29ed-4a2b-a1be-a57b6af805f6&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2150f73a-179b-4afc-b8ce-67c85cc72f04&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e082a024-7850-400b-a5c2-2a140612562a&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=EVEROLIMUS&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=01f7b5fb-be0d-4adf-acd2-6899018de966&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5aaf2fb7-93f6-40d6-941a-9583a3794ce7&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cf2dd54b-7a1d-4401-9a01-75e68235e433&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b7c7c9a4-994f-43dc-8cd6-cd6eeed6fa89&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3dbbc24-7e55-4069-a305-e2cded2db641&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=45d54974-3fc4-4237-a2fa-3158ca1a8105&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3bca36b-29ed-4a2b-a1be-a57b6af805f6&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2150f73a-179b-4afc-b8ce-67c85cc72f04&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e082a024-7850-400b-a5c2-2a140612562a&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=EVEROLIMUS&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=01f7b5fb-be0d-4adf-acd2-6899018de966&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated for the treatment of adult patients with advanced RCC after failure of treatment with sunitinib or sorafenib.\n",
            "{'gene': [], 'protein': [], 'drug': ['sunitinib']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5aaf2fb7-93f6-40d6-941a-9583a3794ce7&audience=consumer\n",
            "\n",
            "Everolimus is indicated for the treatment of adult patients with advanced RCC after failure of treatment with sunitinib or sorafenib.\n",
            "{'gene': [], 'protein': [], 'drug': ['sunitinib']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cf2dd54b-7a1d-4401-9a01-75e68235e433&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b7c7c9a4-994f-43dc-8cd6-cd6eeed6fa89&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3dbbc24-7e55-4069-a305-e2cded2db641&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated for the treatment of adult patients with advanced RCC after failure of treatment with sunitinib or sorafenib.\n",
            "{'gene': [], 'protein': [], 'drug': ['sunitinib']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=45d54974-3fc4-4237-a2fa-3158ca1a8105&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3bca36b-29ed-4a2b-a1be-a57b6af805f6&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated for the treatment of adult patients with advanced RCC after failure of treatment with sunitinib or sorafenib.\n",
            "{'gene': [], 'protein': [], 'drug': ['sunitinib']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2150f73a-179b-4afc-b8ce-67c85cc72f04&audience=consumer\n",
            "\n",
            "AFINITOR is indicated for the treatment of adult patients with advanced RCC after failure of treatment with sunitinib or sorafenib.\n",
            "{'gene': [], 'protein': [], 'drug': ['sunitinib']}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e082a024-7850-400b-a5c2-2a140612562a&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=EVEROLIMUS&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=01f7b5fb-be0d-4adf-acd2-6899018de966&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated in adult and pediatric patients aged 1 year and older with TSC for the treatment of SEGA that requires therapeutic intervention but cannot be curatively resected.\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5aaf2fb7-93f6-40d6-941a-9583a3794ce7&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cf2dd54b-7a1d-4401-9a01-75e68235e433&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b7c7c9a4-994f-43dc-8cd6-cd6eeed6fa89&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated in adult and pediatric patients aged 1 year and older with TSC for the treatment of SEGA that requires therapeutic intervention but cannot be curatively resected.\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3dbbc24-7e55-4069-a305-e2cded2db641&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated in adult and pediatric patients aged 1 year and older with TSC for the treatment of SEGA that requires therapeutic intervention but cannot be curatively resected.\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=45d54974-3fc4-4237-a2fa-3158ca1a8105&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e3bca36b-29ed-4a2b-a1be-a57b6af805f6&audience=consumer\n",
            "\n",
            "Everolimus tablets are indicated in adult and pediatric patients aged 1 year and older with TSC for the treatment of SEGA that requires therapeutic intervention but cannot be curatively resected.\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2150f73a-179b-4afc-b8ce-67c85cc72f04&audience=consumer\n",
            "\n",
            "AFINITOR and AFINITOR DISPERZ are indicated in adult and pediatric patients aged 1 year and older with TSC for the treatment of SEGA that requires therapeutic intervention but cannot be curatively resected.\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e082a024-7850-400b-a5c2-2a140612562a&audience=consumer\n",
            "['Everolimus']\n",
            "Everolimus\n",
            "['Tamoxifen Citrate']\n",
            "Tamoxifen Citrate\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=tamoxifen+citrate&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9b8a4211-120f-4981-ad69-928accb97637\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a596c715-8e30-43c1-aba3-4fe5de6729fa\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=70052366-80f3-41bc-b2dd-c1efe9317243\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8e02d45c-2b87-4c2f-8d75-b848172fafde\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see BOXED WARNING at the beginning of the label).\n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-844-825-8500.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see Table 3 in CLINICAL PHARMACOLOGY).\n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fa87c1c2-f58b-4a12-a658-a957ae7e522d\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7ee3d3d2-85d1-4018-8e70-5ed8a64ae1f0\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen citrate tablets are an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen citrate tablet therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in postmenopausal women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen- and progesterone-receptor values may help to predict whether adjuvant tamoxifen citrate tablet therapy is likely to be beneficial.\n",
            "Tamoxifen citrate tablets reduce the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen citrate tablet therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen citrate tablets on overall or breast cancer-related mortality (see BOXED WARNING).\n",
            "Tamoxifen citrate tablets are indicated only for high risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "Age 60 or older and:\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Healthcare Professionals can obtain a Gail Model Risk Assessment Tool by dialing Mylan Pharmaceuticals Inc. at 1-877-446-3679 (1-877-4-INFO-RX).\n",
            "There are insufficient data available regarding the effect of tamoxifen citrate tablets on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy. In the NSABP P-1 trial, tamoxifen citrate tablet treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see CLINICAL PHARMACOLOGY: Table 3).\n",
            "{'gene': ['tamoxifen citrate'], 'protein': [], 'drug': ['tamoxifen citrate']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8f642753-9e12-433c-a0bc-ab33dac41ddf\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8d731e7e-3b04-49a2-a0f6-3779422174cc\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=716657e8-d809-4671-87bc-8d10ef69a071\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b02f410e-ba92-4853-9b67-2afbde9866be\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen citrate tablets are an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen citrate tablet therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in postmenopausal women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen- and progesterone-receptor values may help to predict whether adjuvant tamoxifen citrate tablet therapy is likely to be beneficial.\n",
            "Tamoxifen citrate tablets reduce the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen citrate tablet therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen citrate tablets on overall or breast cancer-related mortality (see BOXED WARNING).\n",
            "Tamoxifen citrate tablets are indicated only for high risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "Age 60 or older and:\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Healthcare Professionals can obtain a Gail Model Risk Assessment Tool by dialing Mylan Pharmaceuticals Inc. at 1-877-446-3679 (1-877-4-INFO-RX).\n",
            "There are insufficient data available regarding the effect of tamoxifen citrate tablets on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy. In the NSABP P-1 trial, tamoxifen citrate tablet treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see CLINICAL PHARMACOLOGY: Table 3).\n",
            "{'gene': ['tamoxifen citrate'], 'protein': [], 'drug': ['tamoxifen citrate']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=62ec5985-9a78-4a17-bc82-47f1cd676935\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1bb81dd5-8d4e-4d6d-b030-a8ff2a3d4f55\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see BOXED WARNING at the beginning of the label).\n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-844-825-8500.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see Table 3 in CLINICAL PHARMACOLOGY).\n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cfd2e863-544d-4f32-b53d-dda6ef26522f\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see \n",
            "         \n",
            " \n",
            "  BOXED WARNING at the beginning of the label).\n",
            "        \n",
            "\n",
            " \n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. High risk is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-844-825-8500.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see \n",
            "         \n",
            " \n",
            "  Table 3 in \n",
            "         \n",
            " \n",
            "  CLINICAL PHARMACOLOGY).\n",
            "        \n",
            "\n",
            " \n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=151ce76c-2d44-4c82-850d-ab213bea4d6b\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see \n",
            "  BOXED WARNING at the beginning of the label).\n",
            " \n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. High risk is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-888-825-8500.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see \n",
            "  Table 3 in \n",
            "  CLINICAL PHARMACOLOGY).\n",
            " \n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=86dcde0e-347c-49e1-8b38-ad106a25c9af\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see BOXED WARNING at the beginning of the label).\n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. High risk is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-888-838-2872.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see Table 3 in CLINICAL PHARMACOLOGY).\n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8ed73b24-caeb-79ce-e053-2a95a90a2569\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a5615dbb-3c5e-6ccc-e053-2995a90ab698\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=bb53eeb7-83fc-469e-b9e3-df5f27bfec7b\n",
            "\n",
            "Tamoxifen citrate tablets, USP are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets, USP are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets, USP are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see BOXED WARNING at the beginning of the label).\n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. High risk is defined as women at least 35 years of age with a 5-year predicted risk of breast cancer greater than or equal to1.67%, as calculated by the Gail Model. \n",
            "Examples of combinations of factors predicting a 5-year risk greater than or equal to 1.67% are: \n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-800-272-5525.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen citrate in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see Table 3 in CLINICAL PHARMACOLOGY).\n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4fff5311-1129-46ba-a123-c90cfcda56f3\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen citrate tablets are an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen citrate tablet therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in postmenopausal women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen- and progesterone-receptor values may help to predict whether adjuvant tamoxifen citrate tablet therapy is likely to be beneficial.\n",
            "Tamoxifen citrate tablets reduce the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen citrate tablet therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen citrate tablets on overall or breast cancer-related mortality (see \n",
            "         \n",
            " \n",
            "  BOXED WARNING).\n",
            "        \n",
            "\n",
            " \n",
            "Tamoxifen citrate tablets are indicated only for high risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "Age 60 or older and:\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Healthcare Professionals can obtain a Gail Model Risk Assessment Tool by dialing Mylan Pharmaceuticals Inc. at 1-877-446-3679 (1-877-4-INFO-RX).\n",
            "There are insufficient data available regarding the effect of tamoxifen citrate tablets on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy. In the NSABP P-1 trial, tamoxifen citrate tablet treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see CLINICAL PHARMACOLOGY: \n",
            "         \n",
            " \n",
            "  Table 3).\n",
            "        \n",
            "\n",
            " \n",
            "{'gene': ['tamoxifen citrate'], 'protein': [], 'drug': ['tamoxifen citrate']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=bd72fd53-ae0c-4080-9b7f-2d294b14cc86\n",
            "\n",
            "SOLTAMOX is indicated for the treatment of adult patients with estrogen receptor-positive metastatic breast cancer.\n",
            "\n",
            "SOLTAMOX is indicated:\n",
            "\n",
            "SOLTAMOX is indicated to reduce the incidence of breast cancer in adult women at high risk for breast cancer.[see Boxed Warning and Clinical Studies (14.4)].\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=tamoxifen+citrate&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9b8a4211-120f-4981-ad69-928accb97637\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a596c715-8e30-43c1-aba3-4fe5de6729fa\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=70052366-80f3-41bc-b2dd-c1efe9317243\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8e02d45c-2b87-4c2f-8d75-b848172fafde\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see BOXED WARNING at the beginning of the label). The decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy.\n",
            "Current data from clinical trials support 5 years of adjuvant tamoxifen therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fa87c1c2-f58b-4a12-a658-a957ae7e522d\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7ee3d3d2-85d1-4018-8e70-5ed8a64ae1f0\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see BOXED WARNING). The decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy.\n",
            "Current data from clinical trials support 5 years of adjuvant tamoxifen citrate tablet therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8f642753-9e12-433c-a0bc-ab33dac41ddf\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8d731e7e-3b04-49a2-a0f6-3779422174cc\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=716657e8-d809-4671-87bc-8d10ef69a071\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b02f410e-ba92-4853-9b67-2afbde9866be\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see BOXED WARNING). The decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy.\n",
            "Current data from clinical trials support 5 years of adjuvant tamoxifen citrate tablet therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=62ec5985-9a78-4a17-bc82-47f1cd676935\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1bb81dd5-8d4e-4d6d-b030-a8ff2a3d4f55\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see BOXED WARNING at the beginning of the label). The decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy.\n",
            "Current data from clinical trials support 5 years of adjuvant tamoxifen therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cfd2e863-544d-4f32-b53d-dda6ef26522f\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see \n",
            "         \n",
            " \n",
            "  BOXED WARNING at the beginning of the label). The decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy.\n",
            "        \n",
            "\n",
            " \n",
            "Current data from clinical trials support 5 years of adjuvant tamoxifen therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=151ce76c-2d44-4c82-850d-ab213bea4d6b\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see \n",
            "         \n",
            " \n",
            "  \n",
            "         \n",
            " \n",
            "  BOXED WARNING at the beginning of the label). The decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy.\n",
            "        \n",
            "\n",
            " \n",
            "        \n",
            "\n",
            " \n",
            "Current data from clinical trials support 5 years of adjuvant tamoxifen therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=86dcde0e-347c-49e1-8b38-ad106a25c9af\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see BOXED WARNING at the beginning of the label). The decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy.\n",
            "Current data from clinical trials support 5 years of adjuvant tamoxifen therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8ed73b24-caeb-79ce-e053-2a95a90a2569\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a5615dbb-3c5e-6ccc-e053-2995a90ab698\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=bb53eeb7-83fc-469e-b9e3-df5f27bfec7b\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see BOXED WARNING at the beginning of the label). The decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. \n",
            "Current data from clinical trials support 5years of adjuvant tamoxifen therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4fff5311-1129-46ba-a123-c90cfcda56f3\n",
            "\n",
            "In women with DCIS, following breast surgery and radiation, tamoxifen citrate tablets are indicated to reduce the risk of invasive breast cancer (see \n",
            "         \n",
            " \n",
            "  BOXED WARNING). The decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy.\n",
            "        \n",
            "\n",
            " \n",
            "Current data from clinical trials support 5 years of adjuvant tamoxifen citrate tablet therapy for patients with breast cancer.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=bd72fd53-ae0c-4080-9b7f-2d294b14cc86\n",
            "\n",
            "In adult women with DCIS, following breast surgery and radiation, SOLTAMOX is indicated to reduce the risk of invasive breast cancer [see Boxed Warning and Clinical Studies (14.3)].\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=tamoxifen+citrate&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9b8a4211-120f-4981-ad69-928accb97637\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a596c715-8e30-43c1-aba3-4fe5de6729fa\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=70052366-80f3-41bc-b2dd-c1efe9317243\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8e02d45c-2b87-4c2f-8d75-b848172fafde\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see BOXED WARNING at the beginning of the label).\n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-844-825-8500.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see Table 3 in CLINICAL PHARMACOLOGY).\n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fa87c1c2-f58b-4a12-a658-a957ae7e522d\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7ee3d3d2-85d1-4018-8e70-5ed8a64ae1f0\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen citrate tablets are an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen citrate tablet therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in postmenopausal women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen- and progesterone-receptor values may help to predict whether adjuvant tamoxifen citrate tablet therapy is likely to be beneficial.\n",
            "Tamoxifen citrate tablets reduce the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen citrate tablet therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen citrate tablets on overall or breast cancer-related mortality (see BOXED WARNING).\n",
            "Tamoxifen citrate tablets are indicated only for high risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "Age 60 or older and:\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Healthcare Professionals can obtain a Gail Model Risk Assessment Tool by dialing Mylan Pharmaceuticals Inc. at 1-877-446-3679 (1-877-4-INFO-RX).\n",
            "There are insufficient data available regarding the effect of tamoxifen citrate tablets on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy. In the NSABP P-1 trial, tamoxifen citrate tablet treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see CLINICAL PHARMACOLOGY: Table 3).\n",
            "{'gene': ['tamoxifen citrate'], 'protein': [], 'drug': ['tamoxifen citrate']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8f642753-9e12-433c-a0bc-ab33dac41ddf\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8d731e7e-3b04-49a2-a0f6-3779422174cc\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=716657e8-d809-4671-87bc-8d10ef69a071\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b02f410e-ba92-4853-9b67-2afbde9866be\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen citrate tablets are an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen citrate tablet therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in postmenopausal women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen- and progesterone-receptor values may help to predict whether adjuvant tamoxifen citrate tablet therapy is likely to be beneficial.\n",
            "Tamoxifen citrate tablets reduce the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen citrate tablet therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen citrate tablets on overall or breast cancer-related mortality (see BOXED WARNING).\n",
            "Tamoxifen citrate tablets are indicated only for high risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "Age 60 or older and:\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Healthcare Professionals can obtain a Gail Model Risk Assessment Tool by dialing Mylan Pharmaceuticals Inc. at 1-877-446-3679 (1-877-4-INFO-RX).\n",
            "There are insufficient data available regarding the effect of tamoxifen citrate tablets on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy. In the NSABP P-1 trial, tamoxifen citrate tablet treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see CLINICAL PHARMACOLOGY: Table 3).\n",
            "{'gene': ['tamoxifen citrate'], 'protein': [], 'drug': ['tamoxifen citrate']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=62ec5985-9a78-4a17-bc82-47f1cd676935\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1bb81dd5-8d4e-4d6d-b030-a8ff2a3d4f55\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see BOXED WARNING at the beginning of the label).\n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-844-825-8500.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see Table 3 in CLINICAL PHARMACOLOGY).\n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cfd2e863-544d-4f32-b53d-dda6ef26522f\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see \n",
            "         \n",
            " \n",
            "  BOXED WARNING at the beginning of the label).\n",
            "        \n",
            "\n",
            " \n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. High risk is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-844-825-8500.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see \n",
            "         \n",
            " \n",
            "  Table 3 in \n",
            "         \n",
            " \n",
            "  CLINICAL PHARMACOLOGY).\n",
            "        \n",
            "\n",
            " \n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=151ce76c-2d44-4c82-850d-ab213bea4d6b\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see \n",
            "  BOXED WARNING at the beginning of the label).\n",
            " \n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. High risk is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-888-825-8500.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see \n",
            "  Table 3 in \n",
            "  CLINICAL PHARMACOLOGY).\n",
            " \n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=86dcde0e-347c-49e1-8b38-ad106a25c9af\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see BOXED WARNING at the beginning of the label).\n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. High risk is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-888-838-2872.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see Table 3 in CLINICAL PHARMACOLOGY).\n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8ed73b24-caeb-79ce-e053-2a95a90a2569\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a5615dbb-3c5e-6ccc-e053-2995a90ab698\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=bb53eeb7-83fc-469e-b9e3-df5f27bfec7b\n",
            "\n",
            "Tamoxifen citrate tablets, USP are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen is an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen therapy.\n",
            "\n",
            "Tamoxifen citrate tablets, USP are indicated for the treatment of node-positive breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets, USP are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen and progesterone receptor values may help to predict whether adjuvant tamoxifen therapy is likely to be beneficial.\n",
            "Tamoxifen reduces the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen on overall or breast cancer-related mortality (see BOXED WARNING at the beginning of the label).\n",
            "Tamoxifen citrate tablets are indicated only for high-risk women. High risk is defined as women at least 35 years of age with a 5-year predicted risk of breast cancer greater than or equal to1.67%, as calculated by the Gail Model. \n",
            "Examples of combinations of factors predicting a 5-year risk greater than or equal to 1.67% are: \n",
            "\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "\n",
            "\n",
            "Age 60 or older and:\n",
            "\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Health Care Professionals can obtain a Gail Model Risk Assessment Tool by dialing 1-800-272-5525.\n",
            "There are insufficient data available regarding the effect of tamoxifen on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen citrate in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate therapy. In the NSABP P-1 trial, tamoxifen treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see Table 3 in CLINICAL PHARMACOLOGY).\n",
            "{'gene': [], 'protein': [], 'drug': ['tamoxifen']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4fff5311-1129-46ba-a123-c90cfcda56f3\n",
            "\n",
            "Tamoxifen citrate tablets are effective in the treatment of metastatic breast cancer in women and men. In premenopausal women with metastatic breast cancer, tamoxifen citrate tablets are an alternative to oophorectomy or ovarian irradiation. Available evidence indicates that patients whose tumors are estrogen receptor positive are more likely to benefit from tamoxifen citrate tablet therapy.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated for the treatment of node-positive breast cancer in postmenopausal women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation. In some tamoxifen adjuvant studies, most of the benefit to date has been in the subgroup with four or more positive axillary nodes.\n",
            "Tamoxifen citrate tablets are indicated for the treatment of axillary node-negative breast cancer in women following total mastectomy or segmental mastectomy, axillary dissection, and breast irradiation.\n",
            "The estrogen- and progesterone-receptor values may help to predict whether adjuvant tamoxifen citrate tablet therapy is likely to be beneficial.\n",
            "Tamoxifen citrate tablets reduce the occurrence of contralateral breast cancer in patients receiving adjuvant tamoxifen citrate tablet therapy for breast cancer.\n",
            "\n",
            "Tamoxifen citrate tablets are indicated to reduce the incidence of breast cancer in women at high risk for breast cancer. This effect was shown in a study of 5 years planned duration with a median follow-up of 4.2 years. Twenty-five percent of the participants received drug for 5 years. The longer-term effects are not known. In this study, there was no impact of tamoxifen citrate tablets on overall or breast cancer-related mortality (see \n",
            "         \n",
            " \n",
            "  BOXED WARNING).\n",
            "        \n",
            "\n",
            " \n",
            "Tamoxifen citrate tablets are indicated only for high risk women. \"High risk\" is defined as women at least 35 years of age with a 5 year predicted risk of breast cancer  1.67%, as calculated by the Gail Model.\n",
            "Examples of combinations of factors predicting a 5 year risk  1.67% are:\n",
            "Age 35 or older and any of the following combination of factors:\n",
            "Age 40 or older and any of the following combination of factors:\n",
            "Age 45 or older and any of the following combination of factors:\n",
            "Age 50 or older and any of the following combination of factors:\n",
            "Age 55 or older and any of the following combination of factors:\n",
            "Age 60 or older and:\n",
            "For women whose risk factors are not described in the above examples, the Gail Model is necessary to estimate absolute breast cancer risk. Healthcare Professionals can obtain a Gail Model Risk Assessment Tool by dialing Mylan Pharmaceuticals Inc. at 1-877-446-3679 (1-877-4-INFO-RX).\n",
            "There are insufficient data available regarding the effect of tamoxifen citrate tablets on breast cancer incidence in women with inherited mutations (BRCA1, BRCA2) to be able to make specific recommendations on the effectiveness of tamoxifen in these patients.\n",
            "After an assessment of the risk of developing breast cancer, the decision regarding therapy with tamoxifen citrate tablets for the reduction in breast cancer incidence should be based upon an individual assessment of the benefits and risks of tamoxifen citrate tablet therapy. In the NSABP P-1 trial, tamoxifen citrate tablet treatment lowered the risk of developing breast cancer during the follow-up period of the trial, but did not eliminate breast cancer risk (see CLINICAL PHARMACOLOGY: \n",
            "         \n",
            " \n",
            "  Table 3).\n",
            "        \n",
            "\n",
            " \n",
            "{'gene': ['tamoxifen citrate'], 'protein': [], 'drug': ['tamoxifen citrate']}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=bd72fd53-ae0c-4080-9b7f-2d294b14cc86\n",
            "\n",
            "SOLTAMOX is indicated for the treatment of adult patients with estrogen receptor-positive metastatic breast cancer.\n",
            "\n",
            "SOLTAMOX is indicated:\n",
            "\n",
            "SOLTAMOX is indicated to reduce the incidence of breast cancer in adult women at high risk for breast cancer.[see Boxed Warning and Clinical Studies (14.4)].\n",
            "['Toremifene']\n",
            "Toremifene\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2cab8dd1-3a10-48e6-86ce-0e5275ed49e5&audience=consumer\n",
            "['Trastuzumab']\n",
            "Trastuzumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=492dbdb2-077e-4064-bff3-372d6af0a7a2&audience=consumer\n",
            "\n",
            "Herceptin is indicated for adjuvant treatment of HER2 overexpressing node positive or node negative (ER/PR negative or with one high risk feature [see Clinical Studies (14.1)]) breast cancer\n",
            "\tSelect patients for therapy based on an FDA-approved companion diagnostic for Herceptin [see Dosage and Administration (2.1)].\n",
            "\n",
            "Herceptin is indicated:\n",
            "\tSelect patients for therapy based on an FDA-approved companion diagnostic for Herceptin [see Dosage and Administration (2.1)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=492dbdb2-077e-4064-bff3-372d6af0a7a2&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=492dbdb2-077e-4064-bff3-372d6af0a7a2&audience=consumer\n",
            "['Fulvestrant']\n",
            "Fulvestrant\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=83d7a440-e904-4e36-afb5-cb02b1c919f7&audience=consumer\n",
            "['Anastrozole']\n",
            "Anastrozole\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=Anastrozole&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7b3da447-8e7c-2aaa-e053-2a91aa0a2821&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=03586c40-eb32-45c5-beb7-6762a6b78790&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c7666def-a855-4ca3-a322-603a88a44101&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4a23dc55-9877-4b43-ac2c-cd628c0f5038&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=ca9d4aff-ffdf-4a46-ad20-e47e5886a19e&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3cd04ed6-7c5a-4c4a-9955-ee40839dfea0&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=188c6fd5-6695-4d8b-af43-61d9c26abfe9&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=965c4cb1-f6b8-47ac-bf13-c36639aa908f&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=ca16f8de-d9a5-4c6c-957b-ae28497b67ac&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=71f053c2-8d0d-41a1-8630-2d797eced799&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=d4550afa-977e-426e-814b-705435d2af81&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b33520de-f891-43f8-bd04-1b66795d3dd2&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c58c7013-50a0-4734-8a78-38ff20a3e7eb&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8270c509-4c1c-bfd8-e053-2991aa0abb90&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=df55dc4c-9a53-4a37-a90f-c4e88d458b89&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4bd7e768-2d09-44e6-bacb-7390a5bc5dbc&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=07ccde5f-e6f2-4e9a-b1de-d5adc047f6bc&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5df9393e-b51e-1a8e-e053-2a91aa0a0907&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b964f555-25d2-47ad-8942-f195f062445c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=03fb3108-1856-466d-bf88-0eac96139551&audience=consumer\n",
            "['Exemestane']\n",
            "Exemestane\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=EXEMESTANE&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=61d86c55-9946-273f-4f22-42d3fc33d143\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fab5468a-d5b7-45e3-b8b2-ed902fddd3a6\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=abde1ccd-0ec9-40d4-b205-fbc00f5878b9\n",
            "\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=88fb6543-383b-4195-a2c1-2673ac194f4b\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=11d5a86e-7861-43f1-ae04-84daa0de7888\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)]. \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4e3fd382-19f5-4ded-ab22-39d5b1b2f366\n",
            "\n",
            "Exemestane tablets is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9354d1a1-8e8c-4c6a-b7ad-d4fe14b10a8b\n",
            "\n",
            "Exemestane is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3573e3c9-ca05-413f-909f-93607a2ccb79\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2ecdb646-4d0f-4d1b-ac6a-9cb5deb9ac0d\n",
            "\n",
            "Exemestane is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=69abd0ff-b200-4d1a-b14a-80e980a8e781\n",
            "\n",
            "EXEMESTANE is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f8bd17b4-5289-45bb-8095-de89dc8d39ed\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f1bbe1bd-fa92-4fdd-88df-31ebcb200d37\n",
            "\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)]. \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f0b31daa-0792-43d5-af08-8b6a864dc90a\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=15e11379-e681-4f2b-bdd6-dbbe56e563b0\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1841ec12-3f4f-4b87-b719-fd0b4d9eaffb\n",
            "\n",
            "Exemestane tablets is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cf066b7a-032a-416c-8d40-15ba581423e3\n",
            "\n",
            "AROMASIN is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=EXEMESTANE&pagesize=20&page=1\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=61d86c55-9946-273f-4f22-42d3fc33d143\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fab5468a-d5b7-45e3-b8b2-ed902fddd3a6\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=abde1ccd-0ec9-40d4-b205-fbc00f5878b9\n",
            "\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=88fb6543-383b-4195-a2c1-2673ac194f4b\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=11d5a86e-7861-43f1-ae04-84daa0de7888\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)]. \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4e3fd382-19f5-4ded-ab22-39d5b1b2f366\n",
            "\n",
            "Exemestane tablets is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9354d1a1-8e8c-4c6a-b7ad-d4fe14b10a8b\n",
            "\n",
            "Exemestane is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3573e3c9-ca05-413f-909f-93607a2ccb79\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2ecdb646-4d0f-4d1b-ac6a-9cb5deb9ac0d\n",
            "\n",
            "Exemestane is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=69abd0ff-b200-4d1a-b14a-80e980a8e781\n",
            "\n",
            "EXEMESTANE is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f8bd17b4-5289-45bb-8095-de89dc8d39ed\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f1bbe1bd-fa92-4fdd-88df-31ebcb200d37\n",
            "\n",
            "\n",
            "Exemestane tablets are indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)]. \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f0b31daa-0792-43d5-af08-8b6a864dc90a\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=15e11379-e681-4f2b-bdd6-dbbe56e563b0\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1841ec12-3f4f-4b87-b719-fd0b4d9eaffb\n",
            "\n",
            "Exemestane tablets is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cf066b7a-032a-416c-8d40-15ba581423e3\n",
            "\n",
            "AROMASIN is indicated for the treatment of advanced breast cancer in postmenopausal women whose disease has progressed following tamoxifen therapy [see Clinical Studies (14.2)].\n",
            "\n",
            "['Lapatinib Ditosylate']\n",
            "Lapatinib Ditosylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=eee37f88-ec6a-4c30-b8aa-e2c71f93088c&audience=consumer\n",
            "['Letrozole']\n",
            "Letrozole\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=LETROZOLE&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7b00e08c-9a1c-09e0-e053-2a91aa0a7638&audience=consumer\n",
            "\n",
            "Letrozole Tablets, USP are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer.\n",
            "\n",
            "Letrozole Tablets, USP are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole for a median of 60 months [\n",
            "         \n",
            " \n",
            "  see Clinical Studies (\n",
            "          \n",
            "  \n",
            "   14.2, \n",
            "          \n",
            "  \n",
            "   14.3)\n",
            "         \n",
            " \n",
            "  ].\n",
            "        \n",
            "\n",
            " \n",
            "\n",
            "Letrozole Tablets, USP are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer.Letrozole Tablets, USP are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy \n",
            "         \n",
            " \n",
            "  [see Clinical Studies (\n",
            "          \n",
            "  \n",
            "   14.4, \n",
            "          \n",
            "  \n",
            "   14.5)].\n",
            "         \n",
            " \n",
            "  \n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=63e54b62-a30d-4a56-9215-75e7a0b0bf02&audience=consumer\n",
            "\n",
            "Letrozole tablets are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer.\n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months [see Clinical Studies (14.2, 14.3)].\n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [see Clinical Studies (14.4, 14.5)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c391fe11-fd47-4a35-af6d-c823ab7b8838&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=32fd146f-7652-877f-cfd1-db9fa367406a&audience=consumer\n",
            "\n",
            "Letrozole Tablets, USP are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole Tablets, USP are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [see Clinical Studies (14.4, 14.5)]. \n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=968611f1-ef03-47a6-8dae-4834f77d23bd&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=ff7325d2-3383-4ff9-a187-9350e3538640&audience=consumer\n",
            "\n",
            " Letrozole tablets  are indicated for the adjuvant  treatment of postmenopausal  women with  hormone receptor positive early breast  cancer.\n",
            "\n",
            " Letrozole tablets  are indicated for the extended adjuvant  treatment of early breast cancer in  postmenopausal women, who have received 5  years of adjuvant  tamoxifen therapy. The  effectiveness of letrozole tablets in extended adjuvant  treatment of early breast  cancer is based on an analysis of disease-free survival in patients treated with  letrozole tablets for a median of 60  months [see Clinical Studies (14.2, 14.3)].\n",
            "\n",
            " Letrozole tablets are indicated for first-line  treatment of  postmenopausal women with hormone receptor positive or unknown,  locally advanced or metastatic breast  cancer. Letrozole tablets are also indicated for the  treatment of advanced breast  cancer in postmenopausal  women with disease progression following antiestrogen therapy [see Clinical Studies (14.4, 14.5)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6cfd29d7-f519-7b8c-4814-07b39f367f0d&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8ebda73d-82a3-4b17-8a81-a1f258003d32&audience=consumer\n",
            "\n",
            "Letrozole tablets are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer. \n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months [see Clinical Studies (14.2, 14.3)]. \n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [see Clinical Studies (14.4, 14.5)]. \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=87b9906a-1e3f-4bd7-9f93-35d20a386315&audience=consumer\n",
            "\n",
            "Letrozole tablets (letrozole) are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer.\n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months [see Clinical Studies (14.2, 14.3)].\n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [see Clinical Studies(14.4, 14.5)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=63560663-ecc4-4529-bbd7-c63c1877dfba&audience=consumer\n",
            "\n",
            "Letrozole tablets are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer.\n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months [\n",
            "         \n",
            " \n",
            "  \n",
            "         \n",
            " \n",
            "  see Clinical Studies (\n",
            "          \n",
            "  \n",
            "   \n",
            "          \n",
            "  \n",
            "   14.2, \n",
            "          \n",
            "  \n",
            "   \n",
            "          \n",
            "  \n",
            "   14.3)\n",
            "         \n",
            " \n",
            "  \n",
            "         \n",
            " \n",
            "  ].\n",
            "        \n",
            "\n",
            " \n",
            "        \n",
            "\n",
            " \n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [\n",
            "         \n",
            " \n",
            "  \n",
            "         \n",
            " \n",
            "  see Clinical Studies(\n",
            "          \n",
            "  \n",
            "   \n",
            "          \n",
            "  \n",
            "   14.4, \n",
            "          \n",
            "  \n",
            "   \n",
            "          \n",
            "  \n",
            "   14.5)\n",
            "         \n",
            " \n",
            "  \n",
            "         \n",
            " \n",
            "  ].\n",
            "        \n",
            "\n",
            " \n",
            "        \n",
            "\n",
            " \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=35c73219-c39e-4e10-b944-76009b061e79&audience=consumer\n",
            "\n",
            "Letrozole tablets are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer.\n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months [see Clinical Studies (14.2, 14.3)]. \n",
            "\n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [see Clinical Studies (14.4, 14.5)]. \n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=93f317f1-96d1-498e-e053-2995a90ac60d&audience=consumer\n",
            "\n",
            "Letrozole tablets are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer.\n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months\n",
            "         \n",
            " \n",
            "   [see\n",
            "          \n",
            "  \n",
            "    Clinical Studies (14.2,\n",
            "          \n",
            "  \n",
            "    14.3)]\n",
            "         \n",
            " \n",
            "  .\n",
            "        \n",
            "\n",
            " \n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy\n",
            "         \n",
            " \n",
            "   [see\n",
            "          \n",
            "  \n",
            "    Clinical Studies (14.4,\n",
            "          \n",
            "  \n",
            "    14.5)].\n",
            "         \n",
            " \n",
            "  \n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3cc750a5-cc88-4b42-b4ab-a87e760f291d&audience=consumer\n",
            "\n",
            "Letrozole tablets are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer.\n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months [see Clinical Studies (14.2, 14.3)].\n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [see Clinical Studies (14.4, 14.5)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=065cc89f-3a3f-4d78-b9f6-1c32a77ceb1c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c10bf97f-f9cf-468a-8a1f-7637f41022e4&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1524730f-e37d-478c-1956-b113b3ca42a8&audience=consumer\n",
            "\n",
            "Letrozole tablets are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer. \n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months [\n",
            "  see Clinical Studies \n",
            "   (14.2, \n",
            "   14.3)].\n",
            " \n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [\n",
            "  see Clinical Studies \n",
            "   (14.4, \n",
            "   14.5)].\n",
            "  \n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=26dcb6ee-74d8-4804-b151-79a41484979b&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a707ec4b-8d26-55c2-e053-2a95a90a2e07&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b73bf908-7979-4955-af20-0414388920da&audience=consumer\n",
            "\n",
            "\n",
            "Letrozole \n",
            "tablets are indicated for the adjuvant treatment of postmenopausal women with \n",
            "hormone receptor positive early breast cancer.\n",
            "\n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant \n",
            "treatment of early breast cancer in postmenopausal women, who have received 5 \n",
            "years of adjuvant tamoxifen therapy. The effectiveness of letrozole in extended \n",
            "adjuvant treatment of early breast cancer is based on an analysis of \n",
            "disease-free survival in patients treated with letrozole tablets for a median of \n",
            "60 months [see Clinical Studies \n",
            "(14.2, 14.3)]. \n",
            "\n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of \n",
            "postmenopausal women with hormone receptor positive or unknown, locally advanced \n",
            "or metastatic breast cancer. Letrozole tablets are also indicated for the \n",
            "treatment of advanced breast cancer in postmenopausal women with disease \n",
            "progression following antiestrogen therapy [see Clinical Studies \n",
            "(14.4, 14.5)]. \n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0193faa4-9b02-4ccd-9751-de6124cc8c68&audience=consumer\n",
            "\n",
            "Letrozole tablets are indicated for the adjuvant treatment of postmenopausal women with hormone receptor positive early breast cancer. \n",
            "\n",
            "Letrozole tablets are indicated for the extended adjuvant treatment of early breast cancer in postmenopausal women, who have received 5 years of adjuvant tamoxifen therapy. The effectiveness of letrozole tablets in extended adjuvant treatment of early breast cancer is based on an analysis of disease-free survival in patients treated with letrozole tablets for a median of 60 months [see Clinical Studies (14.2, 14.3)]. \n",
            "\n",
            "Letrozole tablets are indicated for first-line treatment of postmenopausal women with hormone receptor positive or unknown, locally advanced or metastatic breast cancer. Letrozole tablets are also indicated for the treatment of advanced breast cancer in postmenopausal women with disease progression following antiestrogen therapy [see Clinical Studies (14.4, 14.5)]. \n",
            "['Pertuzumab']\n",
            "Pertuzumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=17f85d17-ab71-4f5b-9fe3-0b8c822f69ff&audience=consumer\n",
            "\n",
            "PERJETA is indicated for use in combination with trastuzumab and docetaxel for the treatment of patients with HER2-positive metastatic breast cancer who have not received prior anti-HER2 therapy or chemotherapy for metastatic disease [see Dosage and Administration (2.2) and Clinical Studies (14.1)].\n",
            "\n",
            "PERJETA is indicated for use in combination with trastuzumab and chemotherapy for \n",
            "{'gene': [], 'protein': ['anti-HER2'], 'drug': []}\n",
            "['Ado-Trastuzumab Emtansine']\n",
            "Ado-Trastuzumab Emtansine\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=23f3c1f4-0fc8-4804-a9e3-04cf25dd302e&audience=consumer\n",
            "\n",
            "KADCYLA, as a single agent, is indicated for the treatment of patients with HER2-positive, metastatic breast cancer who previously received trastuzumab and a taxane, separately or in combination. Patients should have either:\n",
            "Select patients for therapy based on an FDA-approved companion diagnostic for KADCYLA [see Dosage and Administration (2.1)].\n",
            "\n",
            "KADCYLA, as a single agent, is indicated for the adjuvant treatment of patients with HER2-positive early breast cancer who have residual invasive disease after neoadjuvant taxane and trastuzumab -based treatment.\n",
            "Select patients for therapy based on an FDA-approved companion diagnostic for KADCYLA [see Dosage and Administration (2.1)].\n",
            "['Palbociclib']\n",
            "Palbociclib\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=PALBOCICLIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9c77010e-cf3e-4425-b298-ceb2847872b6&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e0e6412f-50b4-4fd4-9364-62818d121a07&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=86718885-da44-4d3d-b3ca-06cbcedcffd8&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fecbdd7d-b729-41b5-9872-231b8fe104ce&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=89a142d5-7e61-48bf-9e77-e2cd124c4792&audience=consumer\n",
            "['Ribociclib']\n",
            "Ribociclib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aaeaef94-f3f5-4367-8ea2-b181d7be2da8&audience=consumer\n",
            "['Neratinib Maleate']\n",
            "Neratinib Maleate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=d1e41dcf-e82a-47c2-a0ad-6c6eef621834&audience=consumer\n",
            "\n",
            "NERLYNX as a single agent is indicated for the extended adjuvant treatment of adult patients with early-stage human epidermal growth factor receptor 2 (HER2)-positive breast cancer, to follow adjuvant trastuzumab based therapy \n",
            "         \n",
            " \n",
            "  [see Clinical Studies (\n",
            "          \n",
            "  \n",
            "   14.1)]\n",
            "         \n",
            " \n",
            "  .\n",
            "\n",
            "        \n",
            "\n",
            " \n",
            "\n",
            "\n",
            "NERLYNX in combination with capecitabine is indicated for the treatment of adult patients with advanced or metastatic HER2-positive breast cancer who have received two or more prior anti-HER2 based regimens in the metastatic setting \n",
            "         .\n",
            "\n",
            "        NERLYNX in combination with capecitabine is indicated for the treatment of adult patients with advanced or metastatic HER2-positive breast cancer who have received two or more prior anti-HER2 based regimens in the metastatic setting \n",
            "         \n",
            " \n",
            "   [see Clinical Studies (\n",
            "          \n",
            "  \n",
            "    14.2)]\n",
            "         \n",
            " \n",
            "   .\n",
            "\n",
            "        \n",
            "\n",
            "  \n",
            "\n",
            "['Abemaciclib']\n",
            "Abemaciclib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=be4bc0de-0fdc-4d46-8d25-be43c79e6a06&audience=consumer\n",
            "['Olaparib']\n",
            "Olaparib\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=OLAPARIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a242ecac-1e94-4afa-ae0f-40a5018a938c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=741ff3e3-dc1a-45a6-84e5-2481b27131aa&audience=consumer\n",
            "\n",
            "\n",
            "Lynparza is indicated for the treatment of adult patients with deleterious or suspected deleterious gBRCAm, HER2-negative metastatic breast cancer, who have been treated with chemotherapy in the neoadjuvant, adjuvant, or metastatic setting. Patients with hormone receptor (HR)-positive breast cancer should have been treated with a prior endocrine therapy or be considered inappropriate for endocrine therapy. Select patients for therapy based on an FDA-approved companion diagnostic for Lynparza [see Dosage and Administration (2.1)].\n",
            "\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=OLAPARIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a242ecac-1e94-4afa-ae0f-40a5018a938c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=741ff3e3-dc1a-45a6-84e5-2481b27131aa&audience=consumer\n",
            "\n",
            "\n",
            "Lynparza is indicated in combination with bevacizumab for the maintenance treatment of adult patients with advanced epithelial ovarian, fallopian tube or primary peritoneal cancer who are in complete or partial response to first-line platinum-based chemotherapy and whose cancer is associated with homologous recombination deficiency (HRD)-positive status defined by either:\n",
            "\n",
            "\n",
            "Select patients for therapy based on an FDA-approved companion diagnostic for Lynparza [see Dosage and Administration (2.1)].\n",
            "\n",
            "\n",
            "Lynparza is indicated for the maintenance treatment of adult patients with recurrent epithelial ovarian, fallopian tube or primary peritoneal cancer, who are in complete or partial response to platinum-based chemotherapy.\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=OLAPARIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a242ecac-1e94-4afa-ae0f-40a5018a938c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=741ff3e3-dc1a-45a6-84e5-2481b27131aa&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=OLAPARIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a242ecac-1e94-4afa-ae0f-40a5018a938c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=741ff3e3-dc1a-45a6-84e5-2481b27131aa&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=OLAPARIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a242ecac-1e94-4afa-ae0f-40a5018a938c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=741ff3e3-dc1a-45a6-84e5-2481b27131aa&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=OLAPARIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a242ecac-1e94-4afa-ae0f-40a5018a938c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=741ff3e3-dc1a-45a6-84e5-2481b27131aa&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=OLAPARIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a242ecac-1e94-4afa-ae0f-40a5018a938c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=741ff3e3-dc1a-45a6-84e5-2481b27131aa&audience=consumer\n",
            "\n",
            "\n",
            "Lynparza is indicated for the treatment of adult patients with deleterious or suspected deleterious germline or somatic homologous recombination repair (HRR) gene-mutated metastatic castration-resistant prostate cancer (mCRPC) who have progressed following prior treatment with enzalutamide or abiraterone. Select patients for therapy based on an FDA-approved companion diagnostic for Lynparza [see Dosage and Administration (2.1)].\n",
            "\n",
            "['Talazoparib Tosylate']\n",
            "Talazoparib Tosylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f2423edd-6d24-495c-aec1-c2f457f08d9a\n",
            "['Atezolizumab']\n",
            "Atezolizumab\n",
            "['Alpelisib']\n",
            "Alpelisib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b20b4e18-7a4b-4500-a08f-06c6dab0ee5b&audience=consumer\n",
            "['Fam-Trastuzumab Deruxtecan-nxki']\n",
            "Fam-Trastuzumab Deruxtecan-nxki\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7e67e73e-ddf4-4e4d-8b50-09d7514910b6&audience=consumer\n",
            "\n",
            "ENHERTU is indicated for the treatment of adult patients with unresectable or metastatic HER2-positive breast cancer who have received two or more prior anti-HER2-based regimens in the metastatic setting.\n",
            "This indication is approved under accelerated approval based on tumor response rate and duration of response [see Clinical Studies (14.1)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in a confirmatory trial.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7e67e73e-ddf4-4e4d-8b50-09d7514910b6&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7e67e73e-ddf4-4e4d-8b50-09d7514910b6&audience=consumer\n",
            "['Tucatinib']\n",
            "Tucatinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f27eb1b9-b7fc-424e-988f-84dd7bb195a3&audience=consumer\n",
            "['Sacituzumab Govitecan-hziy']\n",
            "Sacituzumab Govitecan-hziy\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=57a597d2-03f0-472e-b148-016d7169169d&audience=consumer\n",
            "['Pertuzumab, Trastuzumab, and Hyaluronidase-zzxf']\n",
            "Pertuzumab, Trastuzumab, and Hyaluronidase-zzxf\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=27dd5e6b-72cd-458d-a015-cf4dab5800da&audience=consumer\n",
            "\n",
            "PHESGO is indicated for use in combination with chemotherapy for\n",
            "Select patients for therapy based on an FDA-approved companion diagnostic test [see Dosage and Administration (2.1)].\n",
            "\n",
            "\n",
            "PHESGO is indicated for use in combination with docetaxel for the treatment of adult patients with HER2-positive metastatic breast cancer who have not received prior anti-HER2 therapy or chemotherapy for metastatic disease [see Dosage and Administration (2.2) and Clinical Studies (14.1)].\n",
            "Select patients for therapy based on an FDA-approved companion diagnostic test [see Dosage and Administration (2.1)].\n",
            "\n",
            "{'gene': [], 'protein': ['anti-HER2'], 'drug': []}\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Margetuximab-cmkb']\n",
            "Margetuximab-cmkb\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e97a5872-eabf-463b-8f4c-5b5aed9c7bf0&audience=consumer\n",
            "['Bevacizumab']\n",
            "Bevacizumab\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Cetuximab']\n",
            "Cetuximab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8bc6397e-4bd8-4d37-a007-a327e4da34d9&audience=consumer\n",
            "\n",
            "ERBITUX is indicated for the treatment of K-Ras wild-type, epidermal growth factor receptor (EGFR)-expressing, metastatic colorectal cancer (mCRC) as determined by an FDA-approved test [see\n",
            "Dosage and Administration (2.2)]:\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8bc6397e-4bd8-4d37-a007-a327e4da34d9&audience=consumer\n",
            "\n",
            "ERBITUX is indicated:\n",
            "\n",
            "['Panitumumab']\n",
            "Panitumumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e0fa4bca-f245-4d92-ae29-b0c630a315c2&audience=consumer\n",
            "\n",
            "\n",
            "Vectibix is indicated for the treatment of patients with wild-type RAS (defined as wild-type in both KRAS and NRASas determined by an FDA-approved test for this use)metastatic colorectal cancer (mCRC)[see Dosage and Administration (2.1)]:\n",
            "\n",
            "\n",
            "As first-line therapy in combination with FOLFOX[see Clinical Studies (14.2)].\n",
            "\n",
            "\n",
            "As monotherapy following disease progression after prior treatment with fluoropyrimidine-, oxaliplatin-, and irinotecan-containing chemotherapy [see Clinical Studies (14.1)].\n",
            "\n",
            "\n",
            "Limitation of Use: Vectibix is not indicated for the treatment of patients with RAS-mutant mCRC or for whom RASmutation status is unknown[see Dosage and Administration (2.1), Warnings and Precautions (5.2), and Clinical Pharmacology (12.1)].\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "['Bevacizumab']\n",
            "Bevacizumab\n",
            "['Ziv-Aflibercept']\n",
            "Ziv-Aflibercept\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f6725df6-50ee-4b0a-b900-d02ba634395d&audience=consumer\n",
            "['Regorafenib']\n",
            "Regorafenib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=824f19c9-0546-4a8a-8d8f-c4055c04f7c7&audience=consumer\n",
            "\n",
            "STIVARGA is indicated for the treatment of patients with metastatic colorectal cancer (CRC) who have been previously treated with fluoropyrimidine-, oxaliplatin- and irinotecan-based chemotherapy, an anti-VEGF therapy, and, if RAS wild- type, an anti-EGFR therapy.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=824f19c9-0546-4a8a-8d8f-c4055c04f7c7&audience=consumer\n",
            "\n",
            "STIVARGA is indicated for the treatment of patients with locally advanced, unresectable or metastatic gastrointestinal stromal tumor (GIST) who have been previously treated with imatinib mesylate and sunitinib malate.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=824f19c9-0546-4a8a-8d8f-c4055c04f7c7&audience=consumer\n",
            "\n",
            "STIVARGA is indicated for the treatment of patients with hepatocellular carcinoma (HCC) who have been previously treated with sorafenib. \n",
            "['Ramucirumab']\n",
            "Ramucirumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c6080942-dee6-423e-b688-1272c2ae90d4&audience=consumer\n",
            "\n",
            "CYRAMZA, in combination with FOLFIRI (irinotecan, folinic acid, and fluorouracil), is indicated for the treatment of patients with metastatic colorectal cancer (mCRC) with disease progression on or after prior therapy with bevacizumab, oxaliplatin, and a fluoropyrimidine.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c6080942-dee6-423e-b688-1272c2ae90d4&audience=consumer\n",
            "\n",
            "\n",
            "CYRAMZA, as a single agent, is indicated for the treatment of patients with hepatocellular carcinoma (HCC) who have an alpha fetoprotein (AFP) of 400ng/mL and have been treated with sorafenib.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c6080942-dee6-423e-b688-1272c2ae90d4&audience=consumer\n",
            "\n",
            "\n",
            "CYRAMZA, in combination with erlotinib, is indicated for the first-line treatment of patients with metastatic non-small cell lung cancer (NSCLC) whose tumors have epidermal growth factor receptor (EGFR) exon 19 deletions or exon 21 (L858R) substitution mutations.\n",
            "\n",
            "CYRAMZA, in combination with docetaxel, is indicated for the treatment of patients with metastatic non-small cell lung cancer (NSCLC) with disease progression on or after platinum-based chemotherapy. Patients with epidermal growth factor receptor (EGFR) or anaplastic lymphoma kinase (ALK) genomic tumor aberrations should have disease progression on FDA-approved therapy for these aberrations prior to receiving CYRAMZA.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c6080942-dee6-423e-b688-1272c2ae90d4&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c6080942-dee6-423e-b688-1272c2ae90d4&audience=consumer\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Ipilimumab']\n",
            "Ipilimumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2265ef30-253e-11df-8a39-0800200c9a66&audience=consumer\n",
            "\n",
            "YERVOY, in combination with nivolumab, is indicated for the treatment of adult and pediatric patients 12 years and older with microsatellite instability-high (MSI-H) or mismatch repair deficient (dMMR) metastatic colorectal cancer (mCRC) that has progressed following treatment with a fluoropyrimidine, oxaliplatin, and irinotecan. \n",
            "This indication is approved under accelerated approval based on overall response rate and duration of response [see Clinical Studies (14.4)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in confirmatory trials.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2265ef30-253e-11df-8a39-0800200c9a66&audience=consumer\n",
            "\n",
            "\n",
            "YERVOY, in combination with nivolumab, is indicated for the treatment of patients with hepatocellular carcinoma (HCC) who have been previously treated with sorafenib. This indication is approved under accelerated approval based on overall response rate and duration of response [see Clinical Studies (14.5)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in the confirmatory trials.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2265ef30-253e-11df-8a39-0800200c9a66&audience=consumer\n",
            "\n",
            "\n",
            "YERVOY, in combination with nivolumab, is indicated for the first-line treatment of adult patients with unresectable malignant pleural mesothelioma.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2265ef30-253e-11df-8a39-0800200c9a66&audience=consumer\n",
            "\n",
            "YERVOY is indicated for the treatment of unresectable or metastatic melanoma in adults and pediatric patients 12 years and older.\n",
            "\n",
            "YERVOY is indicated for the adjuvant treatment of patients with cutaneous melanoma with pathologic involvement of regional lymph nodes of more than 1 mm who have undergone complete resection, including total lymphadenectomy.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2265ef30-253e-11df-8a39-0800200c9a66&audience=consumer\n",
            "\n",
            "\n",
            "YERVOY, in combination with nivolumab, is indicated for the first-line treatment of adult patients with metastatic non-small cell lung cancer (NSCLC) whose tumors express PD-L1 (1%) as determined by an FDA-approved test [see Dosage and Administration (2.1)], with no EGFR or ALK genomic tumor aberrations.\n",
            "\n",
            "YERVOY, in combination with nivolumab and 2 cycles of platinum-doublet chemotherapy, is indicated for the first-line treatment of adult patients with metastatic or recurrent NSCLC, with no EGFR or ALK genomic tumor aberrations.\n",
            "{'gene': ['EGFR', 'ALK genomic'], 'protein': ['EGFR'], 'drug': []}\n",
            "{'gene': ['EGFR', 'ALK genomic'], 'protein': ['EGFR'], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2265ef30-253e-11df-8a39-0800200c9a66&audience=consumer\n",
            "\n",
            "YERVOY, in combination with nivolumab, is indicated for the treatment of patients with intermediate or poor risk, previously untreated advanced renal cell carcinoma (RCC).\n",
            "['Encorafenib']\n",
            "Encorafenib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=235dfc38-0f0b-4037-b501-7a9f4294740c&audience=consumer\n",
            "\n",
            "\n",
            "BRAFTOVI is indicated, in combination with cetuximab, for the treatment of adult patients with metastatic colorectal cancer (CRC) with a BRAF V600E mutation, as detected by an FDA-approved test, after prior therapy [see Dosage and Administration (2.1)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=235dfc38-0f0b-4037-b501-7a9f4294740c&audience=consumer\n",
            "\n",
            "BRAFTOVI is indicated, in combination with binimetinib, for the treatment of patients with unresectable or metastatic melanoma with a BRAF V600E or V600K mutation, as detected by an FDA-approved test [see Dosage and Administration (2.1)].\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Imatinib Mesylate']\n",
            "Imatinib Mesylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=211ef2da-2868-4a77-8055-1cb2cd78e24b&audience=consumer\n",
            "\n",
            "Adult patients with relapsed or refractory Philadelphia chromosome positive acute lymphoblastic leukemia (Ph+ ALL).\n",
            "\n",
            "Pediatric patients with newly diagnosed Philadelphia chromosome positive acute lymphoblastic leukemia (Ph+ ALL) in combination with chemotherapy.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=211ef2da-2868-4a77-8055-1cb2cd78e24b&audience=consumer\n",
            "\n",
            "Adult patients with hypereosinophilic syndrome and/or chronic eosinophilic leukemia who have the FIP1L1-PDGFR fusion kinase (mutational analysis or fluorescence in situ hybridization [FISH] demonstration of CHIC2 allele deletion) and for patients with HES and/or CEL who are FIP1L1-PDGFR fusion kinase negative or unknown.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=211ef2da-2868-4a77-8055-1cb2cd78e24b&audience=consumer\n",
            "\n",
            "Adult patients with hypereosinophilic syndrome and/or chronic eosinophilic leukemia who have the FIP1L1-PDGFR fusion kinase (mutational analysis or fluorescence in situ hybridization [FISH] demonstration of CHIC2 allele deletion) and for patients with HES and/or CEL who are FIP1L1-PDGFR fusion kinase negative or unknown.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=211ef2da-2868-4a77-8055-1cb2cd78e24b&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=211ef2da-2868-4a77-8055-1cb2cd78e24b&audience=consumer\n",
            "\n",
            "Adult patients with unresectable, recurrent and/or metastatic dermatofibrosarcoma protuberans.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=211ef2da-2868-4a77-8055-1cb2cd78e24b&audience=consumer\n",
            "\n",
            "Patients with Kit (CD117) positive unresectable and/or metastatic malignant gastrointestinal stromal tumors. \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=211ef2da-2868-4a77-8055-1cb2cd78e24b&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=211ef2da-2868-4a77-8055-1cb2cd78e24b&audience=consumer\n",
            "\n",
            "\n",
            "Adult patients with aggressive systemic mastocytosis without the D816V c-Kit mutation or with c-Kit mutational status unknown.\n",
            "\n",
            "{'gene': ['D816V c-Kit mutation'], 'protein': ['D816V c-Kit'], 'drug': []}\n",
            "['Lanreotide Acetate']\n",
            "Lanreotide Acetate\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=LANREOTIDE+ACETATE&pagesize=20&page=1&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=LANREOTIDE+ACETATE&pagesize=20&page=1&audience=consumer\n",
            "['Avelumab']\n",
            "Avelumab\n",
            "['Lutetium Lu 177-Dotatate']\n",
            "Lutetium Lu 177-Dotatate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=72d1a024-00b7-418a-b36e-b2cb48f2ab55\n",
            "['Iobenguane I 131']\n",
            "Iobenguane I 131\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3423ac0f-9585-4abf-a8ba-a5123a8277ee\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3423ac0f-9585-4abf-a8ba-a5123a8277ee\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Lenvatinib Mesylate']\n",
            "Lenvatinib Mesylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f4bedd21-efde-44c6-9d9c-b48b78d7ed1e&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f4bedd21-efde-44c6-9d9c-b48b78d7ed1e&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f4bedd21-efde-44c6-9d9c-b48b78d7ed1e&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f4bedd21-efde-44c6-9d9c-b48b78d7ed1e&audience=consumer\n",
            "['Trastuzumab']\n",
            "Trastuzumab\n",
            "['Ramucirumab']\n",
            "Ramucirumab\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Fam-Trastuzumab Deruxtecan-nxki']\n",
            "Fam-Trastuzumab Deruxtecan-nxki\n",
            "['Cetuximab']\n",
            "Cetuximab\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Imatinib Mesylate']\n",
            "Imatinib Mesylate\n",
            "['Sunitinib Malate']\n",
            "Sunitinib Malate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=43a4d7f8-48ae-4a63-9108-2fa8e3ea9d9c&audience=consumer\n",
            "\n",
            "SUTENT is indicated for the treatment of adult patients with gastrointestinal stromal tumor (GIST) after disease progression on or intolerance to imatinib mesylate.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=43a4d7f8-48ae-4a63-9108-2fa8e3ea9d9c&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=43a4d7f8-48ae-4a63-9108-2fa8e3ea9d9c&audience=consumer\n",
            "\n",
            "SUTENT is indicated for the treatment of adult patients with advanced renal cell carcinoma (RCC).\n",
            "\n",
            "SUTENT is indicated for the adjuvant treatment of adult patients at high risk of recurrent RCC following nephrectomy.\n",
            "['Regorafenib']\n",
            "Regorafenib\n",
            "['Avapritinib']\n",
            "Avapritinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=645c887c-8cd4-4623-8da9-ac223d71a8b9&audience=consumer\n",
            "\n",
            "AYVAKIT is indicated for the treatment of adults with unresectable or metastatic GIST harboring a platelet-derived growth factor receptor alpha (PDGFRA) exon 18 mutation, including PDGFRA D842V mutations [see Dosage and Administration (2.1)].\n",
            "['Ripretinib']\n",
            "Ripretinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9f18e462-03dd-4296-a02a-a0577e3ee78d&audience=consumer\n",
            "['Denosumab']\n",
            "Denosumab\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=DENOSUMAB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=49e5afe9-a0c7-40c4-af9f-f287a80c5c88&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=628f0998-1206-4001-aeee-18133aa9f3bf&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=DENOSUMAB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=49e5afe9-a0c7-40c4-af9f-f287a80c5c88&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=628f0998-1206-4001-aeee-18133aa9f3bf&audience=consumer\n",
            "\n",
            "Xgeva is indicated for the treatment of hypercalcemia of malignancy refractory to bisphosphonate therapy.\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=DENOSUMAB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=49e5afe9-a0c7-40c4-af9f-f287a80c5c88&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=628f0998-1206-4001-aeee-18133aa9f3bf&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=DENOSUMAB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=49e5afe9-a0c7-40c4-af9f-f287a80c5c88&audience=consumer\n",
            "\n",
            "Prolia is indicated as a treatment to increase bone mass in women at high risk for fracture receiving adjuvant aromatase inhibitor therapy for breast cancer[see Clinical Studies (14.5)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=628f0998-1206-4001-aeee-18133aa9f3bf&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=DENOSUMAB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=49e5afe9-a0c7-40c4-af9f-f287a80c5c88&audience=consumer\n",
            "\n",
            "Prolia is indicated as a treatment to increase bone mass in men at high risk for fracturereceiving androgen deprivation therapy for nonmetastatic prostate cancer.  In these patients Prolia also reduced the incidence of vertebral fractures [see Clinical Studies (14.4)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=628f0998-1206-4001-aeee-18133aa9f3bf&audience=consumer\n",
            "['Pexidartinib Hydrochloride']\n",
            "Pexidartinib Hydrochloride\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=24c5a6f1-b640-4d14-b44b-dd293ed002b1&audience=consumer\n",
            "['Bevacizumab']\n",
            "Bevacizumab\n",
            "['Sorafenib Tosylate']\n",
            "Sorafenib Tosylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b50667e4-5ebc-4968-a646-d605058dbef0&audience=consumer\n",
            "\n",
            "NEXAVAR is indicated for the treatment of patients with unresectable hepatocellular carcinoma (HCC). \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b50667e4-5ebc-4968-a646-d605058dbef0&audience=consumer\n",
            "\n",
            "NEXAVAR is indicated for the treatment of patients with advanced renal cell carcinoma (RCC). \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b50667e4-5ebc-4968-a646-d605058dbef0&audience=consumer\n",
            "['Sunitinib Malate']\n",
            "Sunitinib Malate\n",
            "['Pazopanib Hydrochloride']\n",
            "Pazopanib Hydrochloride\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=eeaaaf38-fb86-4d9f-a19d-0f61daac2fd7&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=eeaaaf38-fb86-4d9f-a19d-0f61daac2fd7&audience=consumer\n",
            "['Temsirolimus']\n",
            "Temsirolimus\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=95b7dc92-2180-42f1-8699-3c28f609e674&audience=consumer\n",
            "['Everolimus']\n",
            "Everolimus\n",
            "['Axitinib']\n",
            "Axitinib\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=AXITINIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=84137882-e000-47da-bd5b-fa76ab3c76f9&audience=consumer\n",
            "\n",
            "\n",
            "INLYTA in combination with avelumab is indicated for the first-line treatment of patients with advanced renal cell carcinoma (RCC).\n",
            "\n",
            "\n",
            "INLYTA in combination with pembrolizumab is indicated for the first-line treatment of patients with advanced renal cell carcinoma.\n",
            "\n",
            "\n",
            "INLYTA as a single agent is indicated for the treatment of advanced renal cell carcinoma (RCC) after failure of one prior systemic therapy. \n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8a903e31-936e-4ed7-8a59-59f32374f338&audience=consumer\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Cabozantinib-S-Malate']\n",
            "Cabozantinib-S-Malate\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=CABOZANTINIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3850cce2-6137-42e5-a792-d318c4a4b3b5&audience=consumer\n",
            "\n",
            "\n",
            "                           CABOMETYX is indicated for the treatment of patients with hepatocellular carcinoma (HCC) who have been previously treated with sorafenib.\n",
            "                        \n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1a0c3bea-c87b-4d25-bb44-5f0174da6b34&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=CABOZANTINIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3850cce2-6137-42e5-a792-d318c4a4b3b5&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1a0c3bea-c87b-4d25-bb44-5f0174da6b34&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=CABOZANTINIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3850cce2-6137-42e5-a792-d318c4a4b3b5&audience=consumer\n",
            "\n",
            "CABOMETYX is indicated for the treatment of patients with advanced renal cell carcinoma (RCC).\n",
            "\n",
            "CABOMETYX, in combination with nivolumab, is indicated for the first-line treatment of patients with advanced RCC.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1a0c3bea-c87b-4d25-bb44-5f0174da6b34&audience=consumer\n",
            "['Lenvatinib Mesylate']\n",
            "Lenvatinib Mesylate\n",
            "['Ipilimumab']\n",
            "Ipilimumab\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Avelumab']\n",
            "Avelumab\n",
            "[]\n",
            "['Imatinib Mesylate']\n",
            "Imatinib Mesylate\n",
            "['Dasatinib']\n",
            "Dasatinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4764f37b-c9e6-4ede-bcc2-8a03b7c521df&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4764f37b-c9e6-4ede-bcc2-8a03b7c521df&audience=consumer\n",
            "['Nilotinib']\n",
            "Nilotinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6093952a-5248-45cb-ad17-33716a411146&audience=consumer\n",
            "['Bosutinib']\n",
            "Bosutinib\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BOSUTINIB+MONOHYDRATE&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=adc84ad5-a04d-4fee-9ba8-91f7abd928e3&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=d5929f91-6496-4c0e-97e8-0bd524e15763&audience=consumer\n",
            "['Rituximab']\n",
            "Rituximab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b172773b-3905-4a1c-ad95-bab4b6126563&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b172773b-3905-4a1c-ad95-bab4b6126563&audience=consumer\n",
            "\n",
            "RITUXAN, in combination with fludarabine and cyclophosphamide (FC), is indicated for the treatment of adult patients with previously untreated and previously treated CD20-positive CLL.\n",
            "['Alemtuzumab']\n",
            "Alemtuzumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4f5f7255-7abc-4328-bd1a-ceaf139ef3e0&audience=consumer\n",
            "['Ofatumumab']\n",
            "Ofatumumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=77785ce3-e8df-4ca1-8f8e-6c418c6a17de&audience=consumer\n",
            "['Obinutuzumab']\n",
            "Obinutuzumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=df12ceb2-5b4b-4ab5-a317-2a36bf2a3cda&audience=consumer\n",
            "\n",
            "GAZYVA, in combination with chlorambucil, is indicated for the treatment of patients with previously untreated chronic lymphocytic leukemia.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=df12ceb2-5b4b-4ab5-a317-2a36bf2a3cda&audience=consumer\n",
            "\n",
            "GAZYVA, in combination with bendamustine followed by GAZYVA monotherapy, is indicated for the treatment of patients with follicular lymphoma who relapsed after, or are refractory to, a rituximab-containing regimen.\n",
            "\n",
            "GAZYVA, in combination with chemotherapy followed by GAZYVA monotherapy in patients achieving at least a partial remission, is indicated for the treatment of adult patients with previously untreated stage II bulky, III or IV follicular lymphoma.\n",
            "\n",
            "['Ibrutinib']\n",
            "Ibrutinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0dfd0279-ff17-4ea9-89be-9803c71bab44&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0dfd0279-ff17-4ea9-89be-9803c71bab44&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0dfd0279-ff17-4ea9-89be-9803c71bab44&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0dfd0279-ff17-4ea9-89be-9803c71bab44&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0dfd0279-ff17-4ea9-89be-9803c71bab44&audience=consumer\n",
            "['Idelalisib']\n",
            "Idelalisib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=efbdafa9-d18c-4e85-b4a2-1e620fc74e50&audience=consumer\n",
            "\n",
            "Zydelig is indicated, in combination with rituximab, for the treatment of patients with relapsed chronic lymphocytic leukemia (CLL) for whom rituximab alone would be considered appropriate therapy due to other co-morbidities.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=efbdafa9-d18c-4e85-b4a2-1e620fc74e50&audience=consumer\n",
            "\n",
            "Zydelig is indicated for the treatment of patients with relapsed follicular B-cell non-Hodgkin lymphoma (FL) who have received at least two prior systemic therapies.\n",
            "This indication is approved under accelerated approval based on overall response rate [see Clinical Studies (14.2)]. An improvement in patient survival or disease related symptoms has not been established. Continued approval for this indication may be contingent upon verification of clinical benefit in confirmatory trials. \n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "['Blinatumomab']\n",
            "Blinatumomab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=38b482a8-960b-4591-9857-5031ecb830aa&audience=consumer\n",
            "['Venetoclax']\n",
            "Venetoclax\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b118a40d-6b56-cee3-10f6-ded821a97018&audience=consumer\n",
            "\n",
            "\n",
            "VENCLEXTA is indicated in combination with azacitidine, or decitabine, or low-dose cytarabine for the treatment of newly diagnosed acute myeloid leukemia (AML) in adults 75 years or older, or who have comorbidities that preclude use of intensive induction chemotherapy.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b118a40d-6b56-cee3-10f6-ded821a97018&audience=consumer\n",
            "\n",
            "\n",
            "VENCLEXTA is indicated for the treatment of adult patients with chronic lymphocytic leukemia (CLL) or small lymphocytic lymphoma (SLL).\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b118a40d-6b56-cee3-10f6-ded821a97018&audience=consumer\n",
            "\n",
            "\n",
            "VENCLEXTA is indicated for the treatment of adult patients with chronic lymphocytic leukemia (CLL) or small lymphocytic lymphoma (SLL).\n",
            "['Ponatinib Hydrochloride']\n",
            "Ponatinib Hydrochloride\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=PONATINIB+HYDROCHLORIDE&pagesize=20&page=1&audience=consumer\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=PONATINIB+HYDROCHLORIDE&pagesize=20&page=1&audience=consumer\n",
            "['Midostaurin']\n",
            "Midostaurin\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=11fa3fc9-6776-49a6-b1c1-653f627c3e58&audience=consumer\n",
            "\n",
            "RYDAPT is indicated, in combination with standard cytarabine and daunorubicin induction and cytarabine consolidation chemotherapy, for the treatment of adult patients with newly diagnosed acute myeloid leukemia (AML) who are FLT3 mutation-positive, as detected by a FDA approved test [see Dosage and Administration (2.1), Clinical Studies (14.1)].\n",
            "\n",
            "\n",
            "Limitations of Use\n",
            "\n",
            "RYDAPT is not indicated as a single-agent induction therapy for the treatment of patients with AML.\n",
            "{'gene': [], 'protein': ['Use RYDAPT'], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=11fa3fc9-6776-49a6-b1c1-653f627c3e58&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=11fa3fc9-6776-49a6-b1c1-653f627c3e58&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=11fa3fc9-6776-49a6-b1c1-653f627c3e58&audience=consumer\n",
            "['Enasidenib Mesylate']\n",
            "Enasidenib Mesylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a5b4cdf0-3fa8-4c6c-80f6-8d8a00e3a5b6&audience=consumer\n",
            "\n",
            "IDHIFA is indicated for the treatment of adult patients with relapsed or refractory acute myeloid leukemia (AML) with an isocitrate dehydrogenase-2 (IDH2) mutation as detected by an FDA-approved test.\n",
            "['Inotuzumab Ozogamicin']\n",
            "Inotuzumab Ozogamicin\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cc7014b1-c775-411d-b374-8113248b4077&audience=consumer\n",
            "['Tisagenlecleucel']\n",
            "Tisagenlecleucel\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aad3ba54-dfd3-4cb3-9e2b-c5ef89559189&audience=consumer\n",
            "\n",
            "Patients up to 25 years ofagewith B-cell precursor acute lymphoblastic leukemia (ALL) that is refractory or in second or later relapse.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=aad3ba54-dfd3-4cb3-9e2b-c5ef89559189&audience=consumer\n",
            "['Gemtuzumab Ozogamicin']\n",
            "Gemtuzumab Ozogamicin\n",
            "['Rituximab and Hyaluronidase Human']\n",
            "Rituximab and Hyaluronidase Human\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3e5b7e82-f018-4eaf-ae78-d6145a906b20&audience=consumer\n",
            "\n",
            "RITUXAN HYCELA is indicated, in combination with fludarabine and cyclophosphamide (FC), for the treatment of adult patients with previously untreated and previously treated CLL.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3e5b7e82-f018-4eaf-ae78-d6145a906b20&audience=consumer\n",
            "\n",
            "RITUXAN HYCELA is indicated for the treatment of adult patients with previously untreated diffuse large B-cell lymphoma in combination with cyclophosphamide, doxorubicin, vincristine, prednisone (CHOP) or other anthracycline-based chemotherapy regimens.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3e5b7e82-f018-4eaf-ae78-d6145a906b20&audience=consumer\n",
            "\n",
            "RITUXAN HYCELA is indicated for the treatment of adult patients with:\n",
            "['Ivosidenib']\n",
            "Ivosidenib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5b6d8ae8-bac8-4e1b-b1b4-7f9665e62de5\n",
            "\n",
            "\n",
            "TIBSOVO is indicated for the treatment of newly-diagnosed acute myeloid leukemia (AML) with a susceptible isocitrate dehydrogenase-1 (IDH1) mutation as detected by an FDA-approved test in adult patients who are  75 years old or who have comorbidities that preclude use of intensive induction chemotherapy [see Dosage and Administration (2.1), Clinical Pharmacology (12.1) and Clinical Studies (14.1)].\n",
            "\n",
            "\n",
            "TIBSOVO is indicated for the treatment of adult patients with relapsed or refractory acute myeloid leukemia (AML) with a susceptible isocitrate dehydrogenase-1 (IDH1) mutation as detected by an FDA-approved test [see Dosage and Administration (2.1), Clinical Pharmacology (12.1) and Clinical Studies (14.2)].\n",
            "\n",
            "['Duvelisib']\n",
            "Duvelisib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=47547f93-489c-4298-aeea-d6a37ca6cb1d&audience=consumer\n",
            "\n",
            "COPIKTRA is indicated for the treatment of adult patients with relapsed or refractory CLL or SLL after at least two prior therapies.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=47547f93-489c-4298-aeea-d6a37ca6cb1d&audience=consumer\n",
            "\n",
            "COPIKTRA is indicated for the treatment of adult patients with relapsed or refractory CLL or SLL after at least two prior therapies.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=47547f93-489c-4298-aeea-d6a37ca6cb1d&audience=consumer\n",
            "\n",
            "COPIKTRA is indicated for the treatment of adult patients with relapsed or refractory FL after at least two prior systemic therapies.\n",
            "This indication is approved under accelerated approval based on overall response rate (ORR) \n",
            "\t\t\t\t\t\t\t\t\t[see Clinical Studies (14.2)]; continued approval for this indication may be contingent upon verification and description of clinical benefit in confirmatory trials.\n",
            "\t\t\t\t\t\t\t\t\n",
            "['Moxetumomab Pasudotox-tdfk']\n",
            "Moxetumomab Pasudotox-tdfk\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=d6510282-1a57-4614-9859-299a227a089c\n",
            "['Glasdegib Maleate']\n",
            "Glasdegib Maleate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=204a6f7e-c9a4-472b-abd2-9527bda64d17&audience=consumer\n",
            "['Gilteritinib Fumarate']\n",
            "Gilteritinib Fumarate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b5ff59aa-9c0d-49a8-9053-1f179b482383&audience=consumer\n",
            "\n",
            "XOSPATA is indicated for the treatment of adult patients who have relapsed or refractory acute myeloid leukemia (AML) with a FMS-like tyrosine kinase 3 (FLT3) mutation as detected by an FDA-approved test.\n",
            "['Tagraxofusp-erzs']\n",
            "Tagraxofusp-erzs\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=80eb1080-fe20-4af0-b2a5-a67999923fa8&audience=consumer\n",
            "['Acalabrutinib']\n",
            "Acalabrutinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=dd4835ef-e1bc-4997-a399-1ffa2556fbfe&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=dd4835ef-e1bc-4997-a399-1ffa2556fbfe&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=dd4835ef-e1bc-4997-a399-1ffa2556fbfe&audience=consumer\n",
            "\n",
            "CALQUENCE is indicated for the treatment of adult patients with mantle cell lymphoma (MCL) who have received at least one prior therapy.\n",
            "This indication is approved under accelerated approval based on overall response rate [see Clinical Studies (14.1)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in confirmatory trials.\n",
            "['Sorafenib Tosylate']\n",
            "Sorafenib Tosylate\n",
            "['Regorafenib']\n",
            "Regorafenib\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Lenvatinib Mesylate']\n",
            "Lenvatinib Mesylate\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Cabozantinib-S-Malate']\n",
            "Cabozantinib-S-Malate\n",
            "['Ramucirumab']\n",
            "Ramucirumab\n",
            "['Ipilimumab']\n",
            "Ipilimumab\n",
            "['Pemigatinib']\n",
            "Pemigatinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9e1f2222-1d89-4e63-989c-ccebe2ab1eb4&audience=consumer\n",
            "['Atezolizumab']\n",
            "Atezolizumab\n",
            "['Bevacizumab']\n",
            "Bevacizumab\n",
            "['Bevacizumab']\n",
            "Bevacizumab\n",
            "['Crizotinib']\n",
            "Crizotinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2a51b0de-47d6-455e-a94c-d2c737b04ff7&audience=consumer\n",
            "\n",
            "\n",
            "XALKORI is indicated for the treatment of pediatric patients 1 year of age and older and young adults with relapsed or refractory, systemic anaplastic large cell lymphoma (ALCL) that is ALK-positive.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2a51b0de-47d6-455e-a94c-d2c737b04ff7&audience=consumer\n",
            "\n",
            "XALKORI is indicated for the treatment of patients with metastatic non-small cell lung cancer (NSCLC) whose tumors are anaplastic lymphoma kinase (ALK) or ROS1-positive as detected by an FDA-approved test [see Dosage and Administration (2.1)].\n",
            "['Erlotinib Hydrochloride']\n",
            "Erlotinib Hydrochloride\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=57bccb29-1c47-4c64-ab6a-77960a91cc20&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=57bccb29-1c47-4c64-ab6a-77960a91cc20&audience=consumer\n",
            "\n",
            "TARCEVA in combination with gemcitabine is indicated for the first-line treatment of patients with locally advanced, unresectable or metastatic pancreatic cancer [see Clinical Studies (14.5)].\n",
            "['Gefitinib']\n",
            "Gefitinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=827d60e8-7e07-41b7-c28b-49ef1c4a5a41&audience=consumer\n",
            "['Afatinib Dimaleate']\n",
            "Afatinib Dimaleate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fd638e5e-8032-e7ca-0179-95e96ab5d387&audience=consumer\n",
            "\n",
            "GILOTRIF is indicated for the first-line treatment of patients with metastatic non-small cell lung cancer (NSCLC) whose tumors have non-resistant epidermal growth factor receptor (EGFR) mutations as detected by an FDA-approved test [see Dosage and Administration (2.1), Clinical Pharmacology (12.1), Clinical Studies (14.1)].\n",
            "\n",
            "['Ceritinib']\n",
            "Ceritinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fff5d805-4ffd-4e8e-8e63-6f129697563e&audience=consumer\n",
            "['Ramucirumab']\n",
            "Ramucirumab\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Osimertinib Mesylate']\n",
            "Osimertinib Mesylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5e81b4a7-b971-45e1-9c31-29cea8c87ce7&audience=consumer\n",
            "\n",
            "\n",
            "TAGRISSO is indicated as adjuvant therapy after tumor resection in adult patients with non-small cell lung cancer (NSCLC) whose tumors have epidermal growth factor receptor (EGFR) exon 19 deletions or exon 21 L858R mutations, as detected by an FDA-approved test [see Dosage and Administration (2.1)].\n",
            "\n",
            "['Necitumumab']\n",
            "Necitumumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=89bcf553-669a-40b0-a9d7-67a5c1d2f591&audience=consumer\n",
            "\n",
            "PORTRAZZA is indicated, in combination with gemcitabine and cisplatin, for first-line treatment of patients with metastatic squamous non-small cell lung cancer.\n",
            "\n",
            "['Alectinib']\n",
            "Alectinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=42c49deb-713b-427a-9670-08af08adcffb&audience=consumer\n",
            "['Atezolizumab']\n",
            "Atezolizumab\n",
            "['Brigatinib']\n",
            "Brigatinib\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BRIGATINIB&pagesize=20&page=1&audience=consumer\n",
            "['Trametinib Dimethyl Sulfoxide']\n",
            "Trametinib Dimethyl Sulfoxide\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0002ad27-779d-42ab-83b5-bc65453412a1&audience=consumer\n",
            "\n",
            "\n",
            "MEKINIST is indicated, in combination with dabrafenib, for the treatment of patients with locally advanced or metastatic anaplastic thyroid cancer (ATC) with BRAF V600E mutation and with no satisfactory locoregional treatment options [see Dosage and Administration (2.1)].\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0002ad27-779d-42ab-83b5-bc65453412a1&audience=consumer\n",
            "\n",
            "\n",
            "MEKINIST is indicated, as a single agent in BRAF-inhibitor treatment-nave patients or in combination with dabrafenib, for the treatment of patients with unresectable or metastatic melanoma with BRAF V600E or V600K mutations, as detected by an FDA-approved test [see Dosage and Administration (2.1)].\n",
            "\n",
            "\n",
            "\n",
            "MEKINIST is indicated, in combination with dabrafenib, for the adjuvant treatment of patients with melanoma with BRAF V600E or V600K mutations as detected by an FDA-approved test, and involvement of lymph node(s), following complete resection [see Dosage and Administration (2.1)].\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0002ad27-779d-42ab-83b5-bc65453412a1&audience=consumer\n",
            "['Dabrafenib Mesylate']\n",
            "Dabrafenib Mesylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fee1e6b1-e1a5-4254-9f2e-a70e0f8dbdea&audience=consumer\n",
            "\n",
            "TAFINLAR is indicated, in combination with trametinib, for the treatment of patients with locally advanced or metastatic anaplastic thyroid cancer (ATC) with BRAF V600E mutation and with no satisfactory locoregional treatment options [see Dosage and Administration (2.1)].\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fee1e6b1-e1a5-4254-9f2e-a70e0f8dbdea&audience=consumer\n",
            "\n",
            "TAFINLARisindicated as a single agent for the treatment of patients with unresectable or metastatic melanoma withBRAFV600Emutation as detected by an FDA-approved test.\n",
            "\n",
            "TAFINLAR is indicated, in combination with trametinib,for the treatment of patients with unresectable or metastatic melanoma with BRAF V600E orV600K mutations, as detected by an FDA-approved test [see Dosage and Administration (2.1)].\n",
            "\n",
            "TAFINLAR is indicated, in combination with trametinib, for the adjuvant treatment of patients with melanoma with BRAF V600E or V600K mutations, as detected by an FDA-approved test, and involvement of lymph node(s), following complete resection [see Dosage and Administration (2.1)].\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fee1e6b1-e1a5-4254-9f2e-a70e0f8dbdea&audience=consumer\n",
            "['Durvalumab']\n",
            "Durvalumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8baba4ea-2855-42fa-9bd9-5a7548d4cec3&audience=consumer\n",
            "\n",
            "IMFINZI is indicated for the treatment of adult patients with unresectable Stage III non-small cell lung cancer (NSCLC) whose disease has not progressed following concurrent platinum-based chemotherapy and radiation therapy.\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=8baba4ea-2855-42fa-9bd9-5a7548d4cec3&audience=consumer\n",
            "\n",
            "IMFINZI is indicated for the treatment of adult patients with unresectable Stage III non-small cell lung cancer (NSCLC) whose disease has not progressed following concurrent platinum-based chemotherapy and radiation therapy.\n",
            "\n",
            "\n",
            "\n",
            "IMFINZI, in combination with etoposide and either carboplatin or cisplatin, is indicated for the first-line treatment of adult patients with extensive-stage small cell lung cancer (ES-SCLC).\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "['Dacomitinib']\n",
            "Dacomitinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4ab27d2f-e385-4e9c-b324-fa69c10b855a\n",
            "['Lorlatinib']\n",
            "Lorlatinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2b34d62d-e02a-4af3-bc0d-1571dd4ee76d&audience=consumer\n",
            "['Entrectinib']\n",
            "Entrectinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c7c71b0c-2549-4495-86b6-c2807fa54908&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c7c71b0c-2549-4495-86b6-c2807fa54908&audience=consumer\n",
            "['Capmatinib Hydrochloride']\n",
            "Capmatinib Hydrochloride\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=455892c3-d144-4ba8-9ab4-79cabff9876d&audience=consumer\n",
            "['Ipilimumab']\n",
            "Ipilimumab\n",
            "['Selpercatinib']\n",
            "Selpercatinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7fa848ba-a59c-4144-9f52-64d090f4d828&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7fa848ba-a59c-4144-9f52-64d090f4d828&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7fa848ba-a59c-4144-9f52-64d090f4d828&audience=consumer\n",
            "['Pralsetinib']\n",
            "Pralsetinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=beb226ab-f684-42de-93a4-edb2263ddb11&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=beb226ab-f684-42de-93a4-edb2263ddb11&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=beb226ab-f684-42de-93a4-edb2263ddb11&audience=consumer\n",
            "['Cemiplimab-rwlc']\n",
            "Cemiplimab-rwlc\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4347ae1f-d397-4f18-8b70-03897e1c054a\n",
            "\n",
            "\n",
            "LIBTAYO is indicated for the treatment of patients:\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4347ae1f-d397-4f18-8b70-03897e1c054a\n",
            "\n",
            "LIBTAYO is indicated for the treatment of patients with metastatic cutaneous squamous cell carcinoma (mCSCC) or locally advanced CSCC (laCSCC) who are not candidates for curative surgery or curative radiation.\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4347ae1f-d397-4f18-8b70-03897e1c054a\n",
            "\n",
            "\n",
            "LIBTAYO is indicated for the first-line treatment of patients with non-small cell lung cancer (NSCLC) whose tumors have high  PD-L1 expression [Tumor Proportion Score (TPS)  50%] as determined by an FDA-approved test [see Dosage and Administration (2.1)]), with no EGFR, ALK or ROS1 aberrations, and is:\n",
            "\n",
            "{'gene': ['EGFR', 'ALK', 'ROS1'], 'protein': ['EGFR', 'ALK'], 'drug': []}\n",
            "['Tepotinib Hydrochloride']\n",
            "Tepotinib Hydrochloride\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=80a0f1b9-071a-47f5-9e67-32d638a669dc&audience=consumer\n",
            "['Ibritumomab Tiuxetan']\n",
            "Ibritumomab Tiuxetan\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=ccf29429-bf8e-4cce-bd12-1de56853fc5b&audience=consumer\n",
            "['Denileukin Diftitox']\n",
            "Denileukin Diftitox\n",
            "['Brentuximab Vedotin']\n",
            "Brentuximab Vedotin\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3904f8dd-1aef-3490-e48f-bd55f32ed67f&audience=consumer\n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with previously untreated sALCL or other CD30-expressing PTCL, including angioimmunoblastic T-cell lymphoma and PTCL not otherwise specified, in combination with cyclophosphamide, doxorubicin, and prednisone.\n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with sALCL after failure of at least one prior multi-agent chemotherapy regimen.\n",
            "\n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with pcALCL or CD30-expressing MF who have received prior systemic therapy.\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3904f8dd-1aef-3490-e48f-bd55f32ed67f&audience=consumer\n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with pcALCL or CD30-expressing MF who have received prior systemic therapy.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3904f8dd-1aef-3490-e48f-bd55f32ed67f&audience=consumer\n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with previously untreated Stage III or IV cHL, in combination with doxorubicin, vinblastine, and dacarbazine.\n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with cHL at high risk of relapse or progression as post-autologous hematopoietic stem cell transplantation (auto-HSCT) consolidation.  \n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with cHL after failure of auto-HSCT or after failure of at least two prior multi-agent chemotherapy regimens in patients who are not auto-HSCT candidates.  \n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3904f8dd-1aef-3490-e48f-bd55f32ed67f&audience=consumer\n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with pcALCL or CD30-expressing MF who have received prior systemic therapy.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3904f8dd-1aef-3490-e48f-bd55f32ed67f&audience=consumer\n",
            "\n",
            "ADCETRIS is indicated for the treatment of adult patients with previously untreated sALCL or other CD30-expressing PTCL, including angioimmunoblastic T-cell lymphoma and PTCL not otherwise specified, in combination with cyclophosphamide, doxorubicin, and prednisone.\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "['Rituximab']\n",
            "Rituximab\n",
            "['Vorinostat']\n",
            "Vorinostat\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=cd86ee78-2781-468b-930c-3c4677bcc092&audience=consumer\n",
            "['Romidepsin']\n",
            "Romidepsin\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e0e86a20-23f8-4486-8e0c-5c4a268df231\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e0e86a20-23f8-4486-8e0c-5c4a268df231\n",
            "['Bexarotene']\n",
            "Bexarotene\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=Bexarotene&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=d7fc7c7b-2ecf-4543-8bb4-9b4415d66613&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e713c60d-f623-4260-b0ca-db07e1365fd2&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=dc8d68d1-0c5c-4004-b71d-22b83daf4270&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2d1eb4bd-51f9-4c9b-9fb1-0be5e30a3dc5&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7945daa3-b607-48c3-a94f-f3a28b033dad&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7437a43f-172a-4f4e-8132-fe23b00224f9&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e415c4c8-3144-4172-94a0-f4f3243130ae&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1d056725-0576-4338-8089-6336e768ccdc&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=d35b4697-3f3b-4aa1-b5d9-153993e83811&audience=consumer\n",
            "['Bortezomib']\n",
            "Bortezomib\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BORTEZOMIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=287ce008-3d9c-4914-acf0-01c372399b8d&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=042329d6-8077-4af2-816f-cd0a2bbcbbd3&audience=consumer\n",
            "\n",
            "Bortezomib for Injection is indicated for the treatment of patients with mantle cell lymphoma who have received at least 1 prior therapy.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1521d321-e724-4ffc-adad-34bf4f44fac7&audience=consumer\n",
            "\n",
            "VELCADE is indicated for the treatment of adult patients with mantle cell lymphoma.\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=BORTEZOMIB&pagesize=20&page=1&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=287ce008-3d9c-4914-acf0-01c372399b8d&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=042329d6-8077-4af2-816f-cd0a2bbcbbd3&audience=consumer\n",
            "\n",
            "Bortezomib for Injection is indicated for the treatment of patients with multiple myeloma.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1521d321-e724-4ffc-adad-34bf4f44fac7&audience=consumer\n",
            "\n",
            "VELCADE is indicated for the treatment of adult patients with multiple myeloma. \n",
            "['Pralatrexate']\n",
            "Pralatrexate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=5e4cf15b-bf7b-4b83-863e-e9ef27741a51&audience=consumer\n",
            "['Ibrutinib']\n",
            "Ibrutinib\n",
            "['Siltuximab']\n",
            "Siltuximab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=451f535b-8b6b-4ecf-9f19-d921b72eea39&audience=consumer\n",
            "['Idelalisib']\n",
            "Idelalisib\n",
            "['Belinostat']\n",
            "Belinostat\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=84b2e16e-f0d1-4757-8da8-79dfa83aab79&audience=consumer\n",
            "['Obinutuzumab']\n",
            "Obinutuzumab\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Rituximab and Hyaluronidase Human']\n",
            "Rituximab and Hyaluronidase Human\n",
            "['Copanlisib Hydrochloride']\n",
            "Copanlisib Hydrochloride\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=ade50241-2c10-4038-b4e9-72f6bf905f03&audience=consumer\n",
            "['Axicabtagene Ciloleucel']\n",
            "Axicabtagene Ciloleucel\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=9b70606e-b99c-4272-a0f1-b5523cce0c59&audience=consumer\n",
            "['Acalabrutinib']\n",
            "Acalabrutinib\n",
            "['Tisagenlecleucel']\n",
            "Tisagenlecleucel\n",
            "['Venetoclax']\n",
            "Venetoclax\n",
            "['Mogamulizumab-kpkc']\n",
            "Mogamulizumab-kpkc\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e53960ab-42a1-40d1-9c7d-eb013fe7f18f\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=e53960ab-42a1-40d1-9c7d-eb013fe7f18f\n",
            "['Duvelisib']\n",
            "Duvelisib\n",
            "['Polatuzumab Vedotin-piiq']\n",
            "Polatuzumab Vedotin-piiq\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=20a16ab2-f338-4abb-9dcd-254bd949a2bc&audience=consumer\n",
            "['Zanubrutinib']\n",
            "Zanubrutinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=3e08fe23-d70e-424c-bc51-1222e320f902\n",
            "['Tazemetostat Hydrobromide']\n",
            "Tazemetostat Hydrobromide\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7db07b5f-4e22-467c-9c0a-f830b08dbb1d&audience=consumer\n",
            "\n",
            "TAZVERIK is indicated for the treatment of adults and pediatric patients aged 16 years and older with metastatic or locally advanced epithelioid sarcoma not eligible for complete resection.\n",
            "\n",
            "This indication is approved under accelerated approval based on overall response rate and duration of response [see Clinical Studies (14.1)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in a confirmatory trial(s).\n",
            "\n",
            "{'gene': [], 'protein': [], 'drug': []}\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7db07b5f-4e22-467c-9c0a-f830b08dbb1d&audience=consumer\n",
            "\n",
            "These indications are approved under accelerated approval based on overall response rate and duration of response [see Clinical Studies (14.2)]. Continued approval for these indications may be contingent upon verification and description of clinical benefit in a confirmatory trial(s).\n",
            "\n",
            "['Selinexor']\n",
            "Selinexor\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f6dd2682-75a6-4863-90a8-a3197f6f78a8&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f6dd2682-75a6-4863-90a8-a3197f6f78a8&audience=consumer\n",
            "\n",
            "['Tafasitamab-cxix']\n",
            "Tafasitamab-cxix\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=ec13ac6b-bfde-4e84-907a-83bd69584d95&audience=consumer\n",
            "['Brexucabtagene Autoleucel']\n",
            "Brexucabtagene Autoleucel\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a16108c2-7ca7-45af-965e-54bda4713022&audience=consumer\n",
            "['Crizotinib']\n",
            "Crizotinib\n",
            "['Umbralisib Tosylate']\n",
            "Umbralisib Tosylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2cdecefe-c9e4-4af3-bbaa-a1111c768d29&audience=consumer\n",
            "\n",
            "UKONIQ is indicated for the treatment of adult patients with relapsed or refractory marginal zone lymphoma (MZL) who have received at least one prior anti-CD20-based regimen.\n",
            "\n",
            "This indication is approved under accelerated approval based on overall response rate [see Clinical Studies (14.1)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in a confirmatory trial(s).\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=2cdecefe-c9e4-4af3-bbaa-a1111c768d29&audience=consumer\n",
            "\n",
            "UKONIQ is indicated for the treatment of adult patients with relapsed or refractory follicular lymphoma (FL) who have received at least three prior lines of systemic therapy.\n",
            "\n",
            "This indication is approved under accelerated approval based on overall response rate [see Clinical Studies (14.2)]. Continued approval for this indication may be contingent upon verification and description of clinical benefit in a confirmatory trial(s).\n",
            "\n",
            "['Lisocabtagene Maraleucel']\n",
            "Lisocabtagene Maraleucel\n",
            "['Ipilimumab']\n",
            "Ipilimumab\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Bortezomib']\n",
            "Bortezomib\n",
            "['Carfilzomib']\n",
            "Carfilzomib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=ea66eb30-e665-4693-99a1-a9d3b4bbe2d6&audience=consumer\n",
            "\n",
            "['Panobinostat Lactate']\n",
            "Panobinostat Lactate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7774972a-eeaa-4b9a-9e56-3fc1b968e86a&audience=consumer\n",
            "['Daratumumab']\n",
            "Daratumumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a4d0efe9-5e54-467e-9eb4-56fa7d53b60b&audience=consumer\n",
            "['Ixazomib Citrate']\n",
            "Ixazomib Citrate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=fcef9088-ebab-4bd8-933f-c35f9c8bd50b&audience=consumer\n",
            "['Elotuzumab']\n",
            "Elotuzumab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=80686b7e-f6f4-4154-b5c0-c846425e2d91&audience=consumer\n",
            "['Selinexor']\n",
            "Selinexor\n",
            "['Isatuximab-irfc']\n",
            "Isatuximab-irfc\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a0473462-6f9d-4eca-a5bf-8620aea68e8a\n",
            "['Daratumumab and Hyaluronidase-fihj']\n",
            "Daratumumab and Hyaluronidase-fihj\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4bb241af-4299-4373-8762-2d6709515db0\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4bb241af-4299-4373-8762-2d6709515db0\n",
            "\n",
            "DARZALEX FASPRO is indicated for the treatment of adult patients with multiple myeloma:\n",
            "['Belantamab Mafodotin-blmf']\n",
            "Belantamab Mafodotin-blmf\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=16a160a4-3ec0-4ddf-99ce-05912dd3382d&audience=consumer\n",
            "['Imatinib Mesylate']\n",
            "Imatinib Mesylate\n",
            "['Ruxolitinib Phosphate']\n",
            "Ruxolitinib Phosphate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f1c82580-87ae-11e0-bc84-0002a5d5c51b&audience=consumer\n",
            "\n",
            "Jakafi is indicated for treatment of intermediate or high-risk myelofibrosis (MF), including primary MF, post-polycythemia vera MF and post-essential thrombocythemia MF in adults.\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f1c82580-87ae-11e0-bc84-0002a5d5c51b&audience=consumer\n",
            "\n",
            "Jakafi is indicated for treatment of polycythemia vera (PV) in adults who have had an inadequate response to or are intolerant of hydroxyurea.\n",
            "['Fedratinib Hydrochloride']\n",
            "Fedratinib Hydrochloride\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=f0f55a2a-4e0c-4cba-8571-03e1424486d7&audience=consumer\n",
            "['Dinutuximab']\n",
            "Dinutuximab\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=d66bdf0d-9d65-45de-ae5b-a58617c27492&audience=consumer\n",
            "['Naxitamab-gqgk']\n",
            "Naxitamab-gqgk\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=29a80c6b-8bad-4650-8c7f-f18490c868ec&audience=consumer\n",
            "['Bevacizumab']\n",
            "Bevacizumab\n",
            "['Olaparib']\n",
            "Olaparib\n",
            "['Rucaparib Camsylate']\n",
            "Rucaparib Camsylate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a6d46c03-bb1d-417b-b8e5-3bffe352fe29&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a6d46c03-bb1d-417b-b8e5-3bffe352fe29&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a6d46c03-bb1d-417b-b8e5-3bffe352fe29&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a6d46c03-bb1d-417b-b8e5-3bffe352fe29&audience=consumer\n",
            "['Niraparib Tosylate Monohydrate']\n",
            "Niraparib Tosylate Monohydrate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0fe9d533-67a9-4981-978a-1e84755ae30b&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0fe9d533-67a9-4981-978a-1e84755ae30b&audience=consumer\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=0fe9d533-67a9-4981-978a-1e84755ae30b&audience=consumer\n",
            "['Erlotinib Hydrochloride']\n",
            "Erlotinib Hydrochloride\n",
            "['Everolimus']\n",
            "Everolimus\n",
            "['Sunitinib Malate']\n",
            "Sunitinib Malate\n",
            "['Olaparib']\n",
            "Olaparib\n",
            "['Selumetinib Sulfate']\n",
            "Selumetinib Sulfate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=7d042c61-f28f-4ab5-ab10-d7558c0d49ff\n",
            "['Cabazitaxel']\n",
            "Cabazitaxel\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=de3d9c26-572b-4ea4-9b2d-dd58a2b3e8fa&audience=consumer\n",
            "['Enzalutamide']\n",
            "Enzalutamide\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=b129fdc9-1d8e-425c-a5a9-8a2ed36dfbdf&audience=consumer\n",
            "['Abiraterone Acetate']\n",
            "Abiraterone Acetate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=4e338e89-3cf2-48eb-b6e2-a06c608c6513&audience=consumer\n",
            "['Radium 223 Dichloride']\n",
            "Radium 223 Dichloride\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=a398400e-bd31-41a9-9696-4f7c06569ede&audience=consumer\n",
            "['Apalutamide']\n",
            "Apalutamide\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=d1cda4f7-cb33-46ea-b9ac-431f6452b1a5&audience=consumer\n",
            "['Darolutamide']\n",
            "Darolutamide\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=1a7cb212-56e4-4b9d-a73d-bfee7fe4735e\n",
            "['Rucaparib Camsylate']\n",
            "Rucaparib Camsylate\n",
            "['Rucaparib Camsylate']\n",
            "Rucaparib Camsylate\n",
            "['Olaparib']\n",
            "Olaparib\n",
            "['Vismodegib']\n",
            "Vismodegib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=eb368bb6-80e3-4df9-8a85-91df0a2ada6a&audience=consumer\n",
            "['Sonidegib']\n",
            "Sonidegib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=028312dc-d155-4fd5-8abd-6bb9f011d3cc&audience=consumer\n",
            "['Ipilimumab']\n",
            "Ipilimumab\n",
            "['Vemurafenib']\n",
            "Vemurafenib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=38eea320-7e0c-485a-bc30-98c3c45e2763&audience=consumer\n",
            "\n",
            "\n",
            "ZELBORAF is indicated for the treatment of patients with Erdheim-Chester Disease (ECD) with BRAF V600 mutation.\n",
            "\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=38eea320-7e0c-485a-bc30-98c3c45e2763&audience=consumer\n",
            "\n",
            "ZELBORAF is indicated for the treatment of patients with unresectable or metastatic melanoma with BRAF V600E mutation as detected by an FDA-approved test.\n",
            "['Trametinib Dimethyl Sulfoxide']\n",
            "Trametinib Dimethyl Sulfoxide\n",
            "['Dabrafenib Mesylate']\n",
            "Dabrafenib Mesylate\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Nivolumab']\n",
            "Nivolumab\n",
            "['Cobimetinib Fumarate']\n",
            "Cobimetinib Fumarate\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=c387579e-cee0-4334-bd1e-73f93ac1bde6&audience=consumer\n",
            "[]\n",
            "['Avelumab']\n",
            "Avelumab\n",
            "['Encorafenib']\n",
            "Encorafenib\n",
            "['Binimetinib']\n",
            "Binimetinib\n",
            "https://dailymed.nlm.nih.gov/dailymed/drugInfo.cfm?setid=6c3408ac-d401-4925-8a03-26591afbc240&audience=consumer\n",
            "['Cemiplimab-rwlc']\n",
            "Cemiplimab-rwlc\n",
            "['Atezolizumab']\n",
            "Atezolizumab\n",
            "['Pazopanib Hydrochloride']\n",
            "Pazopanib Hydrochloride\n",
            "[]\n",
            "['Tazemetostat Hydrobromide']\n",
            "Tazemetostat Hydrobromide\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Larotrectinib Sulfate']\n",
            "Larotrectinib Sulfate\n",
            "['Entrectinib']\n",
            "Entrectinib\n",
            "['Pembrolizumab']\n",
            "Pembrolizumab\n",
            "['Trastuzumab']\n",
            "Trastuzumab\n",
            "['Ramucirumab']\n",
            "Ramucirumab\n",
            "['Fam-Trastuzumab Deruxtecan-nxki']\n",
            "Fam-Trastuzumab Deruxtecan-nxki\n",
            "['Imatinib Mesylate']\n",
            "Imatinib Mesylate\n",
            "['Midostaurin']\n",
            "Midostaurin\n",
            "['Cabozantinib-S-Malate']\n",
            "Cabozantinib-S-Malate\n",
            "['Vandetanib']\n",
            "Vandetanib\n",
            "multiple choice\n",
            "https://dailymed.nlm.nih.gov/dailymed/search.cfm?labeltype=all&query=VANDETANIB&pagesize=20&page=1&audience=consumer\n",
            "['Sorafenib Tosylate']\n",
            "Sorafenib Tosylate\n",
            "['Lenvatinib Mesylate']\n",
            "Lenvatinib Mesylate\n",
            "['Trametinib Dimethyl Sulfoxide']\n",
            "Trametinib Dimethyl Sulfoxide\n",
            "['Selpercatinib']\n",
            "Selpercatinib\n",
            "['Pralsetinib']\n",
            "Pralsetinib\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TiribISKP0v-",
        "outputId": "20887961-23ca-44a8-eb6c-ab03f80d66ca"
      },
      "source": [
        "len(therapy_list), len(disease_label), len(NDCCode_label), len(geneProtein_label), len(drug_label), len(comDrug_label), len(negGeProDr_label)"
      ],
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(147, 213, 213, 213, 213, 213, 213)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 29
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "Vh9Itg5sP0wC",
        "outputId": "67d58a60-ce34-4bb6-fb40-745b0b4bf111"
      },
      "source": [
        "# put the labels into a dataFrame\n",
        "dict = {'geneProtein_label': geneProtein_label, 'drug_label': drug_label, 'NDCCode_label': NDCCode_label, 'disease_label': disease_label, 'comDrug_label':comDrug_label, 'negGeProDr_label':negGeProDr_label}  \n",
        "targetTherapyDF = pd.DataFrame(dict)\n",
        "targetTherapyDF[:50]"
      ],
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>geneProtein_label</th>\n",
              "      <th>drug_label</th>\n",
              "      <th>NDCCode_label</th>\n",
              "      <th>disease_label</th>\n",
              "      <th>comDrug_label</th>\n",
              "      <th>negGeProDr_label</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>PD-L1, PD-L1, TECENTRIQ, PD-L1, PD-L1</td>\n",
              "      <td>Atezolizumab (TECENTRIQ- atezolizumab injectio...</td>\n",
              "      <td>50242-917-01, 50242-917-86, 50242-918-01, 5024...</td>\n",
              "      <td>Breast cancer</td>\n",
              "      <td>paclitaxel</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>TECENTRIQ</td>\n",
              "      <td>Atezolizumab (TECENTRIQ- atezolizumab injectio...</td>\n",
              "      <td>50242-917-01, 50242-917-86, 50242-918-01, 5024...</td>\n",
              "      <td>Hepatocellular carcinoma</td>\n",
              "      <td>bevacizumab</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>BRAF, TECENTRIQ</td>\n",
              "      <td>Atezolizumab (TECENTRIQ- atezolizumab injectio...</td>\n",
              "      <td>50242-917-01, 50242-917-86, 50242-918-01, 5024...</td>\n",
              "      <td>Melanoma</td>\n",
              "      <td>vemurafenib</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>TECENTRIQ</td>\n",
              "      <td>Atezolizumab (TECENTRIQ- atezolizumab injectio...</td>\n",
              "      <td>50242-917-01, 50242-917-86, 50242-918-01, 5024...</td>\n",
              "      <td>Small cell lung cancer</td>\n",
              "      <td>carboplatin, etoposide</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>TECENTRIQ</td>\n",
              "      <td>Atezolizumab (TECENTRIQ- atezolizumab injectio...</td>\n",
              "      <td>50242-917-01, 50242-917-86, 50242-918-01, 5024...</td>\n",
              "      <td>Urothelial carcinoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>MSI-H, OPDIVO</td>\n",
              "      <td>Nivolumab (OPDIVO- nivolumab injection)</td>\n",
              "      <td>0003-3734-13, 0003-3772-11, 0003-3774-12</td>\n",
              "      <td>Colorectal cancer</td>\n",
              "      <td>ipilimumab</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>OPDIVO</td>\n",
              "      <td>Nivolumab (OPDIVO- nivolumab injection)</td>\n",
              "      <td>0003-3734-13, 0003-3772-11, 0003-3774-12</td>\n",
              "      <td>Hepatocellular carcinoma</td>\n",
              "      <td>ipilimumab</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>OPDIVO</td>\n",
              "      <td>Nivolumab (OPDIVO- nivolumab injection)</td>\n",
              "      <td>0003-3734-13, 0003-3772-11, 0003-3774-12</td>\n",
              "      <td>Malignant pleural mesothelioma</td>\n",
              "      <td>ipilimumab</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>OPDIVO</td>\n",
              "      <td>Nivolumab (OPDIVO- nivolumab injection)</td>\n",
              "      <td>0003-3734-13, 0003-3772-11, 0003-3774-12</td>\n",
              "      <td>Melanoma</td>\n",
              "      <td>ipilimumab</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCCHN, OPDIVO</td>\n",
              "      <td>Nivolumab (OPDIVO- nivolumab injection)</td>\n",
              "      <td>0003-3734-13, 0003-3772-11, 0003-3774-12</td>\n",
              "      <td>Squamous cell carcinoma of the head and neck</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>OPDIVO</td>\n",
              "      <td>Nivolumab (OPDIVO- nivolumab injection)</td>\n",
              "      <td>0003-3734-13, 0003-3772-11, 0003-3774-12</td>\n",
              "      <td>Urothelial carcinoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>BAVENCIO</td>\n",
              "      <td>Avelumab (BAVENCIO- avelumab injection, soluti...</td>\n",
              "      <td>44087-3535-1</td>\n",
              "      <td>Merkel cell carcinoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>BAVENCIO</td>\n",
              "      <td>Avelumab (BAVENCIO- avelumab injection, soluti...</td>\n",
              "      <td>44087-3535-1</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td>axitinib</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>PD-L1, PD-L1, CPS 1</td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Cervical cancer</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>MSI-H</td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Endometrial carcinoma</td>\n",
              "      <td>lenvatinib</td>\n",
              "      <td>MSI-H</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td></td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Hepatocellular carcinoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td></td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Melanoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td></td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Merkel cell carcinoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>EGFR, ALK genomic, PD-L1, TPS, EGFR, ALK genom...</td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Non-small cell lung cancer</td>\n",
              "      <td>pemetrexed, platinum, carboplatin, paclitaxel,...</td>\n",
              "      <td>EGFR, ALK genomic, EGFR</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td></td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Primary mediastinal large B-cell lymphoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td></td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td>axitinib</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>PD-L1, PD-L1, PD-L1, CPS 10, PD-L1, CIS</td>\n",
              "      <td>Pembrolizumab (KEYTRUDA- pembrolizumab injecti...</td>\n",
              "      <td>0006-3026-01, 0006-3026-02, 0006-3026-04, 0006...</td>\n",
              "      <td>Urothelial carcinoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (AVASTIN- bevacizumab injection, s...</td>\n",
              "      <td>50242-060-01, 50242-060-10, 50242-061-01, 5024...</td>\n",
              "      <td>Cervical cancer</td>\n",
              "      <td>paclitaxel, cisplatin, paclitaxel, topotecan</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (MVASI- bevacizumab-awwb injection...</td>\n",
              "      <td>55513-206-01, 55513-207-01</td>\n",
              "      <td>Cervical cancer</td>\n",
              "      <td>paclitaxel, cisplatin, paclitaxel, topotecan, ...</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (ZIRABEV- bevacizumab-bvzr injecti...</td>\n",
              "      <td>0069-0315-01, 0069-0342-01</td>\n",
              "      <td>Cervical cancer</td>\n",
              "      <td>paclitaxel, cisplatin, paclitaxel, topotecan, ...</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (AVASTIN- bevacizumab injection, s...</td>\n",
              "      <td>50242-060-01, 50242-060-10, 50242-061-01, 5024...</td>\n",
              "      <td>Colorectal cancer</td>\n",
              "      <td>intravenous fluorouracil-based</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (MVASI- bevacizumab-awwb injection...</td>\n",
              "      <td>55513-206-01, 55513-207-01</td>\n",
              "      <td>Colorectal cancer</td>\n",
              "      <td>intravenous fluorouracil-based, intravenous fl...</td>\n",
              "      <td>MVASI</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (ZIRABEV- bevacizumab-bvzr injecti...</td>\n",
              "      <td>0069-0315-01, 0069-0342-01</td>\n",
              "      <td>Colorectal cancer</td>\n",
              "      <td>intravenous fluorouracil-based, intravenous fl...</td>\n",
              "      <td>MVASI</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (AVASTIN- bevacizumab injection, s...</td>\n",
              "      <td>50242-060-01, 50242-060-10, 50242-061-01, 5024...</td>\n",
              "      <td>Glioblastoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (MVASI- bevacizumab-awwb injection...</td>\n",
              "      <td>55513-206-01, 55513-207-01</td>\n",
              "      <td>Glioblastoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (ZIRABEV- bevacizumab-bvzr injecti...</td>\n",
              "      <td>0069-0315-01, 0069-0342-01</td>\n",
              "      <td>Glioblastoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (AVASTIN- bevacizumab injection, s...</td>\n",
              "      <td>50242-060-01, 50242-060-10, 50242-061-01, 5024...</td>\n",
              "      <td>Hepatocellular carcinoma</td>\n",
              "      <td>atezolizumab</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (AVASTIN- bevacizumab injection, s...</td>\n",
              "      <td>50242-060-01, 50242-060-10, 50242-061-01, 5024...</td>\n",
              "      <td>fallopian tube</td>\n",
              "      <td>carboplatin, paclitaxel, Avastin, paclitaxel, ...</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (ZIRABEV- bevacizumab-bvzr injecti...</td>\n",
              "      <td>0069-0315-01, 0069-0342-01</td>\n",
              "      <td>fallopian tube</td>\n",
              "      <td>carboplatin, paclitaxel, Avastin, paclitaxel, ...</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (AVASTIN- bevacizumab injection, s...</td>\n",
              "      <td>50242-060-01, 50242-060-10, 50242-061-01, 5024...</td>\n",
              "      <td>primary peritoneal cancer</td>\n",
              "      <td>carboplatin, paclitaxel, Avastin, paclitaxel, ...</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td></td>\n",
              "      <td>Bevacizumab (ZIRABEV- bevacizumab-bvzr injecti...</td>\n",
              "      <td>0069-0315-01, 0069-0342-01</td>\n",
              "      <td>primary peritoneal cancer</td>\n",
              "      <td>carboplatin, paclitaxel, Avastin, paclitaxel, ...</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>interferon alfa, interferon alfa</td>\n",
              "      <td>Bevacizumab (AVASTIN- bevacizumab injection, s...</td>\n",
              "      <td>50242-060-01, 50242-060-10, 50242-061-01, 5024...</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>interferon alfa, interferon alfa, interferon-a...</td>\n",
              "      <td>Bevacizumab (MVASI- bevacizumab-awwb injection...</td>\n",
              "      <td>55513-206-01, 55513-207-01</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td>interferon-alfa</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>interferon alfa, interferon alfa, interferon-a...</td>\n",
              "      <td>Bevacizumab (ZIRABEV- bevacizumab-bvzr injecti...</td>\n",
              "      <td>0069-0315-01, 0069-0342-01</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td>interferon-alfa</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>0378-3096-28, 0378-3096-32, 0378-3096-85, 0378...</td>\n",
              "      <td>Breast cancer</td>\n",
              "      <td>exemestane</td>\n",
              "      <td>letrozole</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>0093-7766-19, 0093-7766-24, 0093-7767-19, 0093...</td>\n",
              "      <td>Breast cancer</td>\n",
              "      <td>exemestane, exemestane</td>\n",
              "      <td>letrozole, letrozole</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>AFINITOR, </td>\n",
              "      <td>Everolimus (AFINITOR- everolimus tablet)</td>\n",
              "      <td>0078-0566-51, 0078-0566-61, 0078-0567-51, 0078...</td>\n",
              "      <td>Breast cancer</td>\n",
              "      <td>exemestane, exemestane, exemestane</td>\n",
              "      <td>letrozole, letrozole, letrozole</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>51991-821-28, 51991-821-99, 51991-822-28, 5199...</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td></td>\n",
              "      <td>sunitinib</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>0054-0480-13, 0054-0480-14, 0054-0480-22, 0054...</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td></td>\n",
              "      <td>sunitinib, sunitinib</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>63850-0058-1, 63850-0059-1, 63850-0060-1, 6385...</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td></td>\n",
              "      <td>sunitinib, sunitinib, sunitinib</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>0093-7766-19, 0093-7766-24, 0093-7767-19, 0093...</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td></td>\n",
              "      <td>sunitinib, sunitinib, sunitinib, sunitinib</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>AFINITOR, sunitinib</td>\n",
              "      <td>Everolimus (AFINITOR- everolimus tablet)</td>\n",
              "      <td>0078-0566-51, 0078-0566-61, 0078-0567-51, 0078...</td>\n",
              "      <td>Renal cell carcinoma</td>\n",
              "      <td></td>\n",
              "      <td>sunitinib, sunitinib, sunitinib, sunitinib, su...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>51991-821-28, 51991-821-99, 51991-822-28, 5199...</td>\n",
              "      <td>Subependymal giant cell astrocytoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>0378-3096-28, 0378-3096-32, 0378-3096-85, 0378...</td>\n",
              "      <td>Subependymal giant cell astrocytoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>49</th>\n",
              "      <td></td>\n",
              "      <td>Everolimus (EVEROLIMUS tablet)</td>\n",
              "      <td>63850-0058-1, 63850-0059-1, 63850-0060-1, 6385...</td>\n",
              "      <td>Subependymal giant cell astrocytoma</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                    geneProtein_label  ...                                   negGeProDr_label\n",
              "0               PD-L1, PD-L1, TECENTRIQ, PD-L1, PD-L1  ...                                                   \n",
              "1                                           TECENTRIQ  ...                                                   \n",
              "2                                     BRAF, TECENTRIQ  ...                                                   \n",
              "3                                           TECENTRIQ  ...                                                   \n",
              "4                                           TECENTRIQ  ...                                                   \n",
              "5                                       MSI-H, OPDIVO  ...                                                   \n",
              "6                                              OPDIVO  ...                                                   \n",
              "7                                              OPDIVO  ...                                                   \n",
              "8                                              OPDIVO  ...                                                   \n",
              "9                                       SCCHN, OPDIVO  ...                                                   \n",
              "10                                             OPDIVO  ...                                                   \n",
              "11                                           BAVENCIO  ...                                                   \n",
              "12                                           BAVENCIO  ...                                                   \n",
              "13                               PD-L1, PD-L1, CPS 1  ...                                                   \n",
              "14                                              MSI-H  ...                                              MSI-H\n",
              "15                                                     ...                                                   \n",
              "16                                                    ...                                                   \n",
              "17                                                     ...                                                   \n",
              "18  EGFR, ALK genomic, PD-L1, TPS, EGFR, ALK genom...  ...                            EGFR, ALK genomic, EGFR\n",
              "19                                                     ...                                                   \n",
              "20                                                     ...                                                   \n",
              "21           PD-L1, PD-L1, PD-L1, CPS 10, PD-L1, CIS  ...                                                   \n",
              "22                                                     ...                                                   \n",
              "23                                                     ...                                                   \n",
              "24                                                     ...                                                   \n",
              "25                                                     ...                                                   \n",
              "26                                                     ...                                              MVASI\n",
              "27                                                     ...                                              MVASI\n",
              "28                                                     ...                                                   \n",
              "29                                                     ...                                                   \n",
              "30                                                     ...                                                   \n",
              "31                                                     ...                                                   \n",
              "32                                                     ...                                                   \n",
              "33                                                     ...                                                   \n",
              "34                                                     ...                                                   \n",
              "35                                                     ...                                                   \n",
              "36                   interferon alfa, interferon alfa  ...                                                   \n",
              "37  interferon alfa, interferon alfa, interferon-a...  ...                                                   \n",
              "38  interferon alfa, interferon alfa, interferon-a...  ...                                                   \n",
              "39                                                     ...                                          letrozole\n",
              "40                                                     ...                               letrozole, letrozole\n",
              "41                                        AFINITOR,   ...                    letrozole, letrozole, letrozole\n",
              "42                                                     ...                                          sunitinib\n",
              "43                                                     ...                               sunitinib, sunitinib\n",
              "44                                                     ...                    sunitinib, sunitinib, sunitinib\n",
              "45                                                     ...         sunitinib, sunitinib, sunitinib, sunitinib\n",
              "46                                AFINITOR, sunitinib  ...  sunitinib, sunitinib, sunitinib, sunitinib, su...\n",
              "47                                                     ...                                                   \n",
              "48                                                     ...                                                   \n",
              "49                                                     ...                                                   \n",
              "\n",
              "[50 rows x 6 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 30
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "yd-lI82GP0wC"
      },
      "source": [
        "# save the data frame\n",
        "#targetTherapyDF.to_csv(r'/content/drive/MyDrive/RA_NLP_drug_label/2021-04-12 target therapy labels.csv')"
      ],
      "execution_count": 31,
      "outputs": []
    }
  ]
}